{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cae7ef1a",
   "metadata": {},
   "source": [
    "# Integrative analysis of ATAC & RNA - Correlation analysis\n",
    "- goal: visualize epigenetic potential & relative transcriptional abundance\n",
    "- input: raw counts, normalized, batch corrected and fitted data\n",
    "- output: correlation between expression and accessibility across stimuli and time points"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "481363c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/sreichl/projects/bmdm-stim\n"
     ]
    }
   ],
   "source": [
    "cd ../"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "eae75315",
   "metadata": {},
   "outputs": [],
   "source": [
    "# libraries\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.patches import Patch\n",
    "import numpy as np\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cf28ccbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# configs\n",
    "dir_data = os.path.join('results','INT')\n",
    "dir_results=os.path.join('results','INT','epigenetic_correlation')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2b033a84",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(dir_results):\n",
    "        os.mkdir(dir_results)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6137d19",
   "metadata": {},
   "source": [
    "# load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7037c585",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(142, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>library</th>\n",
       "      <th>treatment</th>\n",
       "      <th>time</th>\n",
       "      <th>experiment</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RNA_PT76_R1_C_albicans_2h</th>\n",
       "      <td>Quant-seq</td>\n",
       "      <td>C_albicans</td>\n",
       "      <td>2h</td>\n",
       "      <td>PT76</td>\n",
       "      <td>C_albicans_2h</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA_PT82_R1_C_albicans_24h</th>\n",
       "      <td>Quant-seq</td>\n",
       "      <td>C_albicans</td>\n",
       "      <td>24h</td>\n",
       "      <td>PT82</td>\n",
       "      <td>C_albicans_24h</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA_PT76_R1_C_albicans_4h</th>\n",
       "      <td>Quant-seq</td>\n",
       "      <td>C_albicans</td>\n",
       "      <td>4h</td>\n",
       "      <td>PT76</td>\n",
       "      <td>C_albicans_4h</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA_PT76_R1_C_albicans_6h</th>\n",
       "      <td>Quant-seq</td>\n",
       "      <td>C_albicans</td>\n",
       "      <td>6h</td>\n",
       "      <td>PT76</td>\n",
       "      <td>C_albicans_6h</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA_PT82_R1_C_albicans_8h</th>\n",
       "      <td>Quant-seq</td>\n",
       "      <td>C_albicans</td>\n",
       "      <td>8h</td>\n",
       "      <td>PT82</td>\n",
       "      <td>C_albicans_8h</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              library   treatment time experiment  \\\n",
       "RNA_PT76_R1_C_albicans_2h   Quant-seq  C_albicans   2h       PT76   \n",
       "RNA_PT82_R1_C_albicans_24h  Quant-seq  C_albicans  24h       PT82   \n",
       "RNA_PT76_R1_C_albicans_4h   Quant-seq  C_albicans   4h       PT76   \n",
       "RNA_PT76_R1_C_albicans_6h   Quant-seq  C_albicans   6h       PT76   \n",
       "RNA_PT82_R1_C_albicans_8h   Quant-seq  C_albicans   8h       PT82   \n",
       "\n",
       "                                     group  \n",
       "RNA_PT76_R1_C_albicans_2h    C_albicans_2h  \n",
       "RNA_PT82_R1_C_albicans_24h  C_albicans_24h  \n",
       "RNA_PT76_R1_C_albicans_4h    C_albicans_4h  \n",
       "RNA_PT76_R1_C_albicans_6h    C_albicans_6h  \n",
       "RNA_PT82_R1_C_albicans_8h    C_albicans_8h  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load sample annotation\n",
    "annot = pd.read_csv(os.path.join(dir_data, 'INT_annotations.csv'), index_col=0, header=0,)\n",
    "print(annot.shape)\n",
    "annot.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ccf735ed",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(45069, 142)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RNA_PT76_R1_C_albicans_2h</th>\n",
       "      <th>RNA_PT82_R1_C_albicans_24h</th>\n",
       "      <th>RNA_PT76_R1_C_albicans_4h</th>\n",
       "      <th>RNA_PT76_R1_C_albicans_6h</th>\n",
       "      <th>RNA_PT82_R1_C_albicans_8h</th>\n",
       "      <th>RNA_PT76_R1_untreated_0h</th>\n",
       "      <th>RNA_PT82_R1_untreated_24h</th>\n",
       "      <th>RNA_PT76_R1_IFN_beta_2h</th>\n",
       "      <th>RNA_PT82_R1_IFN_beta_24h</th>\n",
       "      <th>RNA_PT76_R1_IFN_beta_4h</th>\n",
       "      <th>...</th>\n",
       "      <th>PT82_R2_IFN_gamma_24h</th>\n",
       "      <th>PT82_R2_IFN_gamma_8h</th>\n",
       "      <th>PT82_R2_LCMV_Cl13_24h</th>\n",
       "      <th>PT82_R2_LCMV_Cl13_8h</th>\n",
       "      <th>PT82_R2_LO28_24h</th>\n",
       "      <th>PT82_R2_LO28_8h</th>\n",
       "      <th>PT82_R2_LPS_24h</th>\n",
       "      <th>PT82_R2_LPS_8h</th>\n",
       "      <th>PT82_R2_untreated_24h</th>\n",
       "      <th>PT82_R2_untreated_8h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000098104</th>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>11</td>\n",
       "      <td>31</td>\n",
       "      <td>14</td>\n",
       "      <td>5</td>\n",
       "      <td>12</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000033845</th>\n",
       "      <td>219</td>\n",
       "      <td>220</td>\n",
       "      <td>201</td>\n",
       "      <td>169</td>\n",
       "      <td>196</td>\n",
       "      <td>204</td>\n",
       "      <td>226</td>\n",
       "      <td>202</td>\n",
       "      <td>124</td>\n",
       "      <td>153</td>\n",
       "      <td>...</td>\n",
       "      <td>210</td>\n",
       "      <td>109</td>\n",
       "      <td>218</td>\n",
       "      <td>185</td>\n",
       "      <td>32</td>\n",
       "      <td>164</td>\n",
       "      <td>127</td>\n",
       "      <td>155</td>\n",
       "      <td>156</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000025903</th>\n",
       "      <td>115</td>\n",
       "      <td>105</td>\n",
       "      <td>106</td>\n",
       "      <td>113</td>\n",
       "      <td>87</td>\n",
       "      <td>144</td>\n",
       "      <td>128</td>\n",
       "      <td>159</td>\n",
       "      <td>98</td>\n",
       "      <td>137</td>\n",
       "      <td>...</td>\n",
       "      <td>366</td>\n",
       "      <td>204</td>\n",
       "      <td>351</td>\n",
       "      <td>328</td>\n",
       "      <td>65</td>\n",
       "      <td>192</td>\n",
       "      <td>282</td>\n",
       "      <td>150</td>\n",
       "      <td>255</td>\n",
       "      <td>205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000033813</th>\n",
       "      <td>63</td>\n",
       "      <td>57</td>\n",
       "      <td>39</td>\n",
       "      <td>74</td>\n",
       "      <td>51</td>\n",
       "      <td>53</td>\n",
       "      <td>68</td>\n",
       "      <td>58</td>\n",
       "      <td>63</td>\n",
       "      <td>124</td>\n",
       "      <td>...</td>\n",
       "      <td>701</td>\n",
       "      <td>414</td>\n",
       "      <td>674</td>\n",
       "      <td>625</td>\n",
       "      <td>113</td>\n",
       "      <td>424</td>\n",
       "      <td>459</td>\n",
       "      <td>357</td>\n",
       "      <td>550</td>\n",
       "      <td>485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000033793</th>\n",
       "      <td>315</td>\n",
       "      <td>222</td>\n",
       "      <td>309</td>\n",
       "      <td>250</td>\n",
       "      <td>312</td>\n",
       "      <td>203</td>\n",
       "      <td>277</td>\n",
       "      <td>266</td>\n",
       "      <td>222</td>\n",
       "      <td>218</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>12</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 142 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    RNA_PT76_R1_C_albicans_2h  RNA_PT82_R1_C_albicans_24h  \\\n",
       "ENSMUSG00000098104                          1                           6   \n",
       "ENSMUSG00000033845                        219                         220   \n",
       "ENSMUSG00000025903                        115                         105   \n",
       "ENSMUSG00000033813                         63                          57   \n",
       "ENSMUSG00000033793                        315                         222   \n",
       "\n",
       "                    RNA_PT76_R1_C_albicans_4h  RNA_PT76_R1_C_albicans_6h  \\\n",
       "ENSMUSG00000098104                          0                          2   \n",
       "ENSMUSG00000033845                        201                        169   \n",
       "ENSMUSG00000025903                        106                        113   \n",
       "ENSMUSG00000033813                         39                         74   \n",
       "ENSMUSG00000033793                        309                        250   \n",
       "\n",
       "                    RNA_PT82_R1_C_albicans_8h  RNA_PT76_R1_untreated_0h  \\\n",
       "ENSMUSG00000098104                          1                         9   \n",
       "ENSMUSG00000033845                        196                       204   \n",
       "ENSMUSG00000025903                         87                       144   \n",
       "ENSMUSG00000033813                         51                        53   \n",
       "ENSMUSG00000033793                        312                       203   \n",
       "\n",
       "                    RNA_PT82_R1_untreated_24h  RNA_PT76_R1_IFN_beta_2h  \\\n",
       "ENSMUSG00000098104                          4                        6   \n",
       "ENSMUSG00000033845                        226                      202   \n",
       "ENSMUSG00000025903                        128                      159   \n",
       "ENSMUSG00000033813                         68                       58   \n",
       "ENSMUSG00000033793                        277                      266   \n",
       "\n",
       "                    RNA_PT82_R1_IFN_beta_24h  RNA_PT76_R1_IFN_beta_4h  ...  \\\n",
       "ENSMUSG00000098104                         4                        0  ...   \n",
       "ENSMUSG00000033845                       124                      153  ...   \n",
       "ENSMUSG00000025903                        98                      137  ...   \n",
       "ENSMUSG00000033813                        63                      124  ...   \n",
       "ENSMUSG00000033793                       222                      218  ...   \n",
       "\n",
       "                    PT82_R2_IFN_gamma_24h  PT82_R2_IFN_gamma_8h  \\\n",
       "ENSMUSG00000098104                      8                     7   \n",
       "ENSMUSG00000033845                    210                   109   \n",
       "ENSMUSG00000025903                    366                   204   \n",
       "ENSMUSG00000033813                    701                   414   \n",
       "ENSMUSG00000033793                      0                     5   \n",
       "\n",
       "                    PT82_R2_LCMV_Cl13_24h  PT82_R2_LCMV_Cl13_8h  \\\n",
       "ENSMUSG00000098104                      3                    11   \n",
       "ENSMUSG00000033845                    218                   185   \n",
       "ENSMUSG00000025903                    351                   328   \n",
       "ENSMUSG00000033813                    674                   625   \n",
       "ENSMUSG00000033793                      6                     4   \n",
       "\n",
       "                    PT82_R2_LO28_24h  PT82_R2_LO28_8h  PT82_R2_LPS_24h  \\\n",
       "ENSMUSG00000098104                31               14                5   \n",
       "ENSMUSG00000033845                32              164              127   \n",
       "ENSMUSG00000025903                65              192              282   \n",
       "ENSMUSG00000033813               113              424              459   \n",
       "ENSMUSG00000033793                12                6                3   \n",
       "\n",
       "                    PT82_R2_LPS_8h  PT82_R2_untreated_24h  \\\n",
       "ENSMUSG00000098104              12                     10   \n",
       "ENSMUSG00000033845             155                    156   \n",
       "ENSMUSG00000025903             150                    255   \n",
       "ENSMUSG00000033813             357                    550   \n",
       "ENSMUSG00000033793               5                      4   \n",
       "\n",
       "                    PT82_R2_untreated_8h  \n",
       "ENSMUSG00000098104                     9  \n",
       "ENSMUSG00000033845                   123  \n",
       "ENSMUSG00000025903                   205  \n",
       "ENSMUSG00000033813                   485  \n",
       "ENSMUSG00000033793                     9  \n",
       "\n",
       "[5 rows x 142 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load raw data\n",
    "data_raw = pd.read_csv(os.path.join(dir_data, 'INT_counts.csv'), index_col=0, header=0,)\n",
    "print(data_raw.shape)\n",
    "data_raw.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b992b3d3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(45069, 142)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RNA_PT76_R1_C_albicans_2h</th>\n",
       "      <th>RNA_PT82_R1_C_albicans_24h</th>\n",
       "      <th>RNA_PT76_R1_C_albicans_4h</th>\n",
       "      <th>RNA_PT76_R1_C_albicans_6h</th>\n",
       "      <th>RNA_PT82_R1_C_albicans_8h</th>\n",
       "      <th>RNA_PT76_R1_untreated_0h</th>\n",
       "      <th>RNA_PT82_R1_untreated_24h</th>\n",
       "      <th>RNA_PT76_R1_IFN_beta_2h</th>\n",
       "      <th>RNA_PT82_R1_IFN_beta_24h</th>\n",
       "      <th>RNA_PT76_R1_IFN_beta_4h</th>\n",
       "      <th>...</th>\n",
       "      <th>PT82_R2_IFN_gamma_24h</th>\n",
       "      <th>PT82_R2_IFN_gamma_8h</th>\n",
       "      <th>PT82_R2_LCMV_Cl13_24h</th>\n",
       "      <th>PT82_R2_LCMV_Cl13_8h</th>\n",
       "      <th>PT82_R2_LO28_24h</th>\n",
       "      <th>PT82_R2_LO28_8h</th>\n",
       "      <th>PT82_R2_LPS_24h</th>\n",
       "      <th>PT82_R2_LPS_8h</th>\n",
       "      <th>PT82_R2_untreated_24h</th>\n",
       "      <th>PT82_R2_untreated_8h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000098104</th>\n",
       "      <td>1.405811</td>\n",
       "      <td>2.827185</td>\n",
       "      <td>-0.734410</td>\n",
       "      <td>2.015832</td>\n",
       "      <td>1.521031</td>\n",
       "      <td>3.228688</td>\n",
       "      <td>2.443284</td>\n",
       "      <td>2.954642</td>\n",
       "      <td>2.672776</td>\n",
       "      <td>-0.736557</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.689852</td>\n",
       "      <td>-1.351086</td>\n",
       "      <td>-1.207880</td>\n",
       "      <td>-0.545239</td>\n",
       "      <td>1.332881</td>\n",
       "      <td>-0.552659</td>\n",
       "      <td>-1.183220</td>\n",
       "      <td>-0.699713</td>\n",
       "      <td>-0.696228</td>\n",
       "      <td>-1.089052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000033845</th>\n",
       "      <td>6.087384</td>\n",
       "      <td>6.123522</td>\n",
       "      <td>6.259480</td>\n",
       "      <td>5.883915</td>\n",
       "      <td>6.154641</td>\n",
       "      <td>6.080316</td>\n",
       "      <td>6.096749</td>\n",
       "      <td>6.199833</td>\n",
       "      <td>5.787196</td>\n",
       "      <td>5.995907</td>\n",
       "      <td>...</td>\n",
       "      <td>2.938267</td>\n",
       "      <td>2.547950</td>\n",
       "      <td>3.147234</td>\n",
       "      <td>2.812987</td>\n",
       "      <td>1.495079</td>\n",
       "      <td>3.336820</td>\n",
       "      <td>2.507135</td>\n",
       "      <td>3.585831</td>\n",
       "      <td>2.746262</td>\n",
       "      <td>2.614992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000025903</th>\n",
       "      <td>5.435192</td>\n",
       "      <td>5.343274</td>\n",
       "      <td>5.586660</td>\n",
       "      <td>5.477830</td>\n",
       "      <td>5.306164</td>\n",
       "      <td>5.720310</td>\n",
       "      <td>5.474042</td>\n",
       "      <td>5.958396</td>\n",
       "      <td>5.553141</td>\n",
       "      <td>5.895476</td>\n",
       "      <td>...</td>\n",
       "      <td>4.454032</td>\n",
       "      <td>4.229109</td>\n",
       "      <td>4.484900</td>\n",
       "      <td>4.482084</td>\n",
       "      <td>4.562424</td>\n",
       "      <td>3.814140</td>\n",
       "      <td>4.631544</td>\n",
       "      <td>3.491669</td>\n",
       "      <td>4.179978</td>\n",
       "      <td>4.060240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000033813</th>\n",
       "      <td>4.858575</td>\n",
       "      <td>4.733815</td>\n",
       "      <td>4.624963</td>\n",
       "      <td>5.058409</td>\n",
       "      <td>4.792464</td>\n",
       "      <td>4.719336</td>\n",
       "      <td>4.838214</td>\n",
       "      <td>4.948055</td>\n",
       "      <td>5.124871</td>\n",
       "      <td>5.798483</td>\n",
       "      <td>...</td>\n",
       "      <td>6.055141</td>\n",
       "      <td>6.044589</td>\n",
       "      <td>6.115250</td>\n",
       "      <td>6.174928</td>\n",
       "      <td>6.027826</td>\n",
       "      <td>5.906553</td>\n",
       "      <td>5.818789</td>\n",
       "      <td>5.896174</td>\n",
       "      <td>6.248030</td>\n",
       "      <td>6.253013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000033793</th>\n",
       "      <td>6.449895</td>\n",
       "      <td>6.132281</td>\n",
       "      <td>6.704468</td>\n",
       "      <td>6.296670</td>\n",
       "      <td>6.653410</td>\n",
       "      <td>6.073398</td>\n",
       "      <td>6.311933</td>\n",
       "      <td>6.476513</td>\n",
       "      <td>6.349170</td>\n",
       "      <td>6.323613</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.559897</td>\n",
       "      <td>-1.744813</td>\n",
       "      <td>-0.795451</td>\n",
       "      <td>-1.249856</td>\n",
       "      <td>-1.101236</td>\n",
       "      <td>-1.361392</td>\n",
       "      <td>-1.610179</td>\n",
       "      <td>-1.572410</td>\n",
       "      <td>-1.394374</td>\n",
       "      <td>-1.089052</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 142 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    RNA_PT76_R1_C_albicans_2h  RNA_PT82_R1_C_albicans_24h  \\\n",
       "ENSMUSG00000098104                   1.405811                    2.827185   \n",
       "ENSMUSG00000033845                   6.087384                    6.123522   \n",
       "ENSMUSG00000025903                   5.435192                    5.343274   \n",
       "ENSMUSG00000033813                   4.858575                    4.733815   \n",
       "ENSMUSG00000033793                   6.449895                    6.132281   \n",
       "\n",
       "                    RNA_PT76_R1_C_albicans_4h  RNA_PT76_R1_C_albicans_6h  \\\n",
       "ENSMUSG00000098104                  -0.734410                   2.015832   \n",
       "ENSMUSG00000033845                   6.259480                   5.883915   \n",
       "ENSMUSG00000025903                   5.586660                   5.477830   \n",
       "ENSMUSG00000033813                   4.624963                   5.058409   \n",
       "ENSMUSG00000033793                   6.704468                   6.296670   \n",
       "\n",
       "                    RNA_PT82_R1_C_albicans_8h  RNA_PT76_R1_untreated_0h  \\\n",
       "ENSMUSG00000098104                   1.521031                  3.228688   \n",
       "ENSMUSG00000033845                   6.154641                  6.080316   \n",
       "ENSMUSG00000025903                   5.306164                  5.720310   \n",
       "ENSMUSG00000033813                   4.792464                  4.719336   \n",
       "ENSMUSG00000033793                   6.653410                  6.073398   \n",
       "\n",
       "                    RNA_PT82_R1_untreated_24h  RNA_PT76_R1_IFN_beta_2h  \\\n",
       "ENSMUSG00000098104                   2.443284                 2.954642   \n",
       "ENSMUSG00000033845                   6.096749                 6.199833   \n",
       "ENSMUSG00000025903                   5.474042                 5.958396   \n",
       "ENSMUSG00000033813                   4.838214                 4.948055   \n",
       "ENSMUSG00000033793                   6.311933                 6.476513   \n",
       "\n",
       "                    RNA_PT82_R1_IFN_beta_24h  RNA_PT76_R1_IFN_beta_4h  ...  \\\n",
       "ENSMUSG00000098104                  2.672776                -0.736557  ...   \n",
       "ENSMUSG00000033845                  5.787196                 5.995907  ...   \n",
       "ENSMUSG00000025903                  5.553141                 5.895476  ...   \n",
       "ENSMUSG00000033813                  5.124871                 5.798483  ...   \n",
       "ENSMUSG00000033793                  6.349170                 6.323613  ...   \n",
       "\n",
       "                    PT82_R2_IFN_gamma_24h  PT82_R2_IFN_gamma_8h  \\\n",
       "ENSMUSG00000098104              -0.689852             -1.351086   \n",
       "ENSMUSG00000033845               2.938267              2.547950   \n",
       "ENSMUSG00000025903               4.454032              4.229109   \n",
       "ENSMUSG00000033813               6.055141              6.044589   \n",
       "ENSMUSG00000033793              -2.559897             -1.744813   \n",
       "\n",
       "                    PT82_R2_LCMV_Cl13_24h  PT82_R2_LCMV_Cl13_8h  \\\n",
       "ENSMUSG00000098104              -1.207880             -0.545239   \n",
       "ENSMUSG00000033845               3.147234              2.812987   \n",
       "ENSMUSG00000025903               4.484900              4.482084   \n",
       "ENSMUSG00000033813               6.115250              6.174928   \n",
       "ENSMUSG00000033793              -0.795451             -1.249856   \n",
       "\n",
       "                    PT82_R2_LO28_24h  PT82_R2_LO28_8h  PT82_R2_LPS_24h  \\\n",
       "ENSMUSG00000098104          1.332881        -0.552659        -1.183220   \n",
       "ENSMUSG00000033845          1.495079         3.336820         2.507135   \n",
       "ENSMUSG00000025903          4.562424         3.814140         4.631544   \n",
       "ENSMUSG00000033813          6.027826         5.906553         5.818789   \n",
       "ENSMUSG00000033793         -1.101236        -1.361392        -1.610179   \n",
       "\n",
       "                    PT82_R2_LPS_8h  PT82_R2_untreated_24h  \\\n",
       "ENSMUSG00000098104       -0.699713              -0.696228   \n",
       "ENSMUSG00000033845        3.585831               2.746262   \n",
       "ENSMUSG00000025903        3.491669               4.179978   \n",
       "ENSMUSG00000033813        5.896174               6.248030   \n",
       "ENSMUSG00000033793       -1.572410              -1.394374   \n",
       "\n",
       "                    PT82_R2_untreated_8h  \n",
       "ENSMUSG00000098104             -1.089052  \n",
       "ENSMUSG00000033845              2.614992  \n",
       "ENSMUSG00000025903              4.060240  \n",
       "ENSMUSG00000033813              6.253013  \n",
       "ENSMUSG00000033793             -1.089052  \n",
       "\n",
       "[5 rows x 142 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load normalized data\n",
    "data_norm = pd.read_csv(os.path.join(dir_data, 'INT_norm.csv'), index_col=0, header=0,)\n",
    "print(data_norm.shape)\n",
    "data_norm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ea3ca4b3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(45069, 142)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RNA_PT76_R1_C_albicans_2h</th>\n",
       "      <th>RNA_PT82_R1_C_albicans_24h</th>\n",
       "      <th>RNA_PT76_R1_C_albicans_4h</th>\n",
       "      <th>RNA_PT76_R1_C_albicans_6h</th>\n",
       "      <th>RNA_PT82_R1_C_albicans_8h</th>\n",
       "      <th>RNA_PT76_R1_untreated_0h</th>\n",
       "      <th>RNA_PT82_R1_untreated_24h</th>\n",
       "      <th>RNA_PT76_R1_IFN_beta_2h</th>\n",
       "      <th>RNA_PT82_R1_IFN_beta_24h</th>\n",
       "      <th>RNA_PT76_R1_IFN_beta_4h</th>\n",
       "      <th>...</th>\n",
       "      <th>PT82_R2_IFN_gamma_24h</th>\n",
       "      <th>PT82_R2_IFN_gamma_8h</th>\n",
       "      <th>PT82_R2_LCMV_Cl13_24h</th>\n",
       "      <th>PT82_R2_LCMV_Cl13_8h</th>\n",
       "      <th>PT82_R2_LO28_24h</th>\n",
       "      <th>PT82_R2_LO28_8h</th>\n",
       "      <th>PT82_R2_LPS_24h</th>\n",
       "      <th>PT82_R2_LPS_8h</th>\n",
       "      <th>PT82_R2_untreated_24h</th>\n",
       "      <th>PT82_R2_untreated_8h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000098104</th>\n",
       "      <td>1.405811</td>\n",
       "      <td>2.827185</td>\n",
       "      <td>-0.734410</td>\n",
       "      <td>2.015832</td>\n",
       "      <td>1.521031</td>\n",
       "      <td>3.228688</td>\n",
       "      <td>2.443284</td>\n",
       "      <td>2.954642</td>\n",
       "      <td>2.672776</td>\n",
       "      <td>-0.736557</td>\n",
       "      <td>...</td>\n",
       "      <td>2.264365</td>\n",
       "      <td>1.603131</td>\n",
       "      <td>1.746338</td>\n",
       "      <td>2.408978</td>\n",
       "      <td>4.287098</td>\n",
       "      <td>2.401559</td>\n",
       "      <td>1.770998</td>\n",
       "      <td>2.254504</td>\n",
       "      <td>2.257990</td>\n",
       "      <td>1.865165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000033845</th>\n",
       "      <td>6.087384</td>\n",
       "      <td>6.123522</td>\n",
       "      <td>6.259480</td>\n",
       "      <td>5.883915</td>\n",
       "      <td>6.154641</td>\n",
       "      <td>6.080316</td>\n",
       "      <td>6.096749</td>\n",
       "      <td>6.199833</td>\n",
       "      <td>5.787196</td>\n",
       "      <td>5.995907</td>\n",
       "      <td>...</td>\n",
       "      <td>5.844907</td>\n",
       "      <td>5.454590</td>\n",
       "      <td>6.053874</td>\n",
       "      <td>5.719627</td>\n",
       "      <td>4.401719</td>\n",
       "      <td>6.243460</td>\n",
       "      <td>5.413775</td>\n",
       "      <td>6.492471</td>\n",
       "      <td>5.652902</td>\n",
       "      <td>5.521632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000025903</th>\n",
       "      <td>5.435192</td>\n",
       "      <td>5.343274</td>\n",
       "      <td>5.586660</td>\n",
       "      <td>5.477830</td>\n",
       "      <td>5.306164</td>\n",
       "      <td>5.720310</td>\n",
       "      <td>5.474042</td>\n",
       "      <td>5.958396</td>\n",
       "      <td>5.553141</td>\n",
       "      <td>5.895476</td>\n",
       "      <td>...</td>\n",
       "      <td>6.012205</td>\n",
       "      <td>5.787282</td>\n",
       "      <td>6.043073</td>\n",
       "      <td>6.040258</td>\n",
       "      <td>6.120597</td>\n",
       "      <td>5.372314</td>\n",
       "      <td>6.189717</td>\n",
       "      <td>5.049843</td>\n",
       "      <td>5.738152</td>\n",
       "      <td>5.618413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000033813</th>\n",
       "      <td>4.858575</td>\n",
       "      <td>4.733815</td>\n",
       "      <td>4.624963</td>\n",
       "      <td>5.058409</td>\n",
       "      <td>4.792464</td>\n",
       "      <td>4.719336</td>\n",
       "      <td>4.838214</td>\n",
       "      <td>4.948055</td>\n",
       "      <td>5.124871</td>\n",
       "      <td>5.798483</td>\n",
       "      <td>...</td>\n",
       "      <td>4.993532</td>\n",
       "      <td>4.982980</td>\n",
       "      <td>5.053641</td>\n",
       "      <td>5.113319</td>\n",
       "      <td>4.966217</td>\n",
       "      <td>4.844944</td>\n",
       "      <td>4.757180</td>\n",
       "      <td>4.834565</td>\n",
       "      <td>5.186421</td>\n",
       "      <td>5.191404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000033793</th>\n",
       "      <td>6.449895</td>\n",
       "      <td>6.132281</td>\n",
       "      <td>6.704468</td>\n",
       "      <td>6.296670</td>\n",
       "      <td>6.653410</td>\n",
       "      <td>6.073398</td>\n",
       "      <td>6.311933</td>\n",
       "      <td>6.476513</td>\n",
       "      <td>6.349170</td>\n",
       "      <td>6.323613</td>\n",
       "      <td>...</td>\n",
       "      <td>5.011635</td>\n",
       "      <td>5.826719</td>\n",
       "      <td>6.776081</td>\n",
       "      <td>6.321676</td>\n",
       "      <td>6.470296</td>\n",
       "      <td>6.210140</td>\n",
       "      <td>5.961353</td>\n",
       "      <td>5.999122</td>\n",
       "      <td>6.177158</td>\n",
       "      <td>6.482480</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 142 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    RNA_PT76_R1_C_albicans_2h  RNA_PT82_R1_C_albicans_24h  \\\n",
       "ENSMUSG00000098104                   1.405811                    2.827185   \n",
       "ENSMUSG00000033845                   6.087384                    6.123522   \n",
       "ENSMUSG00000025903                   5.435192                    5.343274   \n",
       "ENSMUSG00000033813                   4.858575                    4.733815   \n",
       "ENSMUSG00000033793                   6.449895                    6.132281   \n",
       "\n",
       "                    RNA_PT76_R1_C_albicans_4h  RNA_PT76_R1_C_albicans_6h  \\\n",
       "ENSMUSG00000098104                  -0.734410                   2.015832   \n",
       "ENSMUSG00000033845                   6.259480                   5.883915   \n",
       "ENSMUSG00000025903                   5.586660                   5.477830   \n",
       "ENSMUSG00000033813                   4.624963                   5.058409   \n",
       "ENSMUSG00000033793                   6.704468                   6.296670   \n",
       "\n",
       "                    RNA_PT82_R1_C_albicans_8h  RNA_PT76_R1_untreated_0h  \\\n",
       "ENSMUSG00000098104                   1.521031                  3.228688   \n",
       "ENSMUSG00000033845                   6.154641                  6.080316   \n",
       "ENSMUSG00000025903                   5.306164                  5.720310   \n",
       "ENSMUSG00000033813                   4.792464                  4.719336   \n",
       "ENSMUSG00000033793                   6.653410                  6.073398   \n",
       "\n",
       "                    RNA_PT82_R1_untreated_24h  RNA_PT76_R1_IFN_beta_2h  \\\n",
       "ENSMUSG00000098104                   2.443284                 2.954642   \n",
       "ENSMUSG00000033845                   6.096749                 6.199833   \n",
       "ENSMUSG00000025903                   5.474042                 5.958396   \n",
       "ENSMUSG00000033813                   4.838214                 4.948055   \n",
       "ENSMUSG00000033793                   6.311933                 6.476513   \n",
       "\n",
       "                    RNA_PT82_R1_IFN_beta_24h  RNA_PT76_R1_IFN_beta_4h  ...  \\\n",
       "ENSMUSG00000098104                  2.672776                -0.736557  ...   \n",
       "ENSMUSG00000033845                  5.787196                 5.995907  ...   \n",
       "ENSMUSG00000025903                  5.553141                 5.895476  ...   \n",
       "ENSMUSG00000033813                  5.124871                 5.798483  ...   \n",
       "ENSMUSG00000033793                  6.349170                 6.323613  ...   \n",
       "\n",
       "                    PT82_R2_IFN_gamma_24h  PT82_R2_IFN_gamma_8h  \\\n",
       "ENSMUSG00000098104               2.264365              1.603131   \n",
       "ENSMUSG00000033845               5.844907              5.454590   \n",
       "ENSMUSG00000025903               6.012205              5.787282   \n",
       "ENSMUSG00000033813               4.993532              4.982980   \n",
       "ENSMUSG00000033793               5.011635              5.826719   \n",
       "\n",
       "                    PT82_R2_LCMV_Cl13_24h  PT82_R2_LCMV_Cl13_8h  \\\n",
       "ENSMUSG00000098104               1.746338              2.408978   \n",
       "ENSMUSG00000033845               6.053874              5.719627   \n",
       "ENSMUSG00000025903               6.043073              6.040258   \n",
       "ENSMUSG00000033813               5.053641              5.113319   \n",
       "ENSMUSG00000033793               6.776081              6.321676   \n",
       "\n",
       "                    PT82_R2_LO28_24h  PT82_R2_LO28_8h  PT82_R2_LPS_24h  \\\n",
       "ENSMUSG00000098104          4.287098         2.401559         1.770998   \n",
       "ENSMUSG00000033845          4.401719         6.243460         5.413775   \n",
       "ENSMUSG00000025903          6.120597         5.372314         6.189717   \n",
       "ENSMUSG00000033813          4.966217         4.844944         4.757180   \n",
       "ENSMUSG00000033793          6.470296         6.210140         5.961353   \n",
       "\n",
       "                    PT82_R2_LPS_8h  PT82_R2_untreated_24h  \\\n",
       "ENSMUSG00000098104        2.254504               2.257990   \n",
       "ENSMUSG00000033845        6.492471               5.652902   \n",
       "ENSMUSG00000025903        5.049843               5.738152   \n",
       "ENSMUSG00000033813        4.834565               5.186421   \n",
       "ENSMUSG00000033793        5.999122               6.177158   \n",
       "\n",
       "                    PT82_R2_untreated_8h  \n",
       "ENSMUSG00000098104              1.865165  \n",
       "ENSMUSG00000033845              5.521632  \n",
       "ENSMUSG00000025903              5.618413  \n",
       "ENSMUSG00000033813              5.191404  \n",
       "ENSMUSG00000033793              6.482480  \n",
       "\n",
       "[5 rows x 142 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load batch corrected data\n",
    "data_batch = pd.read_csv(os.path.join(dir_data, 'INT_norm_batchCorrected.csv'), index_col=0, header=0,)\n",
    "print(data_batch.shape)\n",
    "data_batch.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d7cfa0cd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1397139, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>AveExpr</th>\n",
       "      <th>t</th>\n",
       "      <th>P.Value</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>B</th>\n",
       "      <th>time</th>\n",
       "      <th>treatment</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rn</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000074141</th>\n",
       "      <td>3.248002</td>\n",
       "      <td>3.743529</td>\n",
       "      <td>8.102705</td>\n",
       "      <td>3.703750e-12</td>\n",
       "      <td>8.346216e-08</td>\n",
       "      <td>16.707399</td>\n",
       "      <td>24h</td>\n",
       "      <td>C_albicans</td>\n",
       "      <td>C_albicans_24h</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000108227</th>\n",
       "      <td>3.200747</td>\n",
       "      <td>2.445842</td>\n",
       "      <td>8.204313</td>\n",
       "      <td>2.315435e-12</td>\n",
       "      <td>8.346216e-08</td>\n",
       "      <td>16.540949</td>\n",
       "      <td>24h</td>\n",
       "      <td>C_albicans</td>\n",
       "      <td>C_albicans_24h</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000039699</th>\n",
       "      <td>2.042018</td>\n",
       "      <td>5.973286</td>\n",
       "      <td>7.750445</td>\n",
       "      <td>1.878408e-11</td>\n",
       "      <td>2.821932e-07</td>\n",
       "      <td>15.488355</td>\n",
       "      <td>24h</td>\n",
       "      <td>C_albicans</td>\n",
       "      <td>C_albicans_24h</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000032883</th>\n",
       "      <td>-1.310611</td>\n",
       "      <td>5.362410</td>\n",
       "      <td>-7.620933</td>\n",
       "      <td>3.404211e-11</td>\n",
       "      <td>3.835610e-07</td>\n",
       "      <td>15.182627</td>\n",
       "      <td>24h</td>\n",
       "      <td>C_albicans</td>\n",
       "      <td>C_albicans_24h</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000079184</th>\n",
       "      <td>-0.853646</td>\n",
       "      <td>6.978168</td>\n",
       "      <td>-7.369471</td>\n",
       "      <td>1.075002e-10</td>\n",
       "      <td>8.074875e-07</td>\n",
       "      <td>14.112045</td>\n",
       "      <td>24h</td>\n",
       "      <td>C_albicans</td>\n",
       "      <td>C_albicans_24h</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       logFC   AveExpr         t       P.Value     adj.P.Val  \\\n",
       "rn                                                                             \n",
       "ENSMUSG00000074141  3.248002  3.743529  8.102705  3.703750e-12  8.346216e-08   \n",
       "ENSMUSG00000108227  3.200747  2.445842  8.204313  2.315435e-12  8.346216e-08   \n",
       "ENSMUSG00000039699  2.042018  5.973286  7.750445  1.878408e-11  2.821932e-07   \n",
       "ENSMUSG00000032883 -1.310611  5.362410 -7.620933  3.404211e-11  3.835610e-07   \n",
       "ENSMUSG00000079184 -0.853646  6.978168 -7.369471  1.075002e-10  8.074875e-07   \n",
       "\n",
       "                            B time   treatment           group  \n",
       "rn                                                              \n",
       "ENSMUSG00000074141  16.707399  24h  C_albicans  C_albicans_24h  \n",
       "ENSMUSG00000108227  16.540949  24h  C_albicans  C_albicans_24h  \n",
       "ENSMUSG00000039699  15.488355  24h  C_albicans  C_albicans_24h  \n",
       "ENSMUSG00000032883  15.182627  24h  C_albicans  C_albicans_24h  \n",
       "ENSMUSG00000079184  14.112045  24h  C_albicans  C_albicans_24h  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load DEA results\n",
    "dea_results = pd.read_csv(os.path.join(dir_data, 'DEA','INT_DEA_all.csv'), index_col=0, header=0,)\n",
    "print(dea_results.shape)\n",
    "dea_results.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "94195540",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_all = {'raw':data_raw, \n",
    "            'norm':data_norm, \n",
    "            'batch':data_batch}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bf7a877",
   "metadata": {},
   "source": [
    "## Epigenetic correlation within treatment and timepoints"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "64a13e66",
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter configs\n",
    "LFC_filter = 1\n",
    "AveExpr_filter = 1 #before: 0\n",
    "adjp_filter = 0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "af6d7255",
   "metadata": {},
   "outputs": [],
   "source": [
    "# correlation within treatment and time\n",
    "for step in data_all.keys():\n",
    "\n",
    "    # config plot dimensions\n",
    "    plots_n = (len(annot['treatment'].unique())-1)*(len(annot['time'].unique())-1)+1\n",
    "    col_n = len(annot['time'].unique())-1\n",
    "    row_n = len(annot['treatment'].unique())\n",
    "\n",
    "    # Create a Position index\n",
    "    position_idx = range(1,plots_n + 1)\n",
    "\n",
    "    # initialize plot counter\n",
    "    k=0\n",
    "\n",
    "    fig = plt.figure(1, figsize=(10, 10))\n",
    "\n",
    "    for treatment in sorted(annot['treatment'].unique()):\n",
    "        for time in [sorted(annot['time'].unique())[i] for i in [0,2,3,4,5,1]]:\n",
    "            if (treatment=='untreated' and time!='0h') or (treatment!='untreated' and time=='0h'):\n",
    "                continue\n",
    "\n",
    "    #         print(\"{} {}\".format(treatment, time))\n",
    "\n",
    "            #get rna & atac mean counts\n",
    "            tmp_rna = data_all[step].loc[:,annot.index[(annot['treatment']==treatment) & (annot['time']==time) & (annot['library']=='Quant-seq')]].mean(axis=1)\n",
    "            tmp_atac = data_all[step].loc[:,annot.index[(annot['treatment']==treatment) & (annot['time']==time) & (annot['library']=='ATAC-seq')]].mean(axis=1)\n",
    "            # calculate pearson's r correlation\n",
    "            tmp_corr=round(np.corrcoef(tmp_atac, tmp_rna)[0,1],2)\n",
    "            # extract hit genes for highlighting\n",
    "            tmp_sig_up = dea_results.loc[(dea_results['group']==\"{}_{}\".format(treatment,time)) & (dea_results['adj.P.Val']<adjp_filter) & (dea_results['logFC']>LFC_filter) & (dea_results['AveExpr']>AveExpr_filter),].index\n",
    "            tmp_sig_down = dea_results.loc[(dea_results['group']==\"{}_{}\".format(treatment,time)) & (dea_results['adj.P.Val']<adjp_filter) & (dea_results['logFC']<(-1*LFC_filter)) & (dea_results['AveExpr']>AveExpr_filter),].index\n",
    "\n",
    "            # check if features align\n",
    "            assert(all(tmp_atac.index==tmp_rna.index))\n",
    "\n",
    "            # plot\n",
    "            ax = fig.add_subplot(row_n,col_n,position_idx[k])\n",
    "            ax.scatter(tmp_atac, tmp_rna, alpha=0.1, s=1, facecolors='none', edgecolors='k')\n",
    "            ax.scatter(tmp_atac[tmp_sig_up], tmp_rna[tmp_sig_up], alpha=0.25, s=1, c='r')\n",
    "            ax.scatter(tmp_atac[tmp_sig_down], tmp_rna[tmp_sig_down], alpha=0.25, s=1, c='c')\n",
    "            ax.set_title(\"{} {}\".format(treatment, time))\n",
    "            ax.text(0.05, 0.95, \"Pearson's R={}\".format(tmp_corr), fontsize=8, transform=ax.transAxes, verticalalignment='top')\n",
    "\n",
    "            k=k+1\n",
    "\n",
    "    fig.text(0.5, 0.001, \"Chromatin accessibility (ATAC-seq signal intensity)\", ha='center')\n",
    "    fig.text(0.001, 0.5, \"Gene expression  (Quant-seq signal intensity)\", va='center', rotation='vertical')\n",
    "    fig.tight_layout()\n",
    "\n",
    "    fig.savefig(fname=os.path.join(dir_results, \"epigeneticCorrelation_{}.png\".format(step)),\n",
    "                format=\"png\",\n",
    "                dpi=300,\n",
    "                bbox_inches=\"tight\",\n",
    "                   )\n",
    "\n",
    "    plt.close('all')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b220cade",
   "metadata": {},
   "source": [
    "## Epigenetic correlation across treatment and timepoints highlighting differential genes of UNTREATED"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "119fae9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "step='batch'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9d224088",
   "metadata": {},
   "outputs": [],
   "source": [
    "### RNA-axis always untreated expression - highlighted genes from untreated DEA\n",
    "\n",
    "# config plot dimensions\n",
    "plots_n = (len(annot['treatment'].unique())-1)*(len(annot['time'].unique())-1)+1\n",
    "col_n = len(annot['time'].unique())-1\n",
    "row_n = len(annot['treatment'].unique())\n",
    "\n",
    "# Create a Position index\n",
    "position_idx = range(1,plots_n + 1)\n",
    "\n",
    "# initialize plot counter\n",
    "k=0\n",
    "\n",
    "fig = plt.figure(1, figsize=(10, 10))\n",
    "\n",
    "for treatment in sorted(annot['treatment'].unique()):\n",
    "    for time in [sorted(annot['time'].unique())[i] for i in [0,2,3,4,5,1]]:\n",
    "        if (treatment=='untreated' and time!='0h') or (treatment!='untreated' and time=='0h'):\n",
    "            continue\n",
    "\n",
    "#         print(\"{} {}\".format(treatment, time))\n",
    "\n",
    "        #get rna & atac mean counts\n",
    "        tmp_rna = data_all[step].loc[:,annot.index[(annot['treatment']=='untreated') & (annot['time']=='0h') & (annot['library']=='Quant-seq')]].mean(axis=1)\n",
    "        tmp_atac = data_all[step].loc[:,annot.index[(annot['treatment']==treatment) & (annot['time']==time) & (annot['library']=='ATAC-seq')]].mean(axis=1)\n",
    "        # calculate pearson's r correlation\n",
    "        tmp_corr=round(np.corrcoef(tmp_atac, tmp_rna)[0,1],2)\n",
    "        # extract hit genes for highlighting in red\n",
    "        tmp_sig_up = dea_results.loc[(dea_results['group']==\"{}_{}\".format('untreated','0h')) & (dea_results['adj.P.Val']<adjp_filter) & (dea_results['logFC']>LFC_filter) & (dea_results['AveExpr']>AveExpr_filter),].index\n",
    "        tmp_sig_down = dea_results.loc[(dea_results['group']==\"{}_{}\".format('untreated','0h')) & (dea_results['adj.P.Val']<adjp_filter) & (dea_results['logFC']<(-1*LFC_filter)) & (dea_results['AveExpr']>AveExpr_filter),].index\n",
    "\n",
    "        # check if features align\n",
    "        assert(all(tmp_atac.index==tmp_rna.index))\n",
    "\n",
    "        # plot\n",
    "        ax = fig.add_subplot(row_n,col_n,position_idx[k])\n",
    "        ax.scatter(tmp_atac, tmp_rna, alpha=0.1, s=1, facecolors='none', edgecolors='k')\n",
    "        ax.scatter(tmp_atac[tmp_sig_up], tmp_rna[tmp_sig_up], alpha=0.5, s=1, c='r')\n",
    "        ax.scatter(tmp_atac[tmp_sig_down], tmp_rna[tmp_sig_down], alpha=0.5, s=1, c='c')\n",
    "        ax.set_title(\"{} {}\".format(treatment, time))\n",
    "        ax.text(0.05, 0.95, \"Pearson's R={}\".format(tmp_corr), fontsize=8, transform=ax.transAxes, verticalalignment='top')\n",
    "\n",
    "        k=k+1\n",
    "\n",
    "fig.text(0.5, 0.001, \"Chromatin accessibility (ATAC-seq signal intensity)\", ha='center')\n",
    "fig.text(0.001, 0.5, \"Gene expression in untreated (Quant-seq signal intensity)\", va='center', rotation='vertical')\n",
    "fig.tight_layout()\n",
    "\n",
    "fig.savefig(fname=os.path.join(dir_results, \"epigeneticCorrelation_{}_untreatedRNA.png\".format(step)),\n",
    "            format=\"png\",\n",
    "            dpi=300,\n",
    "            bbox_inches=\"tight\",\n",
    "               )\n",
    "\n",
    "plt.close('all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3f859c51",
   "metadata": {},
   "outputs": [],
   "source": [
    "### ATAC-axis always untreated expression - highlighted genes from untreated DEA\n",
    "\n",
    "# config plot dimensions\n",
    "plots_n = (len(annot['treatment'].unique())-1)*(len(annot['time'].unique())-1)+1\n",
    "col_n = len(annot['time'].unique())-1\n",
    "row_n = len(annot['treatment'].unique())\n",
    "\n",
    "# Create a Position index\n",
    "position_idx = range(1,plots_n + 1)\n",
    "\n",
    "# initialize plot counter\n",
    "k=0\n",
    "\n",
    "fig = plt.figure(1, figsize=(10, 10))\n",
    "\n",
    "for treatment in sorted(annot['treatment'].unique()):\n",
    "    for time in [sorted(annot['time'].unique())[i] for i in [0,2,3,4,5,1]]:\n",
    "        if (treatment=='untreated' and time!='0h') or (treatment!='untreated' and time=='0h'):\n",
    "            continue\n",
    "\n",
    "#         print(\"{} {}\".format(treatment, time))\n",
    "\n",
    "        #get rna & atac mean counts\n",
    "        tmp_rna = data_all[step].loc[:,annot.index[(annot['treatment']==treatment) & (annot['time']==time) & (annot['library']=='Quant-seq')]].mean(axis=1)\n",
    "        tmp_atac = data_all[step].loc[:,annot.index[(annot['treatment']=='untreated') & (annot['time']=='0h') & (annot['library']=='ATAC-seq')]].mean(axis=1)\n",
    "        # calculate pearson's r correlation\n",
    "        tmp_corr=round(np.corrcoef(tmp_atac, tmp_rna)[0,1],2)\n",
    "        # extract hit genes for highlighting in red\n",
    "        tmp_sig_up = dea_results.loc[(dea_results['group']==\"{}_{}\".format('untreated','0h')) & (dea_results['adj.P.Val']<adjp_filter) & (dea_results['logFC']>LFC_filter) & (dea_results['AveExpr']>AveExpr_filter),].index\n",
    "        tmp_sig_down = dea_results.loc[(dea_results['group']==\"{}_{}\".format('untreated','0h')) & (dea_results['adj.P.Val']<adjp_filter) & (dea_results['logFC']<(-1*LFC_filter)) & (dea_results['AveExpr']>AveExpr_filter),].index\n",
    "\n",
    "        # check if features align\n",
    "        assert(all(tmp_atac.index==tmp_rna.index))\n",
    "\n",
    "        # plot\n",
    "        ax = fig.add_subplot(row_n,col_n,position_idx[k])\n",
    "        ax.scatter(tmp_atac, tmp_rna, alpha=0.1, s=1, facecolors='none', edgecolors='k')\n",
    "        ax.scatter(tmp_atac[tmp_sig_up], tmp_rna[tmp_sig_up], alpha=0.5, s=1, c='r')\n",
    "        ax.scatter(tmp_atac[tmp_sig_down], tmp_rna[tmp_sig_down], alpha=0.5, s=1, c='c')\n",
    "        ax.set_title(\"{} {}\".format(treatment, time))\n",
    "        ax.text(0.05, 0.95, \"Pearson's R={}\".format(tmp_corr), fontsize=8, transform=ax.transAxes, verticalalignment='top')\n",
    "\n",
    "        k=k+1\n",
    "\n",
    "fig.text(0.5, 0.001, \"Chromatin accessibility in untreated (ATAC-seq signal intensity)\", ha='center')\n",
    "fig.text(0.001, 0.5, \"Gene expression (Quant-seq signal intensity)\", va='center', rotation='vertical')\n",
    "fig.tight_layout()\n",
    "\n",
    "fig.savefig(fname=os.path.join(dir_results, \"epigeneticCorrelation_{}_untreatedATAC.png\".format(step)),\n",
    "            format=\"png\",\n",
    "            dpi=300,\n",
    "            bbox_inches=\"tight\",\n",
    "               )\n",
    "\n",
    "plt.close('all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "b8ab969d",
   "metadata": {},
   "outputs": [],
   "source": [
    "### both axis from treatment & timepoint - highlighted genes from untreated DEA\n",
    "\n",
    "# config plot dimensions\n",
    "plots_n = (len(annot['treatment'].unique())-1)*(len(annot['time'].unique())-1)+1\n",
    "col_n = len(annot['time'].unique())-1\n",
    "row_n = len(annot['treatment'].unique())\n",
    "\n",
    "# Create a Position index\n",
    "position_idx = range(1,plots_n + 1)\n",
    "\n",
    "# initialize plot counter\n",
    "k=0\n",
    "\n",
    "fig = plt.figure(1, figsize=(10, 10))\n",
    "\n",
    "for treatment in sorted(annot['treatment'].unique()):\n",
    "    for time in [sorted(annot['time'].unique())[i] for i in [0,2,3,4,5,1]]:\n",
    "        if (treatment=='untreated' and time!='0h') or (treatment!='untreated' and time=='0h'):\n",
    "            continue\n",
    "\n",
    "#         print(\"{} {}\".format(treatment, time))\n",
    "\n",
    "        #get rna & atac mean counts\n",
    "        tmp_rna = data_all[step].loc[:,annot.index[(annot['treatment']==treatment) & (annot['time']==time) & (annot['library']=='Quant-seq')]].mean(axis=1)\n",
    "        tmp_atac = data_all[step].loc[:,annot.index[(annot['treatment']==treatment) & (annot['time']==time) & (annot['library']=='ATAC-seq')]].mean(axis=1)\n",
    "        # calculate pearson's r correlation\n",
    "        tmp_corr=round(np.corrcoef(tmp_atac, tmp_rna)[0,1],2)\n",
    "        # extract hit genes for highlighting in red\n",
    "        tmp_sig_up = dea_results.loc[(dea_results['group']==\"{}_{}\".format('untreated','0h')) & (dea_results['adj.P.Val']<adjp_filter) & (dea_results['logFC']>LFC_filter) & (dea_results['AveExpr']>AveExpr_filter),].index\n",
    "        tmp_sig_down = dea_results.loc[(dea_results['group']==\"{}_{}\".format('untreated','0h')) & (dea_results['adj.P.Val']<adjp_filter) & (dea_results['logFC']<(-1*LFC_filter)) & (dea_results['AveExpr']>AveExpr_filter),].index\n",
    "\n",
    "        # check if features align\n",
    "        assert(all(tmp_atac.index==tmp_rna.index))\n",
    "\n",
    "        # plot\n",
    "        ax = fig.add_subplot(row_n,col_n,position_idx[k])\n",
    "        ax.scatter(tmp_atac, tmp_rna, alpha=0.1, s=1, facecolors='none', edgecolors='k')\n",
    "        ax.scatter(tmp_atac[tmp_sig_up], tmp_rna[tmp_sig_up], alpha=0.5, s=1, c='r')\n",
    "        ax.scatter(tmp_atac[tmp_sig_down], tmp_rna[tmp_sig_down], alpha=0.5, s=1, c='c')\n",
    "        ax.set_title(\"{} {}\".format(treatment, time))\n",
    "        ax.text(0.05, 0.95, \"Pearson's R={}\".format(tmp_corr), fontsize=8, transform=ax.transAxes, verticalalignment='top')\n",
    "\n",
    "        k=k+1\n",
    "\n",
    "fig.text(0.5, 0.001, \"Chromatin accessibility (ATAC-seq signal intensity)\", ha='center')\n",
    "fig.text(0.001, 0.5, \"Gene expression (Quant-seq signal intensity)\", va='center', rotation='vertical')\n",
    "fig.tight_layout()\n",
    "\n",
    "fig.savefig(fname=os.path.join(dir_results, \"epigeneticCorrelation_{}_untreatedGenes.png\".format(step)),\n",
    "            format=\"png\",\n",
    "            dpi=300,\n",
    "            bbox_inches=\"tight\",\n",
    "               )\n",
    "\n",
    "plt.close('all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6593cb98",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python_basics",
   "language": "python",
   "name": "basics"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
