{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a4be2566",
   "metadata": {},
   "source": [
    "# Supplementary Tables\n",
    "- goal: Generate all Supplementary Tables in long format (easier for reader)\n",
    "- input: results\n",
    "- output: Supplementary Tables as excel files with multiple sheets\n",
    "\n",
    "requires `xlsxwriter`\n",
    "- `conda install -c conda-forge xlsxwriter` did not work\n",
    "- `pip install XlsxWriter` works"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8eadab1f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/sreichl/projects/bmdm-stim\n"
     ]
    }
   ],
   "source": [
    "cd ../"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bace3c97",
   "metadata": {},
   "outputs": [],
   "source": [
    "# libraries\n",
    "import pandas as pd\n",
    "import os\n",
    "import re\n",
    "from functools import reduce"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "8a9ebeea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# utility functions\n",
    "\n",
    "# Function to apply replacements to a single index value\n",
    "def apply_replacements(index_value):\n",
    "    if type(index_value)==str:\n",
    "        return re.sub('|'.join(re.escape(key) for key in replacements.keys()), \n",
    "                      lambda m: replacements[m.group(0)], index_value)\n",
    "    else:\n",
    "        return index_value\n",
    "    \n",
    "# Function to apply replacements to a single index value\n",
    "def apply_bulk_sample_name_replacements(index_value):\n",
    "    if type(index_value)==str:\n",
    "        return re.sub('|'.join(re.escape(key) for key in bulk_sample_name_replacements.keys()), \n",
    "                      lambda m: bulk_sample_name_replacements[m.group(0)], index_value)\n",
    "    else:\n",
    "        return index_value\n",
    "\n",
    "def df_to_excel(writer, df, sheet_name, column_width, index_format):\n",
    "    \n",
    "    ### replace terms using pre defined dictionary\n",
    "    # apply replacements to dataframe content\n",
    "    string_columns = df.select_dtypes(include=['object']).columns\n",
    "    df[string_columns] = df[string_columns].replace(replacements, regex=True)\n",
    "    \n",
    "    ## column names\n",
    "    # apply replacements to column names\n",
    "    df.columns = [re.sub('|'.join(re.escape(key) for key in replacements.keys()), \n",
    "                         lambda m: replacements[m.group(0)], col) for col in df.columns]\n",
    "    # Replace underscores with spaces and capitalize the first letter of each column name\n",
    "    df.columns = [col.replace('_', ' ')[0].upper() + col.replace('_', ' ')[1:] for col in df.columns]\n",
    "    \n",
    "    ## indices\n",
    "    # apply replacements to each (Multi)Index content\n",
    "    if isinstance(df.index, pd.MultiIndex):\n",
    "        df.index = df.index.map(lambda x: tuple(apply_replacements(level) for level in x))\n",
    "    else:\n",
    "        df.index = df.index.map(apply_replacements)\n",
    "    # apply replacements to index names\n",
    "    df.index.names = [re.sub('|'.join(re.escape(key) for key in replacements.keys()), \n",
    "                         lambda m: replacements[m.group(0)], col) for col in df.index.names]\n",
    "    # Capitalize only the first letter of each index level name, preserving the rest\n",
    "    df.index.names = [name[0].upper() + name[1:] if name is not None else None for name in df.index.names]\n",
    "    \n",
    "    ### create Excel sheet\n",
    "    idx_nums = len(df.index.names)\n",
    "    \n",
    "    # Write table as a separate sheet to the Excel file\n",
    "    df.to_excel(writer, sheet_name=sheet_name, index=False, header=False, startrow=1, startcol=idx_nums, freeze_panes = (1,idx_nums))\n",
    "    \n",
    "    # set all columns to same width\n",
    "    writer.sheets[sheet_name].set_column(first_col=0, last_col=df.shape[1]+idx_nums-1, width = column_width)\n",
    "    \n",
    "    # Write the column headers with the defined format.\n",
    "    for col_num, value in enumerate(df.columns.values):\n",
    "        writer.sheets[sheet_name].write(0, col_num + idx_nums, value, index_format)\n",
    "\n",
    "    # Write the index names with the defined format\n",
    "    for idx_name_num, value in enumerate(df.index.names):\n",
    "        # write the index name/header\n",
    "        writer.sheets[sheet_name].write(0, idx_name_num, value, index_format)\n",
    "        \n",
    "        # Write the index values with the defined format for each index.\n",
    "        for idx_num, idx_value in enumerate(df.index.values):\n",
    "            if idx_nums == 1:\n",
    "                writer.sheets[sheet_name].write(idx_num + 1, idx_name_num, idx_value, index_format)\n",
    "            else:\n",
    "                writer.sheets[sheet_name].write(idx_num + 1, idx_name_num, idx_value[idx_name_num], index_format)\n",
    "\n",
    "    return writer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "22795d4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rearrange_sample_names(sample_name):\n",
    "#     RNA_{stimulus}_{time}_{replicate}_{batch}\n",
    "#     ATAC_{stimulus}_{time}_{replicate}_{batch}\n",
    "    parts = sample_name.split('_')\n",
    "    rearranged_parts = [parts[0], parts[3], parts[4], parts[2], parts[1]]\n",
    "    return '_'.join(rearranged_parts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "8072705a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# configs\n",
    "results_path = os.path.join('results', 'tables')\n",
    "column_width = 25\n",
    "\n",
    "# define replacement dictionary\n",
    "replacements = {\n",
    "    'PT76': '1',\n",
    "    'PT82': '2',\n",
    "    'C_albicans': 'Candida',\n",
    "    'LCMV_Cl13': 'LCMV',\n",
    "    'experiment_id': 'batch',\n",
    "    'Conditions': 'Time points',\n",
    "    'Condition': 'Time point',\n",
    "    'Treatment_time': 'Time point',\n",
    "    'Treatment': 'Stimulus',\n",
    "    'group': 'Comparison',\n",
    "#      'logFC': 'Coefficient', or 'Log fold change'?\n",
    "    'AveExpr': 'Average expression',\n",
    "    'P.Value': 'P-value',\n",
    "    'adj.P.Val': 'Adjusted P-value',\n",
    "    'untreated': 'Untreated',\n",
    "    'LO28_2h': 'Listeria 2h',\n",
    "    'LO28_6h': 'Listeria 6h',\n",
    "    'LO28-6h': 'Listeria 6h',\n",
    "    'LO28-24h': 'Listeria 24h',\n",
    "    'LO28': 'Listeria',\n",
    "    'p_val_adj': 'Adjusted P-value',\n",
    "    'p_val': 'P-value',\n",
    "    #      'avg_log2FC': 'Coefficient', or 'Average log2 fold change'?\n",
    "    'pct.': 'Percentage expression ',\n",
    "#     \"_2h\": \" 2h\",\n",
    "#     \"_4h\": \" 4h\",\n",
    "#     \"_6h\": \" 6h\",\n",
    "#     \"_8h\": \" 8h\",\n",
    "#     \"_24h\": \" 24h\",\n",
    "    \"mSpi1\": \"Spi1\",\n",
    "    \"mNon-targeting\": \"WT\",\n",
    "}\n",
    "\n",
    "\n",
    "bulk_sample_name_replacements = {\n",
    "    'PT76': '1',\n",
    "    'PT82': '2',\n",
    "    'C_albicans': 'Candida',\n",
    "    'LCMV_Cl13': 'LCMV',\n",
    "    'IFN_beta': 'IFNbeta',\n",
    "    'IFN_gamma': 'IFNgamma',\n",
    "    'LO28': 'Listeria',\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "faa774da",
   "metadata": {},
   "outputs": [],
   "source": [
    "# config the xlsxwriter format for sheets\n",
    "xlsx_format_dict = {'bold': True,\n",
    "   'font_color': 'black',\n",
    "#    'locked': True,\n",
    "#    'text_wrap': True,\n",
    "#    'bg_color': '#FDFD96',\n",
    "   'border': False,\n",
    "  'align': 'left',\n",
    "# 'valign': 'vcenter'\n",
    "  }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e35cba17",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(results_path):\n",
    "        os.mkdir(results_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "71e8ac27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RNA</th>\n",
       "      <th>ATAC</th>\n",
       "      <th>INT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>LO28</th>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LPS</th>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LCMV_Cl13</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C_albicans</th>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IFN_beta</th>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IFN_gamma</th>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            RNA  ATAC  INT\n",
       "LO28         10    10    7\n",
       "LPS           8     6    6\n",
       "LCMV_Cl13     2     3    4\n",
       "C_albicans    2     5    6\n",
       "IFN_beta      4     8    4\n",
       "IFN_gamma     3     6    5"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load bulk time series configs\n",
    "timeseries_k = pd.read_csv(os.path.join('config','BMDM_timeseries_k.csv'), index_col=0)\n",
    "# timeseries_enrichment_terms = pd.read_csv(os.path.join('config','BMDM_timeseries_enrichment_terms.csv'), index_col=0) # NOT USED in main figures\n",
    "timeseries_k"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "cc085639",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>database</th>\n",
       "      <th>terms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Spi1_plot</th>\n",
       "      <td>BioPlanet_2019</td>\n",
       "      <td>Oncostatin M | Interferon alpha/beta signaling...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 database                                              terms\n",
       "Spi1_plot  BioPlanet_2019  Oncostatin M | Interferon alpha/beta signaling..."
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load small CROP-seq configs\n",
    "spi1_enrichment_terms = pd.read_csv(os.path.join('config','Spi1KO_enrichment_terms.csv'), index_col=0)\n",
    "spi1_enrichment_terms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b3e9e4a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(55487, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>version</th>\n",
       "      <th>source</th>\n",
       "      <th>symbol</th>\n",
       "      <th>external_gene_source</th>\n",
       "      <th>description</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>length</th>\n",
       "      <th>gc</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000194</th>\n",
       "      <td>13</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>Gpr107</td>\n",
       "      <td>MGI Symbol</td>\n",
       "      <td>G protein-coupled receptor 107 [Source:MGI Sym...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>12766</td>\n",
       "      <td>0.494595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000247</th>\n",
       "      <td>11</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>Lhx2</td>\n",
       "      <td>MGI Symbol</td>\n",
       "      <td>LIM homeobox protein 2 [Source:MGI Symbol;Acc:...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>4393</td>\n",
       "      <td>0.597086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000544</th>\n",
       "      <td>14</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>Gpa33</td>\n",
       "      <td>MGI Symbol</td>\n",
       "      <td>glycoprotein A33 (transmembrane) [Source:MGI S...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2742</td>\n",
       "      <td>0.530999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000817</th>\n",
       "      <td>10</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>Fasl</td>\n",
       "      <td>MGI Symbol</td>\n",
       "      <td>Fas ligand (TNF superfamily, member 6) [Source...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1937</td>\n",
       "      <td>0.430563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000889</th>\n",
       "      <td>8</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>Dbh</td>\n",
       "      <td>MGI Symbol</td>\n",
       "      <td>dopamine beta hydroxylase [Source:MGI Symbol;A...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2692</td>\n",
       "      <td>0.554978</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    version          source  symbol external_gene_source  \\\n",
       "ensembl_gene_id                                                            \n",
       "ENSMUSG00000000194       13  ensembl_havana  Gpr107           MGI Symbol   \n",
       "ENSMUSG00000000247       11  ensembl_havana    Lhx2           MGI Symbol   \n",
       "ENSMUSG00000000544       14  ensembl_havana   Gpa33           MGI Symbol   \n",
       "ENSMUSG00000000817       10  ensembl_havana    Fasl           MGI Symbol   \n",
       "ENSMUSG00000000889        8  ensembl_havana     Dbh           MGI Symbol   \n",
       "\n",
       "                                                          description  \\\n",
       "ensembl_gene_id                                                         \n",
       "ENSMUSG00000000194  G protein-coupled receptor 107 [Source:MGI Sym...   \n",
       "ENSMUSG00000000247  LIM homeobox protein 2 [Source:MGI Symbol;Acc:...   \n",
       "ENSMUSG00000000544  glycoprotein A33 (transmembrane) [Source:MGI S...   \n",
       "ENSMUSG00000000817  Fas ligand (TNF superfamily, member 6) [Source...   \n",
       "ENSMUSG00000000889  dopamine beta hydroxylase [Source:MGI Symbol;A...   \n",
       "\n",
       "                      gene_biotype  length        gc  \n",
       "ensembl_gene_id                                       \n",
       "ENSMUSG00000000194  protein_coding   12766  0.494595  \n",
       "ENSMUSG00000000247  protein_coding    4393  0.597086  \n",
       "ENSMUSG00000000544  protein_coding    2742  0.530999  \n",
       "ENSMUSG00000000817  protein_coding    1937  0.430563  \n",
       "ENSMUSG00000000889  protein_coding    2692  0.554978  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load gene annotations\n",
    "gene_annot = pd.read_csv(os.path.join('results', 'RNA', 'counts', 'gene_annotation.tsv'), index_col=0, sep = '\\t')\n",
    "# Rename column from 'external_gene_name' to 'symbol'\n",
    "gene_annot.rename(columns={'external_gene_name': 'symbol'}, inplace=True)\n",
    "print(gene_annot.shape)\n",
    "gene_annot.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c733d7f7",
   "metadata": {},
   "source": [
    "# RNA - TableS1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "753cfbf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = os.path.join('results', 'RNA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "7cef87e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create Excel writer\n",
    "writer = pd.ExcelWriter(os.path.join(results_path,'S01_RNAseq.xlsx'), engine='xlsxwriter')\n",
    "\n",
    "# Get the workbook and worksheet objects\n",
    "workbook = writer.book\n",
    "# define specific formatting\n",
    "index_format = workbook.add_format(xlsx_format_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "043a77d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add an empty sheet called LEGEND (fill with manually curated data later)\n",
    "pd.DataFrame().to_excel(writer, sheet_name=\"TableS1_legends\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "8b365716",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(64, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>library</th>\n",
       "      <th>timepoint after thawing</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>Treatment</th>\n",
       "      <th>Treatment_time</th>\n",
       "      <th>sex</th>\n",
       "      <th>organism</th>\n",
       "      <th>bio-replicate</th>\n",
       "      <th>experiment_id</th>\n",
       "      <th>Aligned_percentage</th>\n",
       "      <th>Aligned_million</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RNA_PT76_R1_C_albicans_2h</th>\n",
       "      <td>Quant-seq</td>\n",
       "      <td>d010</td>\n",
       "      <td>BMDM</td>\n",
       "      <td>C_albicans</td>\n",
       "      <td>2h</td>\n",
       "      <td>female</td>\n",
       "      <td>mouse</td>\n",
       "      <td>R1</td>\n",
       "      <td>PT76</td>\n",
       "      <td>0.740</td>\n",
       "      <td>4.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA_PT82_R1_C_albicans_24h</th>\n",
       "      <td>Quant-seq</td>\n",
       "      <td>d010</td>\n",
       "      <td>BMDM</td>\n",
       "      <td>C_albicans</td>\n",
       "      <td>24h</td>\n",
       "      <td>female</td>\n",
       "      <td>mouse</td>\n",
       "      <td>R1</td>\n",
       "      <td>PT82</td>\n",
       "      <td>0.739</td>\n",
       "      <td>4.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA_PT76_R1_C_albicans_4h</th>\n",
       "      <td>Quant-seq</td>\n",
       "      <td>d010</td>\n",
       "      <td>BMDM</td>\n",
       "      <td>C_albicans</td>\n",
       "      <td>4h</td>\n",
       "      <td>female</td>\n",
       "      <td>mouse</td>\n",
       "      <td>R1</td>\n",
       "      <td>PT76</td>\n",
       "      <td>0.745</td>\n",
       "      <td>3.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA_PT76_R1_C_albicans_6h</th>\n",
       "      <td>Quant-seq</td>\n",
       "      <td>d010</td>\n",
       "      <td>BMDM</td>\n",
       "      <td>C_albicans</td>\n",
       "      <td>6h</td>\n",
       "      <td>female</td>\n",
       "      <td>mouse</td>\n",
       "      <td>R1</td>\n",
       "      <td>PT76</td>\n",
       "      <td>0.741</td>\n",
       "      <td>4.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA_PT82_R1_C_albicans_8h</th>\n",
       "      <td>Quant-seq</td>\n",
       "      <td>d010</td>\n",
       "      <td>BMDM</td>\n",
       "      <td>C_albicans</td>\n",
       "      <td>8h</td>\n",
       "      <td>female</td>\n",
       "      <td>mouse</td>\n",
       "      <td>R1</td>\n",
       "      <td>PT82</td>\n",
       "      <td>0.722</td>\n",
       "      <td>3.9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              library timepoint after thawing cell_line  \\\n",
       "sample                                                                    \n",
       "RNA_PT76_R1_C_albicans_2h   Quant-seq                    d010      BMDM   \n",
       "RNA_PT82_R1_C_albicans_24h  Quant-seq                    d010      BMDM   \n",
       "RNA_PT76_R1_C_albicans_4h   Quant-seq                    d010      BMDM   \n",
       "RNA_PT76_R1_C_albicans_6h   Quant-seq                    d010      BMDM   \n",
       "RNA_PT82_R1_C_albicans_8h   Quant-seq                    d010      BMDM   \n",
       "\n",
       "                             Treatment Treatment_time     sex organism  \\\n",
       "sample                                                                   \n",
       "RNA_PT76_R1_C_albicans_2h   C_albicans             2h  female    mouse   \n",
       "RNA_PT82_R1_C_albicans_24h  C_albicans            24h  female    mouse   \n",
       "RNA_PT76_R1_C_albicans_4h   C_albicans             4h  female    mouse   \n",
       "RNA_PT76_R1_C_albicans_6h   C_albicans             6h  female    mouse   \n",
       "RNA_PT82_R1_C_albicans_8h   C_albicans             8h  female    mouse   \n",
       "\n",
       "                           bio-replicate experiment_id  Aligned_percentage  \\\n",
       "sample                                                                       \n",
       "RNA_PT76_R1_C_albicans_2h             R1          PT76               0.740   \n",
       "RNA_PT82_R1_C_albicans_24h            R1          PT82               0.739   \n",
       "RNA_PT76_R1_C_albicans_4h             R1          PT76               0.745   \n",
       "RNA_PT76_R1_C_albicans_6h             R1          PT76               0.741   \n",
       "RNA_PT82_R1_C_albicans_8h             R1          PT82               0.722   \n",
       "\n",
       "                            Aligned_million  \n",
       "sample                                       \n",
       "RNA_PT76_R1_C_albicans_2h               4.6  \n",
       "RNA_PT82_R1_C_albicans_24h              4.3  \n",
       "RNA_PT76_R1_C_albicans_4h               3.5  \n",
       "RNA_PT76_R1_C_albicans_6h               4.3  \n",
       "RNA_PT82_R1_C_albicans_8h               3.9  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# metadata\n",
    "metadata = pd.read_csv(os.path.join('metadata', 'RNA_sample_metadata.csv'), index_col=0)\n",
    "# load statistics\n",
    "statistics = pd.read_csv(os.path.join('metadata', 'RNA_multiqc_report.csv'), index_col=0)\n",
    "# concatenate dataframes\n",
    "metadata = pd.concat([metadata, statistics], axis=1)\n",
    "# remove 2 samples by QC\n",
    "metadata.drop(['RNA_PT82_R1_untreated_8h','RNA_PT82_R1_LPS_24h'], axis=0, inplace=True)\n",
    "# remove unused columns\n",
    "metadata.drop(['mouse_pool'], axis=1, inplace=True)\n",
    "# remove experiment id from sample name\n",
    "# metadata.index = metadata.index.astype(str).str.replace('PT\\d{2}_', '', regex=True)\n",
    "# add index name\n",
    "metadata.index.names = [\"sample\"]\n",
    "\n",
    "metadata_copy = metadata.copy(deep=True)\n",
    "metadata_copy.index = metadata_copy.index.map(apply_bulk_sample_name_replacements)\n",
    "metadata_copy.index = [rearrange_sample_names(col) for col in metadata_copy.index]\n",
    "metadata_copy.index.names = [\"sample\"]\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = metadata_copy,\n",
    "                     sheet_name = \"sample_annotation\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "print(metadata.shape)\n",
    "metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "323e2069",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(263970, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>LogFC</th>\n",
       "      <th>Average expression</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Comparison</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">ENSMUSG00000038437</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">Mllt6</th>\n",
       "      <th>Candida_4h</th>\n",
       "      <td>1.583815</td>\n",
       "      <td>5.750605</td>\n",
       "      <td>1.885853e-10</td>\n",
       "      <td>2.400503e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Candida_2h</th>\n",
       "      <td>1.709446</td>\n",
       "      <td>5.750605</td>\n",
       "      <td>2.153350e-11</td>\n",
       "      <td>3.915714e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Candida_24h</th>\n",
       "      <td>0.173672</td>\n",
       "      <td>5.750605</td>\n",
       "      <td>3.405508e-01</td>\n",
       "      <td>9.999846e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Candida_8h</th>\n",
       "      <td>0.738289</td>\n",
       "      <td>5.750605</td>\n",
       "      <td>1.223462e-04</td>\n",
       "      <td>3.893363e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Candida_6h</th>\n",
       "      <td>1.236302</td>\n",
       "      <td>5.750605</td>\n",
       "      <td>6.920095e-08</td>\n",
       "      <td>8.808589e-05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          LogFC  Average expression  \\\n",
       "Gene               Symbol Comparison                                  \n",
       "ENSMUSG00000038437 Mllt6  Candida_4h   1.583815            5.750605   \n",
       "                          Candida_2h   1.709446            5.750605   \n",
       "                          Candida_24h  0.173672            5.750605   \n",
       "                          Candida_8h   0.738289            5.750605   \n",
       "                          Candida_6h   1.236302            5.750605   \n",
       "\n",
       "                                            P-value  Adjusted P-value  \n",
       "Gene               Symbol Comparison                                   \n",
       "ENSMUSG00000038437 Mllt6  Candida_4h   1.885853e-10      2.400503e-06  \n",
       "                          Candida_2h   2.153350e-11      3.915714e-08  \n",
       "                          Candida_24h  3.405508e-01      9.999846e-01  \n",
       "                          Candida_8h   1.223462e-04      3.893363e-02  \n",
       "                          Candida_6h   6.920095e-08      8.808589e-05  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# DEA results\n",
    "dea_dfs = []\n",
    "\n",
    "for treatment in metadata[\"Treatment\"].unique():\n",
    "    # load dea results\n",
    "    df_tmp = pd.read_csv(os.path.join(data_path, 'DEA', 'DEA_{}.tsv'.format(treatment)), delimiter=\"\\t\")\n",
    "\n",
    "    # Drop columns 'B' and 't'\n",
    "    df_tmp = df_tmp.drop(['B', 't'], axis=1)\n",
    "\n",
    "    # Rename column from 'rn' to 'gene'\n",
    "    df_tmp.rename(columns={'rn': 'gene'}, inplace=True)\n",
    "    \n",
    "    # Add gene symbols using gene annotation\n",
    "    df_tmp = pd.merge(df_tmp, gene_annot[['symbol']], left_on='gene', right_index=True, how='left')\n",
    "    \n",
    "    # set indices\n",
    "    df_tmp.set_index([\"gene\", \"symbol\", \"group\"], inplace=True)\n",
    "\n",
    "    # Pivot the data\n",
    "#     df_tmp = df_tmp.pivot(index='gene', columns='group')\n",
    "\n",
    "    # Flatten the multi-index columns and reverse the order of concatenation\n",
    "#     df_tmp.columns = ['_'.join(col[::-1]).strip() for col in df_pivot.columns.values]\n",
    "    \n",
    "    dea_dfs.append(df_tmp)\n",
    "\n",
    "# dea_df = pd.concat(dea_dfs, axis=1)\n",
    "dea_df = pd.concat(dea_dfs, axis=0)\n",
    "\n",
    "# filter for genes that are at least in one comparison fulfill the most permissive criteria (C.albicans)\n",
    "genes_sig = dea_df.index[(dea_df[\"adj.P.Val\"]<0.05) & (abs(dea_df[\"logFC\"])>1) & (dea_df[\"AveExpr\"]>1)].get_level_values('gene').unique().tolist()\n",
    "dea_df = dea_df.loc[genes_sig,:]\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = dea_df,\n",
    "                     sheet_name = \"differential_expression\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "print(dea_df.shape)\n",
    "dea_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "3a7ac893",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(15418, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Database</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Term</th>\n",
       "      <th>Cluster</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>'de novo' posttranslational protein folding (GO:0051084)</th>\n",
       "      <th>1</th>\n",
       "      <td>1/28</td>\n",
       "      <td>0.987242</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>DNAJB5</td>\n",
       "      <td>0.247448</td>\n",
       "      <td>GO Biological Process 2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2-oxoglutarate metabolic process (GO:0006103)</th>\n",
       "      <th>1</th>\n",
       "      <td>3/11</td>\n",
       "      <td>0.203409</td>\n",
       "      <td>0.867825</td>\n",
       "      <td>IDH1;IDH2;D2HGDH</td>\n",
       "      <td>1.894196</td>\n",
       "      <td>GO Biological Process 2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3'-UTR-mediated mRNA destabilization (GO:0061158)</th>\n",
       "      <th>1</th>\n",
       "      <td>1/13</td>\n",
       "      <td>0.867838</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>ZFP36L1</td>\n",
       "      <td>0.533594</td>\n",
       "      <td>GO Biological Process 2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3'-UTR-mediated mRNA stabilization (GO:0070935)</th>\n",
       "      <th>1</th>\n",
       "      <td>1/11</td>\n",
       "      <td>0.819539</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>MAPK14</td>\n",
       "      <td>0.630710</td>\n",
       "      <td>GO Biological Process 2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3'-phosphoadenosine 5'-phosphosulfate metabolic process (GO:0050427)</th>\n",
       "      <th>1</th>\n",
       "      <td>2/11</td>\n",
       "      <td>0.485046</td>\n",
       "      <td>0.978915</td>\n",
       "      <td>ENPP1;ABHD14B</td>\n",
       "      <td>1.262108</td>\n",
       "      <td>GO Biological Process 2021</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                           Overlap   P-value  \\\n",
       "Term                                               Cluster                     \n",
       "'de novo' posttranslational protein folding (GO... 1          1/28  0.987242   \n",
       "2-oxoglutarate metabolic process (GO:0006103)      1          3/11  0.203409   \n",
       "3'-UTR-mediated mRNA destabilization (GO:0061158)  1          1/13  0.867838   \n",
       "3'-UTR-mediated mRNA stabilization (GO:0070935)    1          1/11  0.819539   \n",
       "3'-phosphoadenosine 5'-phosphosulfate metabolic... 1          2/11  0.485046   \n",
       "\n",
       "                                                            Adjusted P-value  \\\n",
       "Term                                               Cluster                     \n",
       "'de novo' posttranslational protein folding (GO... 1                1.000000   \n",
       "2-oxoglutarate metabolic process (GO:0006103)      1                0.867825   \n",
       "3'-UTR-mediated mRNA destabilization (GO:0061158)  1                1.000000   \n",
       "3'-UTR-mediated mRNA stabilization (GO:0070935)    1                1.000000   \n",
       "3'-phosphoadenosine 5'-phosphosulfate metabolic... 1                0.978915   \n",
       "\n",
       "                                                                       Genes  \\\n",
       "Term                                               Cluster                     \n",
       "'de novo' posttranslational protein folding (GO... 1                  DNAJB5   \n",
       "2-oxoglutarate metabolic process (GO:0006103)      1        IDH1;IDH2;D2HGDH   \n",
       "3'-UTR-mediated mRNA destabilization (GO:0061158)  1                 ZFP36L1   \n",
       "3'-UTR-mediated mRNA stabilization (GO:0070935)    1                  MAPK14   \n",
       "3'-phosphoadenosine 5'-phosphosulfate metabolic... 1           ENPP1;ABHD14B   \n",
       "\n",
       "                                                            Odds Ratio  \\\n",
       "Term                                               Cluster               \n",
       "'de novo' posttranslational protein folding (GO... 1          0.247448   \n",
       "2-oxoglutarate metabolic process (GO:0006103)      1          1.894196   \n",
       "3'-UTR-mediated mRNA destabilization (GO:0061158)  1          0.533594   \n",
       "3'-UTR-mediated mRNA stabilization (GO:0070935)    1          0.630710   \n",
       "3'-phosphoadenosine 5'-phosphosulfate metabolic... 1          1.262108   \n",
       "\n",
       "                                                                              Database  \n",
       "Term                                               Cluster                              \n",
       "'de novo' posttranslational protein folding (GO... 1        GO Biological Process 2021  \n",
       "2-oxoglutarate metabolic process (GO:0006103)      1        GO Biological Process 2021  \n",
       "3'-UTR-mediated mRNA destabilization (GO:0061158)  1        GO Biological Process 2021  \n",
       "3'-UTR-mediated mRNA stabilization (GO:0070935)    1        GO Biological Process 2021  \n",
       "3'-phosphoadenosine 5'-phosphosulfate metabolic... 1        GO Biological Process 2021  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Enrichment results of DEG clusters (term X cluster_statistic)\n",
    "enr_dfs = []\n",
    "\n",
    "for cluster in range(1,9):\n",
    "    # load enr results\n",
    "    df_tmp = pd.read_csv(os.path.join(data_path, 'enrichment_analysis', 'DEG_clusters', 'k_8', 'cluster_{}'.format(cluster), 'GO_Biological_Process_2021', 'Enrichr_GO_Biological_Process_2021_cluster_{}.csv'.format(cluster)))\n",
    "\n",
    "    # Drop columns\n",
    "    df_tmp = df_tmp.drop(['Unnamed: 0', 'Gene_set'], axis=1)\n",
    "#     df_tmp = df_tmp.drop(['Unnamed: 0', 'Gene_set', 'Genes'], axis=1)\n",
    "\n",
    "    # Rename columns \n",
    "#     df_tmp.columns = [\"cluster_{}_{}\".format(cluster,col) for col in df_tmp.columns]\n",
    "\n",
    "    # add cluster column\n",
    "    df_tmp[\"Cluster\"] = cluster\n",
    "    \n",
    "    # set indices\n",
    "    df_tmp.set_index([\"Term\", \"Cluster\"], inplace=True)\n",
    "    \n",
    "    enr_dfs.append(df_tmp)\n",
    "\n",
    "enr_df = pd.concat(enr_dfs, axis=0)\n",
    "# enr_df = pd.concat(enr_dfs, axis=1)\n",
    "\n",
    "# add database column\n",
    "enr_df[\"Database\"] = 'GO Biological Process 2021'\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = enr_df,\n",
    "                     sheet_name = \"diff_expression_enrichments\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "print(enr_df.shape)\n",
    "enr_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "0dc014a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4775, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Candida</th>\n",
       "      <th>IFN beta</th>\n",
       "      <th>IFN gamma</th>\n",
       "      <th>LCMV</th>\n",
       "      <th>Listeria</th>\n",
       "      <th>LPS</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th>Symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028</th>\n",
       "      <th>Cdc45</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000056</th>\n",
       "      <th>Narf</th>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000058</th>\n",
       "      <th>Cav2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000085</th>\n",
       "      <th>Scmh1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000148</th>\n",
       "      <th>Brat1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Candida  IFN beta  IFN gamma  LCMV  Listeria  LPS\n",
       "Gene               Symbol                                                   \n",
       "ENSMUSG00000000028 Cdc45       NaN       1.0        NaN   NaN       8.0  4.0\n",
       "ENSMUSG00000000056 Narf        NaN       4.0        NaN   NaN       1.0  7.0\n",
       "ENSMUSG00000000058 Cav2        NaN       4.0        NaN   NaN       NaN  7.0\n",
       "ENSMUSG00000000085 Scmh1       NaN       1.0        NaN   NaN       7.0  1.0\n",
       "ENSMUSG00000000148 Brat1       NaN       4.0        NaN   NaN       1.0  NaN"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Time series clustering (gene X stimulus matrix with cluster membership or empty)\n",
    "ts_dfs = []\n",
    "\n",
    "for treatment in metadata[\"Treatment\"].unique():\n",
    "    if treatment==\"untreated\":\n",
    "        continue\n",
    "    \n",
    "    # load dea results\n",
    "    df_tmp = pd.read_csv(os.path.join(data_path, 'time_series', treatment, 'k_{}'.format(timeseries_k.loc[treatment,\"RNA\"]), 'clustering_{}.csv'.format(treatment)), index_col=[1,2])\n",
    "    \n",
    "    # Drop columns\n",
    "    df_tmp = df_tmp.drop(['Unnamed: 0'], axis=1)\n",
    "\n",
    "    # Rename columns \n",
    "    df_tmp.columns = [\"{}\".format(treatment)]\n",
    "    \n",
    "    ts_dfs.append(df_tmp)\n",
    "    \n",
    "ts_df = pd.concat(ts_dfs, axis=1)\n",
    "ts_df.index.names = [\"gene\", \"symbol\"]\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = ts_df,\n",
    "                     sheet_name = \"time_series_clusters\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "print(ts_df.shape)\n",
    "ts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "72998395",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(17401, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Database</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Term</th>\n",
       "      <th>Stimulus</th>\n",
       "      <th>Cluster</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4-1BB-dependent immune response</th>\n",
       "      <th>Candida</th>\n",
       "      <th>1</th>\n",
       "      <td>2/17</td>\n",
       "      <td>0.017993</td>\n",
       "      <td>0.057914</td>\n",
       "      <td>NFKB1;NFKBIA</td>\n",
       "      <td>9.710466</td>\n",
       "      <td>BioPlanet 2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABC transporters</th>\n",
       "      <th>Candida</th>\n",
       "      <th>1</th>\n",
       "      <td>1/26</td>\n",
       "      <td>0.274543</td>\n",
       "      <td>0.360688</td>\n",
       "      <td>ABCC5</td>\n",
       "      <td>3.151861</td>\n",
       "      <td>BioPlanet 2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKT signaling pathway</th>\n",
       "      <th>Candida</th>\n",
       "      <th>1</th>\n",
       "      <td>2/19</td>\n",
       "      <td>0.022265</td>\n",
       "      <td>0.065900</td>\n",
       "      <td>NFKB1;NFKBIA</td>\n",
       "      <td>8.686945</td>\n",
       "      <td>BioPlanet 2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AP-1 transcription factor network</th>\n",
       "      <th>Candida</th>\n",
       "      <th>1</th>\n",
       "      <td>2/52</td>\n",
       "      <td>0.133319</td>\n",
       "      <td>0.233126</td>\n",
       "      <td>BCL2L11;FOS</td>\n",
       "      <td>3.165834</td>\n",
       "      <td>BioPlanet 2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATF2 transcription factor network</th>\n",
       "      <th>Candida</th>\n",
       "      <th>1</th>\n",
       "      <td>2/44</td>\n",
       "      <td>0.101181</td>\n",
       "      <td>0.199649</td>\n",
       "      <td>DUSP5;FOS</td>\n",
       "      <td>3.743802</td>\n",
       "      <td>BioPlanet 2019</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   Overlap   P-value  \\\n",
       "Term                              Stimulus Cluster                     \n",
       "4-1BB-dependent immune response   Candida  1          2/17  0.017993   \n",
       "ABC transporters                  Candida  1          1/26  0.274543   \n",
       "AKT signaling pathway             Candida  1          2/19  0.022265   \n",
       "AP-1 transcription factor network Candida  1          2/52  0.133319   \n",
       "ATF2 transcription factor network Candida  1          2/44  0.101181   \n",
       "\n",
       "                                                    Adjusted P-value  \\\n",
       "Term                              Stimulus Cluster                     \n",
       "4-1BB-dependent immune response   Candida  1                0.057914   \n",
       "ABC transporters                  Candida  1                0.360688   \n",
       "AKT signaling pathway             Candida  1                0.065900   \n",
       "AP-1 transcription factor network Candida  1                0.233126   \n",
       "ATF2 transcription factor network Candida  1                0.199649   \n",
       "\n",
       "                                                           Genes  Odds Ratio  \\\n",
       "Term                              Stimulus Cluster                             \n",
       "4-1BB-dependent immune response   Candida  1        NFKB1;NFKBIA    9.710466   \n",
       "ABC transporters                  Candida  1               ABCC5    3.151861   \n",
       "AKT signaling pathway             Candida  1        NFKB1;NFKBIA    8.686945   \n",
       "AP-1 transcription factor network Candida  1         BCL2L11;FOS    3.165834   \n",
       "ATF2 transcription factor network Candida  1           DUSP5;FOS    3.743802   \n",
       "\n",
       "                                                          Database  \n",
       "Term                              Stimulus Cluster                  \n",
       "4-1BB-dependent immune response   Candida  1        BioPlanet 2019  \n",
       "ABC transporters                  Candida  1        BioPlanet 2019  \n",
       "AKT signaling pathway             Candida  1        BioPlanet 2019  \n",
       "AP-1 transcription factor network Candida  1        BioPlanet 2019  \n",
       "ATF2 transcription factor network Candida  1        BioPlanet 2019  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Enrichment results of time-series clusters (term X cluster_statistic)\n",
    "ts_enr_dfs = []\n",
    "\n",
    "for treatment in metadata[\"Treatment\"].unique():\n",
    "    if treatment==\"untreated\":\n",
    "        continue\n",
    "        \n",
    "    k_tmp = timeseries_k.loc[treatment,\"RNA\"]\n",
    "\n",
    "    for cluster in range(1,k_tmp+1):\n",
    "        # load enr results\n",
    "        df_tmp = pd.read_csv(os.path.join(data_path, 'enrichment_analysis', 'time_series', treatment, 'k_{}'.format(k_tmp), 'cluster_{}'.format(cluster), 'BioPlanet_2019', 'Enrichr_BioPlanet_2019_cluster_{}.csv'.format(cluster)))\n",
    "\n",
    "        # Drop columns\n",
    "        df_tmp = df_tmp.drop(['Unnamed: 0', 'Gene_set'], axis=1)\n",
    "    #     df_tmp = df_tmp.drop(['Unnamed: 0', 'Gene_set', 'Genes'], axis=1)\n",
    "\n",
    "        # Rename columns \n",
    "    #     df_tmp.columns = [\"cluster_{}_{}\".format(cluster,col) for col in df_tmp.columns]\n",
    "\n",
    "        # add cluster column\n",
    "        df_tmp[\"Treatment\"] = treatment\n",
    "        df_tmp[\"Cluster\"] = cluster\n",
    "        \n",
    "        # set indices\n",
    "        df_tmp.set_index([\"Term\", \"Treatment\", \"Cluster\"], inplace=True)\n",
    "\n",
    "        ts_enr_dfs.append(df_tmp)\n",
    "\n",
    "ts_enr_df = pd.concat(ts_enr_dfs, axis=0)\n",
    "# ts_enr_df = pd.concat(ts_enr_dfs, axis=1)\n",
    "\n",
    "# add database column\n",
    "ts_enr_df[\"Database\"] = 'BioPlanet 2019'\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = ts_enr_df,\n",
    "                     sheet_name = \"time_series_enrichments\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "print(ts_enr_df.shape)\n",
    "ts_enr_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "a001b28b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save Excel file\n",
    "writer.save()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2901b507",
   "metadata": {},
   "source": [
    "# ATAC - TableS2\n",
    "- Homer_gene annotations in DEA and ts clustering results are not indices as soemtimes NA/Inf related error in XLSX writer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "0ca9167c",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = os.path.join('results', 'ATAC', 'all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "b8c4bd17",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create Excel writer\n",
    "writer = pd.ExcelWriter(os.path.join(results_path,'S02_ATACseq.xlsx'), engine='xlsxwriter')\n",
    "\n",
    "# Get the workbook and worksheet objects\n",
    "workbook = writer.book\n",
    "# define specific formatting\n",
    "index_format = workbook.add_format(xlsx_format_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "a53e9973",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add an empty sheet called LEGEND (fill with manually curated data later)\n",
    "pd.DataFrame().to_excel(writer, sheet_name=\"TableS2_legends\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "bf0b8663",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(78, 26)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>library</th>\n",
       "      <th>timepoint after thawing</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>Treatment</th>\n",
       "      <th>Treatment_time</th>\n",
       "      <th>sex</th>\n",
       "      <th>bio-replicate</th>\n",
       "      <th>organism</th>\n",
       "      <th>experiment_id</th>\n",
       "      <th>M Reads</th>\n",
       "      <th>...</th>\n",
       "      <th>Fragment Length</th>\n",
       "      <th>Treatment Redundancy</th>\n",
       "      <th>% Dups</th>\n",
       "      <th>% Aligned</th>\n",
       "      <th>% Duplication</th>\n",
       "      <th>% &gt; Q30</th>\n",
       "      <th>Mb Q30 bases</th>\n",
       "      <th>GC content</th>\n",
       "      <th>% PF</th>\n",
       "      <th>% Adapter</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PT76_R1_C_albicans_2h</th>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>d010</td>\n",
       "      <td>BMDM</td>\n",
       "      <td>C_albicans</td>\n",
       "      <td>2h</td>\n",
       "      <td>female</td>\n",
       "      <td>R1</td>\n",
       "      <td>mouse</td>\n",
       "      <td>PT76</td>\n",
       "      <td>12.7</td>\n",
       "      <td>...</td>\n",
       "      <td>147.0</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.97</td>\n",
       "      <td>609.5</td>\n",
       "      <td>0.51</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PT76_R1_C_albicans_4h</th>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>d010</td>\n",
       "      <td>BMDM</td>\n",
       "      <td>C_albicans</td>\n",
       "      <td>4h</td>\n",
       "      <td>female</td>\n",
       "      <td>R1</td>\n",
       "      <td>mouse</td>\n",
       "      <td>PT76</td>\n",
       "      <td>15.5</td>\n",
       "      <td>...</td>\n",
       "      <td>147.0</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.15</td>\n",
       "      <td>0.97</td>\n",
       "      <td>748.6</td>\n",
       "      <td>0.50</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PT76_R1_C_albicans_6h</th>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>d010</td>\n",
       "      <td>BMDM</td>\n",
       "      <td>C_albicans</td>\n",
       "      <td>6h</td>\n",
       "      <td>female</td>\n",
       "      <td>R1</td>\n",
       "      <td>mouse</td>\n",
       "      <td>PT76</td>\n",
       "      <td>14.0</td>\n",
       "      <td>...</td>\n",
       "      <td>147.0</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.17</td>\n",
       "      <td>0.96</td>\n",
       "      <td>675.5</td>\n",
       "      <td>0.51</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PT76_R1_C_albicans_8h</th>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>d010</td>\n",
       "      <td>BMDM</td>\n",
       "      <td>C_albicans</td>\n",
       "      <td>8h</td>\n",
       "      <td>female</td>\n",
       "      <td>R1</td>\n",
       "      <td>mouse</td>\n",
       "      <td>PT76</td>\n",
       "      <td>15.2</td>\n",
       "      <td>...</td>\n",
       "      <td>147.0</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.97</td>\n",
       "      <td>733.7</td>\n",
       "      <td>0.50</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PT76_R1_untreated_0h</th>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>d010</td>\n",
       "      <td>BMDM</td>\n",
       "      <td>untreated</td>\n",
       "      <td>0h</td>\n",
       "      <td>female</td>\n",
       "      <td>R1</td>\n",
       "      <td>mouse</td>\n",
       "      <td>PT76</td>\n",
       "      <td>16.0</td>\n",
       "      <td>...</td>\n",
       "      <td>147.0</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.97</td>\n",
       "      <td>769.1</td>\n",
       "      <td>0.50</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        library timepoint after thawing cell_line   Treatment  \\\n",
       "sample                                                                          \n",
       "PT76_R1_C_albicans_2h  ATAC-seq                    d010      BMDM  C_albicans   \n",
       "PT76_R1_C_albicans_4h  ATAC-seq                    d010      BMDM  C_albicans   \n",
       "PT76_R1_C_albicans_6h  ATAC-seq                    d010      BMDM  C_albicans   \n",
       "PT76_R1_C_albicans_8h  ATAC-seq                    d010      BMDM  C_albicans   \n",
       "PT76_R1_untreated_0h   ATAC-seq                    d010      BMDM   untreated   \n",
       "\n",
       "                      Treatment_time     sex bio-replicate organism  \\\n",
       "sample                                                                \n",
       "PT76_R1_C_albicans_2h             2h  female            R1    mouse   \n",
       "PT76_R1_C_albicans_4h             4h  female            R1    mouse   \n",
       "PT76_R1_C_albicans_6h             6h  female            R1    mouse   \n",
       "PT76_R1_C_albicans_8h             8h  female            R1    mouse   \n",
       "PT76_R1_untreated_0h              0h  female            R1    mouse   \n",
       "\n",
       "                      experiment_id  M Reads  ...  Fragment Length  \\\n",
       "sample                                        ...                    \n",
       "PT76_R1_C_albicans_2h          PT76     12.7  ...            147.0   \n",
       "PT76_R1_C_albicans_4h          PT76     15.5  ...            147.0   \n",
       "PT76_R1_C_albicans_6h          PT76     14.0  ...            147.0   \n",
       "PT76_R1_C_albicans_8h          PT76     15.2  ...            147.0   \n",
       "PT76_R1_untreated_0h           PT76     16.0  ...            147.0   \n",
       "\n",
       "                       Treatment Redundancy  % Dups  % Aligned  % Duplication  \\\n",
       "sample                                                                          \n",
       "PT76_R1_C_albicans_2h                  0.06    0.19       0.99           0.14   \n",
       "PT76_R1_C_albicans_4h                  0.06    0.19       0.99           0.15   \n",
       "PT76_R1_C_albicans_6h                  0.08    0.22       0.99           0.17   \n",
       "PT76_R1_C_albicans_8h                  0.06    0.19       0.99           0.14   \n",
       "PT76_R1_untreated_0h                   0.06    0.18       0.99           0.14   \n",
       "\n",
       "                       % > Q30  Mb Q30 bases  GC content  % PF  % Adapter  \n",
       "sample                                                                     \n",
       "PT76_R1_C_albicans_2h     0.97         609.5        0.51   1.0       0.10  \n",
       "PT76_R1_C_albicans_4h     0.97         748.6        0.50   1.0       0.12  \n",
       "PT76_R1_C_albicans_6h     0.96         675.5        0.51   1.0       0.10  \n",
       "PT76_R1_C_albicans_8h     0.97         733.7        0.50   1.0       0.11  \n",
       "PT76_R1_untreated_0h      0.97         769.1        0.50   1.0       0.13  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# metadata\n",
    "metadata = pd.read_csv(os.path.join('metadata', 'ATAC_sample_metadata.csv'), index_col=0)\n",
    "# load statistics\n",
    "statistics = pd.read_csv(os.path.join('metadata', 'ATAC_multiqc_report.csv'), index_col=0)\n",
    "# concatenate dataframes\n",
    "metadata = pd.concat([metadata, statistics], axis=1)\n",
    "# remove unused columns\n",
    "metadata.drop(['mouse_pool','flowcell'], axis=1, inplace=True)\n",
    "# remove experiment id from sample name\n",
    "# metadata.index = metadata.index.astype(str).str.replace('PT\\d{2}', 'ATAC', regex=True)\n",
    "# add index name\n",
    "metadata.index.names = [\"sample\"]\n",
    "\n",
    "metadata_copy = metadata.copy(deep=True)\n",
    "metadata_copy.index = \"ATAC_\"+metadata.index\n",
    "metadata_copy.index = metadata_copy.index.map(apply_bulk_sample_name_replacements)\n",
    "metadata_copy.index = [rearrange_sample_names(col) for col in metadata_copy.index]\n",
    "metadata_copy.index.names = [\"sample\"]\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = metadata_copy,\n",
    "                     sheet_name = \"sample_annotation\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "print(metadata.shape)\n",
    "metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "9f859dfb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(136735, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Homer_gene</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>region</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CONS00000000005</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4414052</td>\n",
       "      <td>4415256</td>\n",
       "      <td>Rp1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CONS00000000006</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4416568</td>\n",
       "      <td>4417221</td>\n",
       "      <td>Rp1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CONS00000000011</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4491871</td>\n",
       "      <td>4492835</td>\n",
       "      <td>Sox17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CONS00000000013</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4622075</td>\n",
       "      <td>4622934</td>\n",
       "      <td>Sox17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CONS00000000017</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4755824</td>\n",
       "      <td>4756551</td>\n",
       "      <td>Mrpl15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Chromosome    Start      End Homer_gene\n",
       "region                                                 \n",
       "CONS00000000005       chr1  4414052  4415256        Rp1\n",
       "CONS00000000006       chr1  4416568  4417221        Rp1\n",
       "CONS00000000011       chr1  4491871  4492835      Sox17\n",
       "CONS00000000013       chr1  4622075  4622934      Sox17\n",
       "CONS00000000017       chr1  4755824  4756551     Mrpl15"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# consensus region annotation\n",
    "region_annot = pd.read_csv(os.path.join(data_path, 'all_consensus_regions_annotation_filtered.csv'))\n",
    "\n",
    "# Rename column from 'rn' to 'gene'\n",
    "region_annot.rename(columns={'Unnamed: 0': 'region'}, inplace=True)\n",
    "\n",
    "# set index\n",
    "region_annot.set_index([\"region\"], inplace=True)\n",
    "\n",
    "# filter & rename columns\n",
    "region_annot = region_annot.loc[:,['gencode_chr','gencode_start','gencode_end', 'homer_Gene Name']]\n",
    "region_annot.columns = ['Chromosome','Start','End','Homer_gene']\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = region_annot.copy(deep=True),\n",
    "                     sheet_name = \"consensus_regions\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "print(region_annot.shape)\n",
    "region_annot.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "a33a24b7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(176075, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>LogFC</th>\n",
       "      <th>Average expression</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Homer gene</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Region</th>\n",
       "      <th>Comparison</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">CONS00000130796</th>\n",
       "      <th>Candida_2h</th>\n",
       "      <td>1.173784</td>\n",
       "      <td>4.369844</td>\n",
       "      <td>2.008776e-15</td>\n",
       "      <td>2.746700e-10</td>\n",
       "      <td>Prdx1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Candida_4h</th>\n",
       "      <td>0.696071</td>\n",
       "      <td>4.369844</td>\n",
       "      <td>4.387840e-07</td>\n",
       "      <td>3.203916e-04</td>\n",
       "      <td>Prdx1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Candida_6h</th>\n",
       "      <td>0.640373</td>\n",
       "      <td>4.369844</td>\n",
       "      <td>4.282775e-06</td>\n",
       "      <td>1.796335e-03</td>\n",
       "      <td>Prdx1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Candida_8h</th>\n",
       "      <td>0.592594</td>\n",
       "      <td>4.369844</td>\n",
       "      <td>2.730619e-08</td>\n",
       "      <td>1.598495e-05</td>\n",
       "      <td>Prdx1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Candida_24h</th>\n",
       "      <td>0.283090</td>\n",
       "      <td>4.369844</td>\n",
       "      <td>1.370579e-02</td>\n",
       "      <td>7.027565e-01</td>\n",
       "      <td>Prdx1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                LogFC  Average expression       P-value  \\\n",
       "Region          Comparison                                                \n",
       "CONS00000130796 Candida_2h   1.173784            4.369844  2.008776e-15   \n",
       "                Candida_4h   0.696071            4.369844  4.387840e-07   \n",
       "                Candida_6h   0.640373            4.369844  4.282775e-06   \n",
       "                Candida_8h   0.592594            4.369844  2.730619e-08   \n",
       "                Candida_24h  0.283090            4.369844  1.370579e-02   \n",
       "\n",
       "                             Adjusted P-value Homer gene  \n",
       "Region          Comparison                                \n",
       "CONS00000130796 Candida_2h       2.746700e-10      Prdx1  \n",
       "                Candida_4h       3.203916e-04      Prdx1  \n",
       "                Candida_6h       1.796335e-03      Prdx1  \n",
       "                Candida_8h       1.598495e-05      Prdx1  \n",
       "                Candida_24h      7.027565e-01      Prdx1  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# DEA results\n",
    "dea_dfs = []\n",
    "\n",
    "for treatment in metadata[\"Treatment\"].unique():\n",
    "    # load dea results\n",
    "    df_tmp = pd.read_csv(os.path.join(data_path, 'DEA', 'DEA_{}.tsv'.format(treatment)), delimiter=\"\\t\")\n",
    "\n",
    "    # Drop columns 'B' and 't'\n",
    "    df_tmp = df_tmp.drop(['B', 't'], axis=1)\n",
    "\n",
    "    # Rename column from 'rn' to 'gene'\n",
    "    df_tmp.rename(columns={'rn': 'region'}, inplace=True)\n",
    "    \n",
    "     # Add gene symbols using region annotation\n",
    "    df_tmp = pd.merge(df_tmp, region_annot[['Homer_gene']], left_on='region', right_index=True, how='left')\n",
    "\n",
    "    # Pivot the data\n",
    "#     df_tmp = df_tmp.pivot(index='gene', columns='group')\n",
    "\n",
    "    # Flatten the multi-index columns and reverse the order of concatenation\n",
    "#     df_tmp.columns = ['_'.join(col[::-1]).strip() for col in df_pivot.columns.values]\n",
    "\n",
    "    # set indices\n",
    "    df_tmp.set_index([\"region\", \"group\"], inplace=True)\n",
    "    \n",
    "    # filter regions\n",
    "    if treatment!='C_albicans':\n",
    "        regions_sig = df_tmp.index[(df_tmp[\"adj.P.Val\"]<0.05) & (abs(df_tmp[\"logFC\"])>2) & (df_tmp[\"AveExpr\"]>0)].get_level_values('region').unique().tolist()\n",
    "    else:\n",
    "        regions_sig = df_tmp.index[(df_tmp[\"adj.P.Val\"]<0.05) & (abs(df_tmp[\"logFC\"])>1) & (df_tmp[\"AveExpr\"]>0)].get_level_values('region').unique().tolist()\n",
    "    \n",
    "    df_tmp = df_tmp.loc[regions_sig,:]\n",
    "    \n",
    "    dea_dfs.append(df_tmp)\n",
    "\n",
    "# dea_df = pd.concat(dea_dfs, axis=1)\n",
    "dea_df = pd.concat(dea_dfs, axis=0)\n",
    "\n",
    "# # filter for regions that at least in one comparison fulfill the standard criteria -> switched to treatment-wise filtering to reduce size\n",
    "# regions_sig = dea_df.index[(dea_df[\"adj.P.Val\"]<0.05) & (abs(dea_df[\"logFC\"])>2) & (dea_df[\"AveExpr\"]>0)].get_level_values('region').unique().tolist()\n",
    "# # find C albicans specific regions with more leniant filtering\n",
    "# Calbicans_regions_sig = dea_df.index[(dea_df[\"adj.P.Val\"]<0.05) & (abs(dea_df[\"logFC\"])>1) & (dea_df[\"AveExpr\"]>0) & (dea_df.index.get_level_values('group').str.contains('C_albicans'))].get_level_values('region').unique().tolist()\n",
    "# # make union of final relevant regions\n",
    "# regions_sig = list(set(regions_sig).union(Calbicans_regions_sig))\n",
    "# dea_df = dea_df.loc[regions_sig,:]\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = dea_df,\n",
    "                     sheet_name = \"differential_accessibility\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "print(dea_df.shape)\n",
    "dea_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "e55c3887",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(12222, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Hyper Fold Enrichment</th>\n",
       "      <th>Hyper Raw PValue</th>\n",
       "      <th>Hyper Adjp BH</th>\n",
       "      <th>Database</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Term</th>\n",
       "      <th>Cluster</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>regulation of innate immune response (GO:0045088)</th>\n",
       "      <th>1</th>\n",
       "      <td>1.947448</td>\n",
       "      <td>1.769698e-14</td>\n",
       "      <td>2.316535e-10</td>\n",
       "      <td>GO Biological Process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>immune response (GO:0006955)</th>\n",
       "      <th>1</th>\n",
       "      <td>1.419579</td>\n",
       "      <td>1.906900e-11</td>\n",
       "      <td>1.248066e-07</td>\n",
       "      <td>GO Biological Process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>regulation of response to stress (GO:0080134)</th>\n",
       "      <th>1</th>\n",
       "      <td>1.310585</td>\n",
       "      <td>8.841341e-11</td>\n",
       "      <td>3.857772e-07</td>\n",
       "      <td>GO Biological Process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>regulation of defense response (GO:0031347)</th>\n",
       "      <th>1</th>\n",
       "      <td>1.482990</td>\n",
       "      <td>4.155906e-10</td>\n",
       "      <td>1.322796e-06</td>\n",
       "      <td>GO Biological Process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>innate immune response (GO:0045087)</th>\n",
       "      <th>1</th>\n",
       "      <td>1.581043</td>\n",
       "      <td>5.202335e-10</td>\n",
       "      <td>1.322796e-06</td>\n",
       "      <td>GO Biological Process</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                           Hyper Fold Enrichment  \\\n",
       "Term                                              Cluster                          \n",
       "regulation of innate immune response (GO:0045088) 1                     1.947448   \n",
       "immune response (GO:0006955)                      1                     1.419579   \n",
       "regulation of response to stress (GO:0080134)     1                     1.310585   \n",
       "regulation of defense response (GO:0031347)       1                     1.482990   \n",
       "innate immune response (GO:0045087)               1                     1.581043   \n",
       "\n",
       "                                                           Hyper Raw PValue  \\\n",
       "Term                                              Cluster                     \n",
       "regulation of innate immune response (GO:0045088) 1            1.769698e-14   \n",
       "immune response (GO:0006955)                      1            1.906900e-11   \n",
       "regulation of response to stress (GO:0080134)     1            8.841341e-11   \n",
       "regulation of defense response (GO:0031347)       1            4.155906e-10   \n",
       "innate immune response (GO:0045087)               1            5.202335e-10   \n",
       "\n",
       "                                                           Hyper Adjp BH  \\\n",
       "Term                                              Cluster                  \n",
       "regulation of innate immune response (GO:0045088) 1         2.316535e-10   \n",
       "immune response (GO:0006955)                      1         1.248066e-07   \n",
       "regulation of response to stress (GO:0080134)     1         3.857772e-07   \n",
       "regulation of defense response (GO:0031347)       1         1.322796e-06   \n",
       "innate immune response (GO:0045087)               1         1.322796e-06   \n",
       "\n",
       "                                                                        Database  \n",
       "Term                                              Cluster                         \n",
       "regulation of innate immune response (GO:0045088) 1        GO Biological Process  \n",
       "immune response (GO:0006955)                      1        GO Biological Process  \n",
       "regulation of response to stress (GO:0080134)     1        GO Biological Process  \n",
       "regulation of defense response (GO:0031347)       1        GO Biological Process  \n",
       "innate immune response (GO:0045087)               1        GO Biological Process  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Enrichment results of DEG clusters (term X cluster_statistic)\n",
    "enr_dfs = []\n",
    "\n",
    "for cluster in range(1,10):\n",
    "    # load enr results\n",
    "    df_tmp = pd.read_csv(os.path.join(data_path, 'enrichment_analysis', 'DAR_clusters', 'k9_cluster_{}'.format(cluster), 'GREAT', 'GREAT_GO Biological Process.tsv'), sep=\"\\t\")\n",
    "\n",
    "    # Rename columns \n",
    "#     df_tmp.columns = [\"cluster_{}_{}\".format(cluster,col) for col in df_tmp.columns]\n",
    "\n",
    "    # add cluster column\n",
    "    df_tmp[\"Cluster\"] = cluster\n",
    "    \n",
    "    # set Term = name+ID\n",
    "    df_tmp[\"Term\"] = df_tmp[\"name\"]+' ('+df_tmp[\"ID\"]+')'\n",
    "    \n",
    "    # Drop columns\n",
    "#     df_tmp = df_tmp.drop(['name', 'ID'], axis=1)\n",
    "    # keep only selecteed columns\n",
    "    df_tmp = df_tmp.loc[:,['Term','Cluster','Hyper_Fold_Enrichment','Hyper_Raw_PValue','Hyper_Adjp_BH']]\n",
    "    \n",
    "    # set indices\n",
    "    df_tmp.set_index([\"Term\", \"Cluster\"], inplace=True)\n",
    "    \n",
    "    enr_dfs.append(df_tmp)\n",
    "\n",
    "enr_df = pd.concat(enr_dfs, axis=0)\n",
    "# enr_df = pd.concat(enr_dfs, axis=1)\n",
    "\n",
    "# add database column\n",
    "enr_df[\"Database\"] = 'GO Biological Process'\n",
    "\n",
    "# filter enrichment results by significance ie keep only terms that have at least one significant value\n",
    "significant_terms = enr_df.groupby(level='Term').apply(lambda group: (group['Hyper_Adjp_BH'] < 0.05).any())\n",
    "significant_terms = significant_terms[significant_terms]  \n",
    "enr_df = enr_df.loc[enr_df.index.get_level_values('Term').isin(significant_terms.index)]\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = enr_df,\n",
    "                     sheet_name = \"diff_accessibility_enrichments\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "print(enr_df.shape)\n",
    "enr_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "3801db40",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(19238, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Candida</th>\n",
       "      <th>IFN beta</th>\n",
       "      <th>IFN gamma</th>\n",
       "      <th>LCMV</th>\n",
       "      <th>Listeria</th>\n",
       "      <th>LPS</th>\n",
       "      <th>Homer gene</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Region</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CONS00000130796</th>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Prdx1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CONS00000032052</th>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Ccl5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CONS00000021736</th>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Tmcc3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CONS00000173048</th>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lpl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CONS00000017578</th>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Prdm1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Candida  IFN beta  IFN gamma  LCMV  Listeria  LPS Homer gene\n",
       "Region                                                                       \n",
       "CONS00000130796      2.0       NaN        NaN   NaN       NaN  NaN      Prdx1\n",
       "CONS00000032052      2.0       4.0        NaN   NaN       6.0  4.0       Ccl5\n",
       "CONS00000021736      2.0       NaN        NaN   NaN       NaN  NaN      Tmcc3\n",
       "CONS00000173048      2.0       5.0        NaN   NaN       NaN  NaN        Lpl\n",
       "CONS00000017578      2.0       NaN        NaN   NaN       NaN  NaN      Prdm1"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Time series clustering (gene X stimulus matrix with cluster membership or empty)\n",
    "ts_dfs = []\n",
    "\n",
    "for treatment in metadata[\"Treatment\"].unique():\n",
    "    if treatment==\"untreated\":\n",
    "        continue\n",
    "    \n",
    "    # load dea results\n",
    "    df_tmp = pd.read_csv(os.path.join(data_path, 'time_series', treatment, 'k_{}'.format(timeseries_k.loc[treatment,\"ATAC\"]), 'clustering_{}.csv'.format(treatment)), index_col=1)\n",
    "    \n",
    "    # Drop columns\n",
    "    df_tmp = df_tmp.drop(['Unnamed: 0'], axis=1)\n",
    "\n",
    "    # Rename columns \n",
    "    df_tmp.columns = [\"{}\".format(treatment)]\n",
    "    \n",
    "    ts_dfs.append(df_tmp)\n",
    "    \n",
    "ts_df = pd.concat(ts_dfs, axis=1)\n",
    "ts_df.index.names = [\"region\"]\n",
    "\n",
    "# Add gene symbols using region annotation\n",
    "ts_df = pd.merge(ts_df, region_annot[['Homer_gene']], left_on='region', right_index=True, how='left')\n",
    "\n",
    "# set indices\n",
    "# ts_df.reset_index(inplace=True)\n",
    "# ts_df.set_index([\"region\",\"Homer_gene\"], inplace=True)\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = ts_df,\n",
    "                     sheet_name = \"time_series_clusters\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "print(ts_df.shape)\n",
    "ts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "58b31776",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(59888, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Hyper Fold Enrichment</th>\n",
       "      <th>Hyper Raw PValue</th>\n",
       "      <th>Hyper Adjp BH</th>\n",
       "      <th>Database</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Term</th>\n",
       "      <th>Stimulus</th>\n",
       "      <th>Cluster</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>regulation of substrate adhesion-dependent cell spreading (GO:1900024)</th>\n",
       "      <th>Candida</th>\n",
       "      <th>1</th>\n",
       "      <td>4.858430</td>\n",
       "      <td>7.389429e-07</td>\n",
       "      <td>0.009673</td>\n",
       "      <td>GO Biological Process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>negative regulation of amyloid precursor protein biosynthetic process (GO:0042985)</th>\n",
       "      <th>Candida</th>\n",
       "      <th>1</th>\n",
       "      <td>13.947210</td>\n",
       "      <td>2.112510e-04</td>\n",
       "      <td>0.420393</td>\n",
       "      <td>GO Biological Process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>macromolecule catabolic process (GO:0009057)</th>\n",
       "      <th>Candida</th>\n",
       "      <th>1</th>\n",
       "      <td>1.727161</td>\n",
       "      <td>3.556155e-04</td>\n",
       "      <td>0.479560</td>\n",
       "      <td>GO Biological Process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>positive regulation of cell adhesion (GO:0045785)</th>\n",
       "      <th>Candida</th>\n",
       "      <th>1</th>\n",
       "      <td>1.765998</td>\n",
       "      <td>8.793954e-04</td>\n",
       "      <td>0.639516</td>\n",
       "      <td>GO Biological Process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cellular macromolecule catabolic process (GO:0044265)</th>\n",
       "      <th>Candida</th>\n",
       "      <th>1</th>\n",
       "      <td>1.707647</td>\n",
       "      <td>9.779085e-04</td>\n",
       "      <td>0.673727</td>\n",
       "      <td>GO Biological Process</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                     Hyper Fold Enrichment  \\\n",
       "Term                                               Stimulus Cluster                          \n",
       "regulation of substrate adhesion-dependent cell... Candida  1                     4.858430   \n",
       "negative regulation of amyloid precursor protei... Candida  1                    13.947210   \n",
       "macromolecule catabolic process (GO:0009057)       Candida  1                     1.727161   \n",
       "positive regulation of cell adhesion (GO:0045785)  Candida  1                     1.765998   \n",
       "cellular macromolecule catabolic process (GO:00... Candida  1                     1.707647   \n",
       "\n",
       "                                                                     Hyper Raw PValue  \\\n",
       "Term                                               Stimulus Cluster                     \n",
       "regulation of substrate adhesion-dependent cell... Candida  1            7.389429e-07   \n",
       "negative regulation of amyloid precursor protei... Candida  1            2.112510e-04   \n",
       "macromolecule catabolic process (GO:0009057)       Candida  1            3.556155e-04   \n",
       "positive regulation of cell adhesion (GO:0045785)  Candida  1            8.793954e-04   \n",
       "cellular macromolecule catabolic process (GO:00... Candida  1            9.779085e-04   \n",
       "\n",
       "                                                                     Hyper Adjp BH  \\\n",
       "Term                                               Stimulus Cluster                  \n",
       "regulation of substrate adhesion-dependent cell... Candida  1             0.009673   \n",
       "negative regulation of amyloid precursor protei... Candida  1             0.420393   \n",
       "macromolecule catabolic process (GO:0009057)       Candida  1             0.479560   \n",
       "positive regulation of cell adhesion (GO:0045785)  Candida  1             0.639516   \n",
       "cellular macromolecule catabolic process (GO:00... Candida  1             0.673727   \n",
       "\n",
       "                                                                                  Database  \n",
       "Term                                               Stimulus Cluster                         \n",
       "regulation of substrate adhesion-dependent cell... Candida  1        GO Biological Process  \n",
       "negative regulation of amyloid precursor protei... Candida  1        GO Biological Process  \n",
       "macromolecule catabolic process (GO:0009057)       Candida  1        GO Biological Process  \n",
       "positive regulation of cell adhesion (GO:0045785)  Candida  1        GO Biological Process  \n",
       "cellular macromolecule catabolic process (GO:00... Candida  1        GO Biological Process  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Enrichment results of time-series clusters (term X cluster_statistic)\n",
    "ts_enr_dfs = []\n",
    "\n",
    "for treatment in metadata[\"Treatment\"].unique():\n",
    "    if treatment==\"untreated\":\n",
    "        continue\n",
    "        \n",
    "    k_tmp = timeseries_k.loc[treatment,\"ATAC\"]\n",
    "\n",
    "    for cluster in range(1,k_tmp+1):\n",
    "        # load enr results\n",
    "        df_tmp = pd.read_csv(os.path.join(data_path, 'enrichment_analysis', 'time_series', '{}_k{}_cluster_{}'.format(treatment, k_tmp, cluster), 'GREAT', 'GREAT_GO Biological Process.tsv'), sep=\"\\t\")\n",
    "\n",
    "        # Rename columns \n",
    "    #     df_tmp.columns = [\"cluster_{}_{}\".format(cluster,col) for col in df_tmp.columns]\n",
    "\n",
    "        # add cluster column\n",
    "        df_tmp[\"Treatment\"] = treatment\n",
    "        df_tmp[\"Cluster\"] = cluster\n",
    "        \n",
    "        # set Term = name+ID\n",
    "        df_tmp[\"Term\"] = df_tmp[\"name\"]+' ('+df_tmp[\"ID\"]+')'\n",
    "    \n",
    "        # Drop columns\n",
    "#         df_tmp = df_tmp.drop(['name', 'ID'], axis=1)\n",
    "        # keep only selecteed columns\n",
    "        df_tmp = df_tmp.loc[:,['Term','Treatment','Cluster','Hyper_Fold_Enrichment','Hyper_Raw_PValue','Hyper_Adjp_BH']]\n",
    "    \n",
    "        # set indices\n",
    "        df_tmp.set_index([\"Term\", \"Treatment\", \"Cluster\"], inplace=True)\n",
    "\n",
    "        ts_enr_dfs.append(df_tmp)\n",
    "\n",
    "ts_enr_df = pd.concat(ts_enr_dfs, axis=0)\n",
    "# ts_enr_df = pd.concat(ts_enr_dfs, axis=1)\n",
    "\n",
    "# add database column\n",
    "ts_enr_df[\"Database\"] = 'GO Biological Process'\n",
    "\n",
    "# filter enrichment results by significance ie keep only terms that have at least one significant value\n",
    "significant_terms = ts_enr_df.groupby(level='Term').apply(lambda group: (group['Hyper_Adjp_BH'] < 0.05).any())\n",
    "significant_terms = significant_terms[significant_terms]  \n",
    "ts_enr_df = ts_enr_df.loc[ts_enr_df.index.get_level_values('Term').isin(significant_terms.index)]\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = ts_enr_df,\n",
    "                     sheet_name = \"time_series_enrichments\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "print(ts_enr_df.shape)\n",
    "ts_enr_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "3210479a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save Excel file\n",
    "writer.save()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d264c10",
   "metadata": {},
   "source": [
    "# INT - TableS3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "e6f07d86",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = os.path.join('results', 'INT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "618152c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create Excel writer\n",
    "writer = pd.ExcelWriter(os.path.join(results_path,'S03_RNAseq_ATACseq_integrative_analysis.xlsx'), engine='xlsxwriter')\n",
    "\n",
    "# Get the workbook and worksheet objects\n",
    "workbook = writer.book\n",
    "# define specific formatting\n",
    "index_format = workbook.add_format(xlsx_format_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "5f6f866e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add an empty sheet called LEGEND (fill with manually curated data later)\n",
    "pd.DataFrame().to_excel(writer, sheet_name=\"TableS3_legends\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "b4ed50bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(225122, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>LogFC</th>\n",
       "      <th>Average expression</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Comparison</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">ENSMUSG00000074141</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">Il4i1</th>\n",
       "      <th>Candida 24h</th>\n",
       "      <td>3.248002</td>\n",
       "      <td>3.743529</td>\n",
       "      <td>3.703750e-12</td>\n",
       "      <td>8.346216e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Candida 2h</th>\n",
       "      <td>0.039455</td>\n",
       "      <td>3.743529</td>\n",
       "      <td>9.081105e-01</td>\n",
       "      <td>9.886195e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Candida 4h</th>\n",
       "      <td>0.171906</td>\n",
       "      <td>3.743529</td>\n",
       "      <td>6.198277e-01</td>\n",
       "      <td>9.499374e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Candida 6h</th>\n",
       "      <td>0.307128</td>\n",
       "      <td>3.743529</td>\n",
       "      <td>3.742488e-01</td>\n",
       "      <td>9.231326e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Candida 8h</th>\n",
       "      <td>1.448020</td>\n",
       "      <td>3.743529</td>\n",
       "      <td>2.089307e-04</td>\n",
       "      <td>2.263534e-02</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          LogFC  Average expression  \\\n",
       "Gene               Symbol Comparison                                  \n",
       "ENSMUSG00000074141 Il4i1  Candida 24h  3.248002            3.743529   \n",
       "                          Candida 2h   0.039455            3.743529   \n",
       "                          Candida 4h   0.171906            3.743529   \n",
       "                          Candida 6h   0.307128            3.743529   \n",
       "                          Candida 8h   1.448020            3.743529   \n",
       "\n",
       "                                            P-value  Adjusted P-value  \n",
       "Gene               Symbol Comparison                                   \n",
       "ENSMUSG00000074141 Il4i1  Candida 24h  3.703750e-12      8.346216e-08  \n",
       "                          Candida 2h   9.081105e-01      9.886195e-01  \n",
       "                          Candida 4h   6.198277e-01      9.499374e-01  \n",
       "                          Candida 6h   3.742488e-01      9.231326e-01  \n",
       "                          Candida 8h   2.089307e-04      2.263534e-02  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# DEA results\n",
    "\n",
    "dea_df = pd.read_csv(os.path.join(data_path, 'DEA', 'INT_DEA_all.csv'))\n",
    "\n",
    "# Rename column from 'rn' to 'gene'\n",
    "dea_df.rename(columns={'rn': 'gene'}, inplace=True)\n",
    "\n",
    "# Add gene symbols using gene annotation\n",
    "dea_df = pd.merge(dea_df, gene_annot[['symbol']], left_on='gene', right_index=True, how='left')\n",
    "\n",
    "# set indices\n",
    "dea_df.set_index([\"gene\",\"symbol\", \"group\"], inplace=True)\n",
    "\n",
    "# Drop columns\n",
    "dea_df = dea_df.drop(['time', 'treatment'], axis=1)\n",
    "dea_df = dea_df.drop(['B', 't'], axis=1)\n",
    "\n",
    "# filter for genes that are at least in one comparison fulfill the criteria\n",
    "genes_sig = dea_df.index[(dea_df[\"adj.P.Val\"]<0.05) & (abs(dea_df[\"logFC\"])>1) & (dea_df[\"AveExpr\"]>1)].get_level_values('gene').unique().tolist()\n",
    "dea_df = dea_df.loc[genes_sig,:]\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = dea_df,\n",
    "                     sheet_name = \"divergent_gene_identification\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "print(dea_df.shape)\n",
    "dea_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "3e38e568",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(7252, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Candida</th>\n",
       "      <th>IFN beta</th>\n",
       "      <th>IFN gamma</th>\n",
       "      <th>LCMV</th>\n",
       "      <th>Listeria</th>\n",
       "      <th>LPS</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th>Symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028</th>\n",
       "      <th>Cdc45</th>\n",
       "      <td>NaN</td>\n",
       "      <td>B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000056</th>\n",
       "      <th>Narf</th>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000058</th>\n",
       "      <th>Cav2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000127</th>\n",
       "      <th>Fer</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000149</th>\n",
       "      <th>Gna12</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Candida IFN beta IFN gamma LCMV Listeria  LPS\n",
       "Gene               Symbol                                              \n",
       "ENSMUSG00000000028 Cdc45      NaN        B       NaN  NaN        G    B\n",
       "ENSMUSG00000000056 Narf       NaN        A       NaN  NaN      NaN  NaN\n",
       "ENSMUSG00000000058 Cav2       NaN      NaN       NaN  NaN      NaN    F\n",
       "ENSMUSG00000000127 Fer        NaN      NaN       NaN  NaN        D  NaN\n",
       "ENSMUSG00000000149 Gna12      NaN      NaN       NaN  NaN        E  NaN"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Time series clustering (gene X stimulus matrix with cluster membership or empty)\n",
    "ts_dfs = []\n",
    "\n",
    "# Create a mapping from numbers to letters for clusters\n",
    "num_to_letter = {1: 'A', 2: 'B', 3: 'C', 4: 'D', 5: 'E', 6: 'F', 7: 'G'}\n",
    "listeria_changes = {\n",
    "    'G': 'A',\n",
    "    'A': 'B',\n",
    "    'B': 'C',\n",
    "    'D': 'D',\n",
    "    'E': 'E',\n",
    "    'C': 'F',\n",
    "    'F': 'G',\n",
    "}\n",
    "\n",
    "for treatment in metadata[\"Treatment\"].unique():\n",
    "    if treatment==\"untreated\":\n",
    "        continue\n",
    "    \n",
    "    # load dea results\n",
    "    df_tmp = pd.read_csv(os.path.join(data_path, 'time_series', treatment, 'k_{}'.format(timeseries_k.loc[treatment,\"INT\"]), 'clustering_{}.csv'.format(treatment)), index_col=[1,2])\n",
    "    \n",
    "    # Drop columns\n",
    "    df_tmp = df_tmp.drop(['Unnamed: 0'], axis=1)\n",
    "\n",
    "    # Rename columns \n",
    "    df_tmp.columns = [\"{}\".format(treatment)]\n",
    "    \n",
    "    ts_dfs.append(df_tmp)\n",
    "    \n",
    "ts_df = pd.concat(ts_dfs, axis=1)\n",
    "ts_df.index.names = [\"gene\", \"symbol\"]\n",
    "\n",
    "# Replace cluster numbers with letters\n",
    "ts_df = ts_df.replace(num_to_letter)\n",
    "# replace Listeria clusters accordingly\n",
    "ts_df['LO28'] = ts_df['LO28'].replace(listeria_changes)\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = ts_df,\n",
    "                     sheet_name = \"time_series_clusters\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "print(ts_df.shape)\n",
    "ts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "5c047a9a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(57202, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Database</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Term</th>\n",
       "      <th>Stimulus</th>\n",
       "      <th>Cluster</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3'-UTR-mediated mRNA destabilization (GO:0061158)</th>\n",
       "      <th>Candida</th>\n",
       "      <th>A</th>\n",
       "      <td>1/16</td>\n",
       "      <td>0.032235</td>\n",
       "      <td>0.074873</td>\n",
       "      <td>30.880495</td>\n",
       "      <td>ZC3H12A</td>\n",
       "      <td>GO Biological Process 2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B cell activation (GO:0042113)</th>\n",
       "      <th>Candida</th>\n",
       "      <th>A</th>\n",
       "      <td>1/78</td>\n",
       "      <td>0.147724</td>\n",
       "      <td>0.178144</td>\n",
       "      <td>6.325726</td>\n",
       "      <td>CD40</td>\n",
       "      <td>GO Biological Process 2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B cell proliferation (GO:0042100)</th>\n",
       "      <th>Candida</th>\n",
       "      <th>A</th>\n",
       "      <td>1/26</td>\n",
       "      <td>0.051858</td>\n",
       "      <td>0.092375</td>\n",
       "      <td>18.999155</td>\n",
       "      <td>CD40</td>\n",
       "      <td>GO Biological Process 2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B cell receptor signaling pathway (GO:0050853)</th>\n",
       "      <th>Candida</th>\n",
       "      <th>A</th>\n",
       "      <td>1/32</td>\n",
       "      <td>0.063442</td>\n",
       "      <td>0.100254</td>\n",
       "      <td>15.434753</td>\n",
       "      <td>LCK</td>\n",
       "      <td>GO Biological Process 2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD40 signaling pathway (GO:0023035)</th>\n",
       "      <th>Candida</th>\n",
       "      <th>A</th>\n",
       "      <td>1/9</td>\n",
       "      <td>0.018261</td>\n",
       "      <td>0.065709</td>\n",
       "      <td>54.907204</td>\n",
       "      <td>CD40</td>\n",
       "      <td>GO Biological Process 2021</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                   Overlap  \\\n",
       "Term                                              Stimulus Cluster           \n",
       "3'-UTR-mediated mRNA destabilization (GO:0061158) Candida  A          1/16   \n",
       "B cell activation (GO:0042113)                    Candida  A          1/78   \n",
       "B cell proliferation (GO:0042100)                 Candida  A          1/26   \n",
       "B cell receptor signaling pathway (GO:0050853)    Candida  A          1/32   \n",
       "CD40 signaling pathway (GO:0023035)               Candida  A           1/9   \n",
       "\n",
       "                                                                     P-value  \\\n",
       "Term                                              Stimulus Cluster             \n",
       "3'-UTR-mediated mRNA destabilization (GO:0061158) Candida  A        0.032235   \n",
       "B cell activation (GO:0042113)                    Candida  A        0.147724   \n",
       "B cell proliferation (GO:0042100)                 Candida  A        0.051858   \n",
       "B cell receptor signaling pathway (GO:0050853)    Candida  A        0.063442   \n",
       "CD40 signaling pathway (GO:0023035)               Candida  A        0.018261   \n",
       "\n",
       "                                                                    Adjusted P-value  \\\n",
       "Term                                              Stimulus Cluster                     \n",
       "3'-UTR-mediated mRNA destabilization (GO:0061158) Candida  A                0.074873   \n",
       "B cell activation (GO:0042113)                    Candida  A                0.178144   \n",
       "B cell proliferation (GO:0042100)                 Candida  A                0.092375   \n",
       "B cell receptor signaling pathway (GO:0050853)    Candida  A                0.100254   \n",
       "CD40 signaling pathway (GO:0023035)               Candida  A                0.065709   \n",
       "\n",
       "                                                                    Odds Ratio  \\\n",
       "Term                                              Stimulus Cluster               \n",
       "3'-UTR-mediated mRNA destabilization (GO:0061158) Candida  A         30.880495   \n",
       "B cell activation (GO:0042113)                    Candida  A          6.325726   \n",
       "B cell proliferation (GO:0042100)                 Candida  A         18.999155   \n",
       "B cell receptor signaling pathway (GO:0050853)    Candida  A         15.434753   \n",
       "CD40 signaling pathway (GO:0023035)               Candida  A         54.907204   \n",
       "\n",
       "                                                                      Genes  \\\n",
       "Term                                              Stimulus Cluster            \n",
       "3'-UTR-mediated mRNA destabilization (GO:0061158) Candida  A        ZC3H12A   \n",
       "B cell activation (GO:0042113)                    Candida  A           CD40   \n",
       "B cell proliferation (GO:0042100)                 Candida  A           CD40   \n",
       "B cell receptor signaling pathway (GO:0050853)    Candida  A            LCK   \n",
       "CD40 signaling pathway (GO:0023035)               Candida  A           CD40   \n",
       "\n",
       "                                                                                      Database  \n",
       "Term                                              Stimulus Cluster                              \n",
       "3'-UTR-mediated mRNA destabilization (GO:0061158) Candida  A        GO Biological Process 2021  \n",
       "B cell activation (GO:0042113)                    Candida  A        GO Biological Process 2021  \n",
       "B cell proliferation (GO:0042100)                 Candida  A        GO Biological Process 2021  \n",
       "B cell receptor signaling pathway (GO:0050853)    Candida  A        GO Biological Process 2021  \n",
       "CD40 signaling pathway (GO:0023035)               Candida  A        GO Biological Process 2021  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Enrichment results of time-series clusters (term X cluster_statistic)\n",
    "ts_enr_dfs = []\n",
    "\n",
    "for treatment in metadata[\"Treatment\"].unique():\n",
    "    if treatment==\"untreated\":\n",
    "        continue\n",
    "        \n",
    "    k_tmp = timeseries_k.loc[treatment,\"INT\"]\n",
    "\n",
    "    for cluster in range(1,k_tmp+1):\n",
    "        # load enr results\n",
    "        df_tmp = pd.read_csv(os.path.join(data_path, 'enrichment_analysis', 'time_series', treatment, 'k_{}'.format(k_tmp), 'cluster_{}'.format(cluster), 'GO_Biological_Process_2021', 'Enrichr_GO_Biological_Process_2021_cluster_{}.csv'.format(cluster)))\n",
    "\n",
    "        # Drop columns\n",
    "        df_tmp = df_tmp.drop(['Unnamed: 0', 'Gene_set'], axis=1)\n",
    "    #     df_tmp = df_tmp.drop(['Unnamed: 0', 'Gene_set', 'Genes'], axis=1)\n",
    "\n",
    "        # Rename columns \n",
    "    #     df_tmp.columns = [\"cluster_{}_{}\".format(cluster,col) for col in df_tmp.columns]\n",
    "\n",
    "        # add cluster column\n",
    "        df_tmp[\"Treatment\"] = treatment\n",
    "        df_tmp[\"Cluster\"] = cluster\n",
    "        # rename clusters from numbers to letters\n",
    "        df_tmp['Cluster'] = df_tmp['Cluster'].replace(num_to_letter)\n",
    "        if treatment==\"LO28\":\n",
    "            df_tmp['Cluster'] = df_tmp['Cluster'].replace(listeria_changes)\n",
    "        \n",
    "        # set indices\n",
    "        df_tmp.set_index([\"Term\", \"Treatment\", \"Cluster\"], inplace=True)\n",
    "\n",
    "        ts_enr_dfs.append(df_tmp)\n",
    "\n",
    "ts_enr_df = pd.concat(ts_enr_dfs, axis=0)\n",
    "# ts_enr_df = pd.concat(ts_enr_dfs, axis=1)\n",
    "\n",
    "# add database column\n",
    "ts_enr_df[\"Database\"] = 'GO Biological Process 2021'\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = ts_enr_df,\n",
    "                     sheet_name = \"time_series_enrichments\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "print(ts_enr_df.shape)\n",
    "ts_enr_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "22ad615e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(6229, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Motif</th>\n",
       "      <th>NES</th>\n",
       "      <th>TF highConf</th>\n",
       "      <th>NEnrGenes</th>\n",
       "      <th>EnrichedGenes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Stimulus</th>\n",
       "      <th>Cluster</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">Candida</th>\n",
       "      <th>A</th>\n",
       "      <td>predrem__nrMotif347</td>\n",
       "      <td>4.73</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15</td>\n",
       "      <td>Arhgef3;Cdc42ep2;Chd7;Fmnl2;Gadd45b;Hivep3;Kcn...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>transfac_pro__M04860</td>\n",
       "      <td>4.55</td>\n",
       "      <td>Rxra (inferredBy_Orthology).</td>\n",
       "      <td>13</td>\n",
       "      <td>Abtb2;Arhgap28;Cd40;Ciita;Gadd45b;Gm12216;Kdm6...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>predrem__nrMotif82</td>\n",
       "      <td>4.37</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "      <td>Arhgap28;Fscn1;Hivep3;Lck;Lmo4;Lpar1;Mxd1;Prdm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>neph__UW.Motif.0464</td>\n",
       "      <td>4.24</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14</td>\n",
       "      <td>Abtb2;Arhgap28;Arhgef3;Gadd45b;Gm12216;Iqsec2;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>hdpi__ABCF2</td>\n",
       "      <td>4.00</td>\n",
       "      <td>Abcf2 (inferredBy_Orthology).</td>\n",
       "      <td>20</td>\n",
       "      <td>Abtb2;Arhgap28;Arhgef3;Cd40;Chd7;Ciita;Cish;Fc...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 Motif   NES                     TF highConf  \\\n",
       "Stimulus Cluster                                                               \n",
       "Candida  A         predrem__nrMotif347  4.73                             NaN   \n",
       "         A        transfac_pro__M04860  4.55   Rxra (inferredBy_Orthology).    \n",
       "         A          predrem__nrMotif82  4.37                             NaN   \n",
       "         A         neph__UW.Motif.0464  4.24                             NaN   \n",
       "         A                 hdpi__ABCF2  4.00  Abcf2 (inferredBy_Orthology).    \n",
       "\n",
       "                  NEnrGenes                                      EnrichedGenes  \n",
       "Stimulus Cluster                                                                \n",
       "Candida  A               15  Arhgef3;Cdc42ep2;Chd7;Fmnl2;Gadd45b;Hivep3;Kcn...  \n",
       "         A               13  Abtb2;Arhgap28;Cd40;Ciita;Gadd45b;Gm12216;Kdm6...  \n",
       "         A               11  Arhgap28;Fscn1;Hivep3;Lck;Lmo4;Lpar1;Mxd1;Prdm...  \n",
       "         A               14  Abtb2;Arhgap28;Arhgef3;Gadd45b;Gm12216;Iqsec2;...  \n",
       "         A               20  Abtb2;Arhgap28;Arhgef3;Cd40;Chd7;Ciita;Cish;Fc...  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TFBS Enrichment results of time-series clusters (term X cluster_statistic)\n",
    "tf_enr_dfs = []\n",
    "\n",
    "cluster_to_letter = {\"cluster_1\": 'A', \"cluster_2\": 'B', \"cluster_3\": 'C', \"cluster_4\": 'D', \"cluster_5\": 'E', \"cluster_6\": 'F', \"cluster_7\": 'G'}\n",
    "\n",
    "for treatment in metadata[\"Treatment\"].unique():\n",
    "    if treatment==\"untreated\":\n",
    "        continue\n",
    "        \n",
    "    k_tmp = timeseries_k.loc[treatment,\"INT\"]\n",
    "\n",
    "    # load TF enr results\n",
    "    df_tmp = pd.read_csv(os.path.join(data_path, 'TF', 'INT_TF_time_series_{}_k{}.csv'.format(treatment, k_tmp)))\n",
    "\n",
    "    # Drop columns\n",
    "    df_tmp = df_tmp.drop(['AUC', 'TF_lowConf','rankAtMax'], axis=1)\n",
    "\n",
    "    # Rename columns \n",
    "    df_tmp.rename(columns={\"geneSet\": \"Cluster\"}, inplace=True)\n",
    "    \n",
    "    # rename clusters from numbers to letters\n",
    "    df_tmp['Cluster'] = df_tmp['Cluster'].replace(cluster_to_letter)\n",
    "    if treatment==\"LO28\":\n",
    "        df_tmp['Cluster'] = df_tmp['Cluster'].replace(listeria_changes)\n",
    "\n",
    "    # add columns\n",
    "    df_tmp[\"Treatment\"] = treatment\n",
    "        \n",
    "    # set indices\n",
    "    df_tmp.set_index([\"Treatment\", \"Cluster\"], inplace=True)\n",
    "\n",
    "    tf_enr_dfs.append(df_tmp)\n",
    "\n",
    "tf_enr_df = pd.concat(tf_enr_dfs, axis=0)\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = tf_enr_df,\n",
    "                     sheet_name = \"TFBS_enrichments\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "print(tf_enr_df.shape)\n",
    "tf_enr_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "5fe36c7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save Excel file\n",
    "writer.save()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb234bd4",
   "metadata": {},
   "source": [
    "# CROP-seq small - TableS4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "4b667106",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = os.path.join('results', 'KO15')\n",
    "conditions = [\"untreated\", \"LO28_2h\", \"LO28_6h\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "16ced000",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create Excel writer\n",
    "writer = pd.ExcelWriter(os.path.join(results_path,'S04_CROPseq_small.xlsx'), engine='xlsxwriter')\n",
    "\n",
    "# Get the workbook and worksheet objects\n",
    "workbook = writer.book\n",
    "# define specific formatting\n",
    "index_format = workbook.add_format(xlsx_format_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "510a54bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add an empty sheet called LEGEND (fill with manually curated data later)\n",
    "pd.DataFrame().to_excel(writer, sheet_name=\"TableS4_legends\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "bb09d622",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(9153, 19)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Condition</th>\n",
       "      <th>KO_call</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>nCount_AB</th>\n",
       "      <th>nFeature_AB</th>\n",
       "      <th>nCount_gRNA</th>\n",
       "      <th>nFeature_gRNA</th>\n",
       "      <th>batch</th>\n",
       "      <th>percent.mt</th>\n",
       "      <th>guide_call</th>\n",
       "      <th>nCount_SCT</th>\n",
       "      <th>nFeature_SCT</th>\n",
       "      <th>S.Score</th>\n",
       "      <th>G2M.Score</th>\n",
       "      <th>Phase</th>\n",
       "      <th>mixscape_class</th>\n",
       "      <th>mixscape_class_p_ko</th>\n",
       "      <th>mixscape_class.global</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_AAACCTGAGAAGAAGC-1</th>\n",
       "      <td>LO28_2h</td>\n",
       "      <td>Stat1</td>\n",
       "      <td>9229</td>\n",
       "      <td>2626</td>\n",
       "      <td>277</td>\n",
       "      <td>8</td>\n",
       "      <td>186</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>0.736808</td>\n",
       "      <td>Stat1_3</td>\n",
       "      <td>8846</td>\n",
       "      <td>2624</td>\n",
       "      <td>0.030562</td>\n",
       "      <td>-0.317354</td>\n",
       "      <td>S</td>\n",
       "      <td>Stat1 KO</td>\n",
       "      <td>0.86101</td>\n",
       "      <td>KO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_AAACCTGAGACTGTAA-1</th>\n",
       "      <td>untreated</td>\n",
       "      <td>Irf8</td>\n",
       "      <td>10672</td>\n",
       "      <td>3572</td>\n",
       "      <td>389</td>\n",
       "      <td>10</td>\n",
       "      <td>1312</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>2.398801</td>\n",
       "      <td>Irf8_4</td>\n",
       "      <td>9274</td>\n",
       "      <td>3570</td>\n",
       "      <td>-0.093887</td>\n",
       "      <td>-0.181579</td>\n",
       "      <td>G1</td>\n",
       "      <td>Irf8 NP</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>NP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_AAACCTGAGGTGTTAA-1</th>\n",
       "      <td>untreated</td>\n",
       "      <td>Kdm1b</td>\n",
       "      <td>9867</td>\n",
       "      <td>3564</td>\n",
       "      <td>135</td>\n",
       "      <td>9</td>\n",
       "      <td>1069</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>1.530354</td>\n",
       "      <td>Kdm1b_3</td>\n",
       "      <td>9155</td>\n",
       "      <td>3561</td>\n",
       "      <td>-0.023333</td>\n",
       "      <td>0.346600</td>\n",
       "      <td>G2M</td>\n",
       "      <td>Kdm1b NP</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>NP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_AAACCTGCAGGCAGTA-1</th>\n",
       "      <td>LO28_6h</td>\n",
       "      <td>Non-targeting</td>\n",
       "      <td>8754</td>\n",
       "      <td>2577</td>\n",
       "      <td>232</td>\n",
       "      <td>9</td>\n",
       "      <td>43</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>1.028101</td>\n",
       "      <td>Non-targeting_6</td>\n",
       "      <td>8673</td>\n",
       "      <td>2575</td>\n",
       "      <td>-0.248403</td>\n",
       "      <td>-0.246516</td>\n",
       "      <td>G1</td>\n",
       "      <td>NT</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>NT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_AAACCTGCAGTCGTGC-1</th>\n",
       "      <td>LO28_6h</td>\n",
       "      <td>Hdac6</td>\n",
       "      <td>5200</td>\n",
       "      <td>1848</td>\n",
       "      <td>96</td>\n",
       "      <td>8</td>\n",
       "      <td>945</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>1.634615</td>\n",
       "      <td>Hdac6_2</td>\n",
       "      <td>7772</td>\n",
       "      <td>1868</td>\n",
       "      <td>-0.073559</td>\n",
       "      <td>-0.148819</td>\n",
       "      <td>G1</td>\n",
       "      <td>Hdac6 NP</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>NP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Condition        KO_call  nCount_RNA  nFeature_RNA  \\\n",
       "Cell barcode                                                               \n",
       "1_AAACCTGAGAAGAAGC-1    LO28_2h          Stat1        9229          2626   \n",
       "1_AAACCTGAGACTGTAA-1  untreated           Irf8       10672          3572   \n",
       "1_AAACCTGAGGTGTTAA-1  untreated          Kdm1b        9867          3564   \n",
       "1_AAACCTGCAGGCAGTA-1    LO28_6h  Non-targeting        8754          2577   \n",
       "1_AAACCTGCAGTCGTGC-1    LO28_6h          Hdac6        5200          1848   \n",
       "\n",
       "                      nCount_AB  nFeature_AB  nCount_gRNA  nFeature_gRNA  \\\n",
       "Cell barcode                                                               \n",
       "1_AAACCTGAGAAGAAGC-1        277            8          186             14   \n",
       "1_AAACCTGAGACTGTAA-1        389           10         1312             15   \n",
       "1_AAACCTGAGGTGTTAA-1        135            9         1069             11   \n",
       "1_AAACCTGCAGGCAGTA-1        232            9           43             16   \n",
       "1_AAACCTGCAGTCGTGC-1         96            8          945             10   \n",
       "\n",
       "                      batch  percent.mt       guide_call  nCount_SCT  \\\n",
       "Cell barcode                                                           \n",
       "1_AAACCTGAGAAGAAGC-1      1    0.736808          Stat1_3        8846   \n",
       "1_AAACCTGAGACTGTAA-1      1    2.398801           Irf8_4        9274   \n",
       "1_AAACCTGAGGTGTTAA-1      1    1.530354          Kdm1b_3        9155   \n",
       "1_AAACCTGCAGGCAGTA-1      1    1.028101  Non-targeting_6        8673   \n",
       "1_AAACCTGCAGTCGTGC-1      1    1.634615          Hdac6_2        7772   \n",
       "\n",
       "                      nFeature_SCT   S.Score  G2M.Score Phase mixscape_class  \\\n",
       "Cell barcode                                                                   \n",
       "1_AAACCTGAGAAGAAGC-1          2624  0.030562  -0.317354     S       Stat1 KO   \n",
       "1_AAACCTGAGACTGTAA-1          3570 -0.093887  -0.181579    G1        Irf8 NP   \n",
       "1_AAACCTGAGGTGTTAA-1          3561 -0.023333   0.346600   G2M       Kdm1b NP   \n",
       "1_AAACCTGCAGGCAGTA-1          2575 -0.248403  -0.246516    G1             NT   \n",
       "1_AAACCTGCAGTCGTGC-1          1868 -0.073559  -0.148819    G1       Hdac6 NP   \n",
       "\n",
       "                      mixscape_class_p_ko mixscape_class.global  \n",
       "Cell barcode                                                     \n",
       "1_AAACCTGAGAAGAAGC-1              0.86101                    KO  \n",
       "1_AAACCTGAGACTGTAA-1              0.00000                    NP  \n",
       "1_AAACCTGAGGTGTTAA-1              0.00000                    NP  \n",
       "1_AAACCTGCAGGCAGTA-1              0.00000                    NT  \n",
       "1_AAACCTGCAGTCGTGC-1              0.00000                    NP  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# metadata\n",
    "metadata = pd.read_csv(os.path.join(data_path, 'KO15_NORMALIZED_metadata.csv'), index_col=0)\n",
    "# load mixscape classifications\n",
    "mixscape_dfs = []\n",
    "for cond in conditions:\n",
    "    cond_tmp = pd.read_csv(os.path.join(data_path, 'mixscape', cond, 'KO15_mixscape_annotations.csv'), index_col=0)\n",
    "    mixscape_dfs.append(cond_tmp)\n",
    "\n",
    "# concatenate dfs\n",
    "mixscape_df = pd.concat(mixscape_dfs, axis=0)\n",
    "metadata = pd.concat([metadata, mixscape_df], axis=1)\n",
    "\n",
    "# set condition column\n",
    "metadata['Condition'] = metadata['hto_demux'].str.replace(\"HTO-RAW-SZ_\", \"\")\n",
    "\n",
    "# remove leading m in KO target gene names\n",
    "metadata['KO_call'] = metadata['KO_call'].str[1:]\n",
    "metadata['guide_call'] = metadata['guide_call'].str[1:]\n",
    "\n",
    "# Drop columns\n",
    "metadata = metadata.drop(['old.ident', 'NT_Spi1', \"orig.ident\", \"pass_QC\", \"hto_demux\", \"nCount_HTO\", \"nFeature_HTO\"], axis=1)\n",
    "\n",
    "# add index name\n",
    "metadata.index.names = [\"Cell barcode\"]\n",
    "\n",
    "# set indices\n",
    "metadata.reset_index(inplace=True)\n",
    "metadata.set_index([\"Cell barcode\", \"Condition\", \"KO_call\"], inplace=True)\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = metadata.copy(deep=True),\n",
    "                     sheet_name = \"single_cell_annotations\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "\n",
    "# reset indices\n",
    "metadata.reset_index(inplace=True)\n",
    "metadata.set_index([\"Cell barcode\"], inplace=True)\n",
    "\n",
    "print(metadata.shape)\n",
    "metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "e5c74d0e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(6663, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>P-value</th>\n",
       "      <th>Avg log2FC</th>\n",
       "      <th>Percentage expression 1</th>\n",
       "      <th>Percentage expression 2</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th>Time point</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ccl4</th>\n",
       "      <th>Listeria 2h</th>\n",
       "      <td>6.375826e-92</td>\n",
       "      <td>4.746794</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.859</td>\n",
       "      <td>9.045384e-88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tnf</th>\n",
       "      <th>Listeria 2h</th>\n",
       "      <td>1.676377e-90</td>\n",
       "      <td>4.563107</td>\n",
       "      <td>0.997</td>\n",
       "      <td>0.718</td>\n",
       "      <td>2.378276e-86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Junb</th>\n",
       "      <th>Listeria 2h</th>\n",
       "      <td>4.342102e-90</td>\n",
       "      <td>3.075322</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.725</td>\n",
       "      <td>6.160140e-86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ccl3</th>\n",
       "      <th>Listeria 2h</th>\n",
       "      <td>1.005605e-89</td>\n",
       "      <td>3.470721</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.912</td>\n",
       "      <td>1.426651e-85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cxcl2</th>\n",
       "      <th>Listeria 2h</th>\n",
       "      <td>3.237057e-83</td>\n",
       "      <td>5.308387</td>\n",
       "      <td>0.984</td>\n",
       "      <td>0.656</td>\n",
       "      <td>4.592413e-79</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        P-value  Avg log2FC  Percentage expression 1  \\\n",
       "Gene  Time point                                                       \n",
       "Ccl4  Listeria 2h  6.375826e-92    4.746794                    1.000   \n",
       "Tnf   Listeria 2h  1.676377e-90    4.563107                    0.997   \n",
       "Junb  Listeria 2h  4.342102e-90    3.075322                    1.000   \n",
       "Ccl3  Listeria 2h  1.005605e-89    3.470721                    1.000   \n",
       "Cxcl2 Listeria 2h  3.237057e-83    5.308387                    0.984   \n",
       "\n",
       "                   Percentage expression 2  Adjusted P-value  \n",
       "Gene  Time point                                              \n",
       "Ccl4  Listeria 2h                    0.859      9.045384e-88  \n",
       "Tnf   Listeria 2h                    0.718      2.378276e-86  \n",
       "Junb  Listeria 2h                    0.725      6.160140e-86  \n",
       "Ccl3  Listeria 2h                    0.912      1.426651e-85  \n",
       "Cxcl2 Listeria 2h                    0.656      4.592413e-79  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# DEA - NT: 2h/6h vs UT results\n",
    "dea_dfs = []\n",
    "\n",
    "for cond in conditions:\n",
    "    \n",
    "    if cond==\"untreated\":\n",
    "        continue\n",
    "    \n",
    "    # load dea results\n",
    "    df_tmp = pd.read_csv(os.path.join(data_path, 'DEA_KO', 'KO15_DEA_RNA_mNon-targeting_{}_untreated.csv'.format(cond)), index_col=0)\n",
    "\n",
    "    # set index name\n",
    "    df_tmp.index.names = [\"Gene\"]\n",
    "    \n",
    "    # add column\n",
    "    df_tmp[\"Time point\"]=cond\n",
    "    \n",
    "    # set indices\n",
    "    df_tmp.reset_index(inplace=True)\n",
    "    df_tmp.set_index([\"Gene\", \"Time point\"], inplace=True)\n",
    "    \n",
    "    dea_dfs.append(df_tmp)\n",
    "\n",
    "# dea_df = pd.concat(dea_dfs, axis=1)\n",
    "dea_df = pd.concat(dea_dfs, axis=0)\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = dea_df,\n",
    "                     sheet_name = \"DEA_in_WT\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "print(dea_df.shape)\n",
    "dea_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "4063b2ce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(8086, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>P-value</th>\n",
       "      <th>Avg log2FC</th>\n",
       "      <th>Percentage expression 1</th>\n",
       "      <th>Percentage expression 2</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th>Time point</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ptgds</th>\n",
       "      <th>Untreated</th>\n",
       "      <td>1.346231e-46</td>\n",
       "      <td>2.130750</td>\n",
       "      <td>0.730</td>\n",
       "      <td>0.050</td>\n",
       "      <td>1.909898e-42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cst3</th>\n",
       "      <th>Untreated</th>\n",
       "      <td>2.992243e-28</td>\n",
       "      <td>-1.410878</td>\n",
       "      <td>0.992</td>\n",
       "      <td>1.000</td>\n",
       "      <td>4.245095e-24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Csf1r</th>\n",
       "      <th>Untreated</th>\n",
       "      <td>8.989721e-25</td>\n",
       "      <td>-1.197709</td>\n",
       "      <td>0.230</td>\n",
       "      <td>0.821</td>\n",
       "      <td>1.275372e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tyrobp</th>\n",
       "      <th>Untreated</th>\n",
       "      <td>1.816720e-24</td>\n",
       "      <td>-0.955098</td>\n",
       "      <td>0.951</td>\n",
       "      <td>0.996</td>\n",
       "      <td>2.577380e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Coro1a</th>\n",
       "      <th>Untreated</th>\n",
       "      <td>2.867002e-24</td>\n",
       "      <td>-0.923344</td>\n",
       "      <td>0.910</td>\n",
       "      <td>0.996</td>\n",
       "      <td>4.067416e-20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        P-value  Avg log2FC  Percentage expression 1  \\\n",
       "Gene   Time point                                                      \n",
       "Ptgds  Untreated   1.346231e-46    2.130750                    0.730   \n",
       "Cst3   Untreated   2.992243e-28   -1.410878                    0.992   \n",
       "Csf1r  Untreated   8.989721e-25   -1.197709                    0.230   \n",
       "Tyrobp Untreated   1.816720e-24   -0.955098                    0.951   \n",
       "Coro1a Untreated   2.867002e-24   -0.923344                    0.910   \n",
       "\n",
       "                   Percentage expression 2  Adjusted P-value  \n",
       "Gene   Time point                                             \n",
       "Ptgds  Untreated                     0.050      1.909898e-42  \n",
       "Cst3   Untreated                     1.000      4.245095e-24  \n",
       "Csf1r  Untreated                     0.821      1.275372e-20  \n",
       "Tyrobp Untreated                     0.996      2.577380e-20  \n",
       "Coro1a Untreated                     0.996      4.067416e-20  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# DEA - UT/2h/6h: Spi1 vs NT results\n",
    "dea_dfs = []\n",
    "\n",
    "# KO15_DEA_RNA_LO28_2h_mSpi1.csv\n",
    "\n",
    "for cond in conditions:\n",
    "    \n",
    "    # load dea results\n",
    "    df_tmp = pd.read_csv(os.path.join(data_path, 'DEA_condition', 'KO15_DEA_RNA_{}_mSpi1.csv'.format(cond)), index_col=0)\n",
    "\n",
    "    # set index name\n",
    "    df_tmp.index.names = [\"Gene\"]\n",
    "    \n",
    "    # add column\n",
    "    df_tmp[\"Time point\"]=cond\n",
    "    \n",
    "    # set indices\n",
    "    df_tmp.reset_index(inplace=True)\n",
    "    df_tmp.set_index([\"Gene\", \"Time point\"], inplace=True)\n",
    "    \n",
    "    dea_dfs.append(df_tmp)\n",
    "\n",
    "# dea_df = pd.concat(dea_dfs, axis=1)\n",
    "dea_df = pd.concat(dea_dfs, axis=0)\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = dea_df,\n",
    "                     sheet_name = \"DEA_Spi1\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "print(dea_df.shape)\n",
    "dea_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "578a5cc3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(163, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Database</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Term</th>\n",
       "      <th>Control</th>\n",
       "      <th>Perturbation</th>\n",
       "      <th>Direction</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Antigen processing and presentation</th>\n",
       "      <th>Untreated</th>\n",
       "      <th>Spi1 KO</th>\n",
       "      <th>up</th>\n",
       "      <td>3/36</td>\n",
       "      <td>7.618237e-04</td>\n",
       "      <td>1.318541e-02</td>\n",
       "      <td>17.341912</td>\n",
       "      <td>HSP90AB1;CTSL;HSPA8</td>\n",
       "      <td>BioPlanet_2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Arachidonic acid metabolism</th>\n",
       "      <th>Untreated</th>\n",
       "      <th>Spi1 KO</th>\n",
       "      <th>up</th>\n",
       "      <td>2/25</td>\n",
       "      <td>6.877169e-03</td>\n",
       "      <td>8.144015e-02</td>\n",
       "      <td>16.419710</td>\n",
       "      <td>GPX4;PTGDS</td>\n",
       "      <td>BioPlanet_2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cytoplasmic ribosomal proteins</th>\n",
       "      <th>Untreated</th>\n",
       "      <th>Spi1 KO</th>\n",
       "      <th>up</th>\n",
       "      <td>15/98</td>\n",
       "      <td>8.518237e-19</td>\n",
       "      <td>9.583016e-17</td>\n",
       "      <td>38.508564</td>\n",
       "      <td>RPS24;RPL28;RPL23;RPL32;RPS10;RPL36;RPL21;RPS1...</td>\n",
       "      <td>BioPlanet_2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inhibitor of DNA binding (ID) signaling pathway</th>\n",
       "      <th>Untreated</th>\n",
       "      <th>Spi1 KO</th>\n",
       "      <th>up</th>\n",
       "      <td>1/12</td>\n",
       "      <td>5.845127e-02</td>\n",
       "      <td>2.378717e-01</td>\n",
       "      <td>16.875000</td>\n",
       "      <td>ID3</td>\n",
       "      <td>BioPlanet_2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Innate immune system</th>\n",
       "      <th>Untreated</th>\n",
       "      <th>Spi1 KO</th>\n",
       "      <th>up</th>\n",
       "      <td>2/185</td>\n",
       "      <td>2.367215e-01</td>\n",
       "      <td>3.301595e-01</td>\n",
       "      <td>2.193811</td>\n",
       "      <td>HSP90AB1;CTSL</td>\n",
       "      <td>BioPlanet_2019</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                 Overlap  \\\n",
       "Term                                            Control   Perturbation Direction           \n",
       "Antigen processing and presentation             Untreated Spi1 KO      up           3/36   \n",
       "Arachidonic acid metabolism                     Untreated Spi1 KO      up           2/25   \n",
       "Cytoplasmic ribosomal proteins                  Untreated Spi1 KO      up          15/98   \n",
       "Inhibitor of DNA binding (ID) signaling pathway Untreated Spi1 KO      up           1/12   \n",
       "Innate immune system                            Untreated Spi1 KO      up          2/185   \n",
       "\n",
       "                                                                                       P-value  \\\n",
       "Term                                            Control   Perturbation Direction                 \n",
       "Antigen processing and presentation             Untreated Spi1 KO      up         7.618237e-04   \n",
       "Arachidonic acid metabolism                     Untreated Spi1 KO      up         6.877169e-03   \n",
       "Cytoplasmic ribosomal proteins                  Untreated Spi1 KO      up         8.518237e-19   \n",
       "Inhibitor of DNA binding (ID) signaling pathway Untreated Spi1 KO      up         5.845127e-02   \n",
       "Innate immune system                            Untreated Spi1 KO      up         2.367215e-01   \n",
       "\n",
       "                                                                                  Adjusted P-value  \\\n",
       "Term                                            Control   Perturbation Direction                     \n",
       "Antigen processing and presentation             Untreated Spi1 KO      up             1.318541e-02   \n",
       "Arachidonic acid metabolism                     Untreated Spi1 KO      up             8.144015e-02   \n",
       "Cytoplasmic ribosomal proteins                  Untreated Spi1 KO      up             9.583016e-17   \n",
       "Inhibitor of DNA binding (ID) signaling pathway Untreated Spi1 KO      up             2.378717e-01   \n",
       "Innate immune system                            Untreated Spi1 KO      up             3.301595e-01   \n",
       "\n",
       "                                                                                  Odds Ratio  \\\n",
       "Term                                            Control   Perturbation Direction               \n",
       "Antigen processing and presentation             Untreated Spi1 KO      up          17.341912   \n",
       "Arachidonic acid metabolism                     Untreated Spi1 KO      up          16.419710   \n",
       "Cytoplasmic ribosomal proteins                  Untreated Spi1 KO      up          38.508564   \n",
       "Inhibitor of DNA binding (ID) signaling pathway Untreated Spi1 KO      up          16.875000   \n",
       "Innate immune system                            Untreated Spi1 KO      up           2.193811   \n",
       "\n",
       "                                                                                                                              Genes  \\\n",
       "Term                                            Control   Perturbation Direction                                                      \n",
       "Antigen processing and presentation             Untreated Spi1 KO      up                                       HSP90AB1;CTSL;HSPA8   \n",
       "Arachidonic acid metabolism                     Untreated Spi1 KO      up                                                GPX4;PTGDS   \n",
       "Cytoplasmic ribosomal proteins                  Untreated Spi1 KO      up         RPS24;RPL28;RPL23;RPL32;RPS10;RPL36;RPL21;RPS1...   \n",
       "Inhibitor of DNA binding (ID) signaling pathway Untreated Spi1 KO      up                                                       ID3   \n",
       "Innate immune system                            Untreated Spi1 KO      up                                             HSP90AB1;CTSL   \n",
       "\n",
       "                                                                                        Database  \n",
       "Term                                            Control   Perturbation Direction                  \n",
       "Antigen processing and presentation             Untreated Spi1 KO      up         BioPlanet_2019  \n",
       "Arachidonic acid metabolism                     Untreated Spi1 KO      up         BioPlanet_2019  \n",
       "Cytoplasmic ribosomal proteins                  Untreated Spi1 KO      up         BioPlanet_2019  \n",
       "Inhibitor of DNA binding (ID) signaling pathway Untreated Spi1 KO      up         BioPlanet_2019  \n",
       "Innate immune system                            Untreated Spi1 KO      up         BioPlanet_2019  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Enrichment results Spi1\n",
    "spi1_enr_dfs = []\n",
    "\n",
    "db = spi1_enrichment_terms.loc[\"Spi1_plot\",\"database\"]\n",
    "terms = spi1_enrichment_terms.loc[\"Spi1_plot\",\"terms\"].split(\" | \")\n",
    "\n",
    "for direction in ['up','down']:\n",
    "    \n",
    "    # genetic perturbation effect in untreated and Listeria 6h\n",
    "    KO = \"mSpi1\"\n",
    "    for cond in conditions:\n",
    "        if cond==\"LO28_2h\":\n",
    "            continue\n",
    "\n",
    "        # load enr results\n",
    "        df_tmp = pd.read_csv(os.path.join(data_path, 'enrichment_analysis', 'DEA_condition', \"{}_{}_{}\".format(cond, KO, direction), db, 'Enrichr_{}_{}_{}_{}.csv'.format(db, cond, KO, direction)))\n",
    "\n",
    "        # subset by list of terms\n",
    "        df_tmp = df_tmp[df_tmp['Term'].isin(terms)]\n",
    "        \n",
    "        # Drop columns\n",
    "        df_tmp = df_tmp.drop(['Unnamed: 0', 'Gene_set'], axis=1)\n",
    "    #     df_tmp = df_tmp.drop(['Unnamed: 0', 'Gene_set', 'Genes'], axis=1)\n",
    "\n",
    "        # Rename columns \n",
    "    #     df_tmp.columns = [\"cluster_{}_{}\".format(cluster,col) for col in df_tmp.columns]\n",
    "\n",
    "        # add cluster column\n",
    "        df_tmp[\"Control\"] = cond\n",
    "        df_tmp[\"Perturbation\"] = KO+\" KO\"\n",
    "        df_tmp[\"Direction\"] = direction\n",
    "        \n",
    "        # set indices\n",
    "        df_tmp.set_index([\"Term\", \"Control\", \"Perturbation\", \"Direction\"], inplace=True)\n",
    "\n",
    "        spi1_enr_dfs.append(df_tmp)\n",
    "\n",
    "    # treatment effect in WT and Spi1 KO\n",
    "    cond = \"LO28_6h\"\n",
    "    for KO in ['mSpi1','mNon-targeting']:\n",
    "\n",
    "        # load enr results\n",
    "        df_tmp = pd.read_csv(os.path.join(data_path, 'enrichment_analysis', 'DEA_KO', \"{}_{}_untreated_{}\".format(KO, cond, direction), db, 'Enrichr_{}_{}_{}_untreated_{}.csv'.format(db, KO, cond, direction)))\n",
    "\n",
    "        # subset by list of terms\n",
    "        df_tmp = df_tmp[df_tmp['Term'].isin(terms)]\n",
    "        \n",
    "        # Drop columns\n",
    "        df_tmp = df_tmp.drop(['Unnamed: 0', 'Gene_set'], axis=1)\n",
    "    #     df_tmp = df_tmp.drop(['Unnamed: 0', 'Gene_set', 'Genes'], axis=1)\n",
    "\n",
    "        # Rename columns \n",
    "    #     df_tmp.columns = [\"cluster_{}_{}\".format(cluster,col) for col in df_tmp.columns]\n",
    "\n",
    "        # add cluster column\n",
    "        df_tmp[\"Control\"] = KO+\" KO\"\n",
    "        df_tmp[\"Perturbation\"] = cond\n",
    "        df_tmp[\"Direction\"] = direction\n",
    "        \n",
    "        # set indices\n",
    "        df_tmp.set_index([\"Term\", \"Control\", \"Perturbation\", \"Direction\"], inplace=True)\n",
    "\n",
    "        spi1_enr_dfs.append(df_tmp)\n",
    "        \n",
    "        \n",
    "        \n",
    "spi1_enr_df = pd.concat(spi1_enr_dfs, axis=0)\n",
    "\n",
    "# add database column\n",
    "spi1_enr_df[\"Database\"] = db\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = spi1_enr_df,\n",
    "                     sheet_name = \"Spi1_Listeria_enrichments\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "print(spi1_enr_df.shape)\n",
    "spi1_enr_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "b5e8484d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1115, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>P-value</th>\n",
       "      <th>Avg log2FC</th>\n",
       "      <th>Percentage expression 1</th>\n",
       "      <th>Percentage expression 2</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th>Time point</th>\n",
       "      <th>Assay</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Csf1r</th>\n",
       "      <th>Untreated</th>\n",
       "      <th>RNA</th>\n",
       "      <td>2.355798e-23</td>\n",
       "      <td>-1.029912</td>\n",
       "      <td>0.383</td>\n",
       "      <td>0.821</td>\n",
       "      <td>3.342170e-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ei24</th>\n",
       "      <th>Untreated</th>\n",
       "      <th>RNA</th>\n",
       "      <td>5.717037e-05</td>\n",
       "      <td>0.301344</td>\n",
       "      <td>0.506</td>\n",
       "      <td>0.317</td>\n",
       "      <td>8.110760e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Nudt21</th>\n",
       "      <th>Untreated</th>\n",
       "      <th>RNA</th>\n",
       "      <td>1.745362e-04</td>\n",
       "      <td>0.288053</td>\n",
       "      <td>0.851</td>\n",
       "      <td>0.718</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lta4h</th>\n",
       "      <th>Untreated</th>\n",
       "      <th>RNA</th>\n",
       "      <td>1.947871e-04</td>\n",
       "      <td>0.256561</td>\n",
       "      <td>0.487</td>\n",
       "      <td>0.305</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mkrn1</th>\n",
       "      <th>Untreated</th>\n",
       "      <th>RNA</th>\n",
       "      <td>3.136294e-04</td>\n",
       "      <td>0.328329</td>\n",
       "      <td>0.708</td>\n",
       "      <td>0.565</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              P-value  Avg log2FC  Percentage expression 1  \\\n",
       "Gene   Time point Assay                                                      \n",
       "Csf1r  Untreated  RNA    2.355798e-23   -1.029912                    0.383   \n",
       "Ei24   Untreated  RNA    5.717037e-05    0.301344                    0.506   \n",
       "Nudt21 Untreated  RNA    1.745362e-04    0.288053                    0.851   \n",
       "Lta4h  Untreated  RNA    1.947871e-04    0.256561                    0.487   \n",
       "Mkrn1  Untreated  RNA    3.136294e-04    0.328329                    0.708   \n",
       "\n",
       "                         Percentage expression 2  Adjusted P-value  \n",
       "Gene   Time point Assay                                             \n",
       "Csf1r  Untreated  RNA                      0.821      3.342170e-19  \n",
       "Ei24   Untreated  RNA                      0.317      8.110760e-01  \n",
       "Nudt21 Untreated  RNA                      0.718      1.000000e+00  \n",
       "Lta4h  Untreated  RNA                      0.305      1.000000e+00  \n",
       "Mkrn1  Untreated  RNA                      0.565      1.000000e+00  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# DEA - UT: Csf1r vs NT DEA results\n",
    "dea_dfs = []\n",
    "\n",
    "cond = \"untreated\"\n",
    "\n",
    "for assay in ['RNA', 'AB']:\n",
    "    \n",
    "    # load dea results\n",
    "    df_tmp = pd.read_csv(os.path.join(data_path, 'DEA_condition', 'KO15_DEA_{}_{}_mCsf1r.csv'.format(assay, cond)), index_col=0)\n",
    "\n",
    "    # set index name\n",
    "    df_tmp.index.names = [\"Gene\"]\n",
    "    \n",
    "    # add column\n",
    "    df_tmp[\"Condition\"]=cond\n",
    "    df_tmp[\"Assay\"]=assay\n",
    "    \n",
    "    # set indices\n",
    "    df_tmp.reset_index(inplace=True)\n",
    "    df_tmp.set_index([\"Gene\", \"Condition\", \"Assay\"], inplace=True)\n",
    "    \n",
    "    dea_dfs.append(df_tmp)\n",
    "\n",
    "# dea_df = pd.concat(dea_dfs, axis=1)\n",
    "dea_df = pd.concat(dea_dfs, axis=0)\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = dea_df,\n",
    "                     sheet_name = \"DEA_Csf1r_in_Untreated\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "print(dea_df.shape)\n",
    "dea_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "5ca3611f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(8987, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Database</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Term</th>\n",
       "      <th>Time point</th>\n",
       "      <th>Direction</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>'de novo' posttranslational protein folding (GO:0051084)</th>\n",
       "      <th>Listeria 2h</th>\n",
       "      <th>up</th>\n",
       "      <td>3/28</td>\n",
       "      <td>0.008184</td>\n",
       "      <td>0.037269</td>\n",
       "      <td>7.258544</td>\n",
       "      <td>HSPA5;HSPA8;DNAJB1</td>\n",
       "      <td>GO_Biological_Process_2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3'-UTR-mediated mRNA destabilization (GO:0061158)</th>\n",
       "      <th>Listeria 2h</th>\n",
       "      <th>up</th>\n",
       "      <td>4/14</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000828</td>\n",
       "      <td>19.468407</td>\n",
       "      <td>ZFP36L1;ZC3H12A;RC3H1;ZFP36</td>\n",
       "      <td>GO_Biological_Process_2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3'-UTR-mediated mRNA stabilization (GO:0070935)</th>\n",
       "      <th>Listeria 2h</th>\n",
       "      <th>up</th>\n",
       "      <td>2/11</td>\n",
       "      <td>0.011185</td>\n",
       "      <td>0.045893</td>\n",
       "      <td>12.273593</td>\n",
       "      <td>MAPKAPK2;ZFP36</td>\n",
       "      <td>GO_Biological_Process_2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATF6-mediated unfolded protein response (GO:0036500)</th>\n",
       "      <th>Listeria 2h</th>\n",
       "      <th>up</th>\n",
       "      <td>1/9</td>\n",
       "      <td>0.126758</td>\n",
       "      <td>0.210114</td>\n",
       "      <td>7.466035</td>\n",
       "      <td>HSPA5</td>\n",
       "      <td>GO_Biological_Process_2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATP-dependent chromatin remodeling (GO:0043044)</th>\n",
       "      <th>Listeria 2h</th>\n",
       "      <th>up</th>\n",
       "      <td>1/39</td>\n",
       "      <td>0.444551</td>\n",
       "      <td>0.485050</td>\n",
       "      <td>1.719285</td>\n",
       "      <td>ACTB</td>\n",
       "      <td>GO_Biological_Process_2021</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                         Overlap  \\\n",
       "Term                                               Time point  Direction           \n",
       "'de novo' posttranslational protein folding (GO... Listeria 2h up           3/28   \n",
       "3'-UTR-mediated mRNA destabilization (GO:0061158)  Listeria 2h up           4/14   \n",
       "3'-UTR-mediated mRNA stabilization (GO:0070935)    Listeria 2h up           2/11   \n",
       "ATF6-mediated unfolded protein response (GO:003... Listeria 2h up            1/9   \n",
       "ATP-dependent chromatin remodeling (GO:0043044)    Listeria 2h up           1/39   \n",
       "\n",
       "                                                                           P-value  \\\n",
       "Term                                               Time point  Direction             \n",
       "'de novo' posttranslational protein folding (GO... Listeria 2h up         0.008184   \n",
       "3'-UTR-mediated mRNA destabilization (GO:0061158)  Listeria 2h up         0.000043   \n",
       "3'-UTR-mediated mRNA stabilization (GO:0070935)    Listeria 2h up         0.011185   \n",
       "ATF6-mediated unfolded protein response (GO:003... Listeria 2h up         0.126758   \n",
       "ATP-dependent chromatin remodeling (GO:0043044)    Listeria 2h up         0.444551   \n",
       "\n",
       "                                                                          Adjusted P-value  \\\n",
       "Term                                               Time point  Direction                     \n",
       "'de novo' posttranslational protein folding (GO... Listeria 2h up                 0.037269   \n",
       "3'-UTR-mediated mRNA destabilization (GO:0061158)  Listeria 2h up                 0.000828   \n",
       "3'-UTR-mediated mRNA stabilization (GO:0070935)    Listeria 2h up                 0.045893   \n",
       "ATF6-mediated unfolded protein response (GO:003... Listeria 2h up                 0.210114   \n",
       "ATP-dependent chromatin remodeling (GO:0043044)    Listeria 2h up                 0.485050   \n",
       "\n",
       "                                                                          Odds Ratio  \\\n",
       "Term                                               Time point  Direction               \n",
       "'de novo' posttranslational protein folding (GO... Listeria 2h up           7.258544   \n",
       "3'-UTR-mediated mRNA destabilization (GO:0061158)  Listeria 2h up          19.468407   \n",
       "3'-UTR-mediated mRNA stabilization (GO:0070935)    Listeria 2h up          12.273593   \n",
       "ATF6-mediated unfolded protein response (GO:003... Listeria 2h up           7.466035   \n",
       "ATP-dependent chromatin remodeling (GO:0043044)    Listeria 2h up           1.719285   \n",
       "\n",
       "                                                                                                Genes  \\\n",
       "Term                                               Time point  Direction                                \n",
       "'de novo' posttranslational protein folding (GO... Listeria 2h up                  HSPA5;HSPA8;DNAJB1   \n",
       "3'-UTR-mediated mRNA destabilization (GO:0061158)  Listeria 2h up         ZFP36L1;ZC3H12A;RC3H1;ZFP36   \n",
       "3'-UTR-mediated mRNA stabilization (GO:0070935)    Listeria 2h up                      MAPKAPK2;ZFP36   \n",
       "ATF6-mediated unfolded protein response (GO:003... Listeria 2h up                               HSPA5   \n",
       "ATP-dependent chromatin remodeling (GO:0043044)    Listeria 2h up                                ACTB   \n",
       "\n",
       "                                                                                            Database  \n",
       "Term                                               Time point  Direction                              \n",
       "'de novo' posttranslational protein folding (GO... Listeria 2h up         GO_Biological_Process_2021  \n",
       "3'-UTR-mediated mRNA destabilization (GO:0061158)  Listeria 2h up         GO_Biological_Process_2021  \n",
       "3'-UTR-mediated mRNA stabilization (GO:0070935)    Listeria 2h up         GO_Biological_Process_2021  \n",
       "ATF6-mediated unfolded protein response (GO:003... Listeria 2h up         GO_Biological_Process_2021  \n",
       "ATP-dependent chromatin remodeling (GO:0043044)    Listeria 2h up         GO_Biological_Process_2021  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Enrichment results of NT/WT with treatment\n",
    "NT_enr_dfs = []\n",
    "\n",
    "db = \"GO_Biological_Process_2021\"\n",
    "KO = \"mNon-targeting\"\n",
    "\n",
    "for direction in ['up','down']:\n",
    "    \n",
    "    # treatment effect in NT/WT\n",
    "    for cond in conditions:\n",
    "        if cond==\"untreated\":\n",
    "            continue\n",
    "\n",
    "        # load enr results\n",
    "        df_tmp = pd.read_csv(os.path.join(data_path, 'enrichment_analysis', 'DEA_KO', \"{}_{}_untreated_{}\".format(KO, cond, direction), db, 'Enrichr_{}_{}_{}_untreated_{}.csv'.format(db, KO, cond, direction)))\n",
    "        \n",
    "        # Drop columns\n",
    "        df_tmp = df_tmp.drop(['Unnamed: 0', 'Gene_set'], axis=1)\n",
    "    #     df_tmp = df_tmp.drop(['Unnamed: 0', 'Gene_set', 'Genes'], axis=1)\n",
    "\n",
    "        # Rename columns \n",
    "    #     df_tmp.columns = [\"cluster_{}_{}\".format(cluster,col) for col in df_tmp.columns]\n",
    "\n",
    "        # add cluster column\n",
    "        df_tmp[\"Time point\"] = cond\n",
    "        df_tmp[\"Direction\"] = direction\n",
    "        \n",
    "        # set indices\n",
    "        df_tmp.set_index([\"Term\", \"Time point\", \"Direction\"], inplace=True)\n",
    "\n",
    "        NT_enr_dfs.append(df_tmp)\n",
    "\n",
    "NT_enr_df = pd.concat(NT_enr_dfs, axis=0)\n",
    "\n",
    "# add database column\n",
    "NT_enr_df[\"Database\"] = db\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = NT_enr_df,\n",
    "                     sheet_name = \"WT_Listeria_enrichments\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "print(NT_enr_df.shape)\n",
    "NT_enr_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "cc4461e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(44298, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>P-value</th>\n",
       "      <th>Avg log2FC</th>\n",
       "      <th>Percentage expression 1</th>\n",
       "      <th>Percentage expression 2</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th>Time point</th>\n",
       "      <th>KO</th>\n",
       "      <th>Assay</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ifit3</th>\n",
       "      <th>Untreated</th>\n",
       "      <th>Stat1</th>\n",
       "      <th>RNA</th>\n",
       "      <td>1.022991e-29</td>\n",
       "      <td>1.348961</td>\n",
       "      <td>0.736</td>\n",
       "      <td>0.160</td>\n",
       "      <td>1.370297e-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cxcl10</th>\n",
       "      <th>Untreated</th>\n",
       "      <th>Stat1</th>\n",
       "      <th>RNA</th>\n",
       "      <td>1.977560e-27</td>\n",
       "      <td>1.818780</td>\n",
       "      <td>0.925</td>\n",
       "      <td>0.527</td>\n",
       "      <td>2.648942e-23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rsad2</th>\n",
       "      <th>Untreated</th>\n",
       "      <th>Stat1</th>\n",
       "      <th>RNA</th>\n",
       "      <td>8.796826e-24</td>\n",
       "      <td>1.345955</td>\n",
       "      <td>0.849</td>\n",
       "      <td>0.382</td>\n",
       "      <td>1.178335e-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ifit1</th>\n",
       "      <th>Untreated</th>\n",
       "      <th>Stat1</th>\n",
       "      <th>RNA</th>\n",
       "      <td>6.827681e-23</td>\n",
       "      <td>1.197628</td>\n",
       "      <td>0.717</td>\n",
       "      <td>0.210</td>\n",
       "      <td>9.145678e-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Phf11b</th>\n",
       "      <th>Untreated</th>\n",
       "      <th>Stat1</th>\n",
       "      <th>RNA</th>\n",
       "      <td>1.169138e-20</td>\n",
       "      <td>1.070399</td>\n",
       "      <td>0.802</td>\n",
       "      <td>0.374</td>\n",
       "      <td>1.566060e-16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    P-value  Avg log2FC  \\\n",
       "Gene   Time point KO    Assay                             \n",
       "Ifit3  Untreated  Stat1 RNA    1.022991e-29    1.348961   \n",
       "Cxcl10 Untreated  Stat1 RNA    1.977560e-27    1.818780   \n",
       "Rsad2  Untreated  Stat1 RNA    8.796826e-24    1.345955   \n",
       "Ifit1  Untreated  Stat1 RNA    6.827681e-23    1.197628   \n",
       "Phf11b Untreated  Stat1 RNA    1.169138e-20    1.070399   \n",
       "\n",
       "                               Percentage expression 1  \\\n",
       "Gene   Time point KO    Assay                            \n",
       "Ifit3  Untreated  Stat1 RNA                      0.736   \n",
       "Cxcl10 Untreated  Stat1 RNA                      0.925   \n",
       "Rsad2  Untreated  Stat1 RNA                      0.849   \n",
       "Ifit1  Untreated  Stat1 RNA                      0.717   \n",
       "Phf11b Untreated  Stat1 RNA                      0.802   \n",
       "\n",
       "                               Percentage expression 2  Adjusted P-value  \n",
       "Gene   Time point KO    Assay                                             \n",
       "Ifit3  Untreated  Stat1 RNA                      0.160      1.370297e-25  \n",
       "Cxcl10 Untreated  Stat1 RNA                      0.527      2.648942e-23  \n",
       "Rsad2  Untreated  Stat1 RNA                      0.382      1.178335e-19  \n",
       "Ifit1  Untreated  Stat1 RNA                      0.210      9.145678e-19  \n",
       "Phf11b Untreated  Stat1 RNA                      0.374      1.566060e-16  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# DEA post-mixscape KO effects within conditions\n",
    "dea_dfs = []\n",
    "\n",
    "for assay in ['RNA', 'AB']:\n",
    "    for cond in conditions:\n",
    "        KOs = list(metadata.loc[(metadata[\"mixscape_class.global\"]==\"KO\") & (metadata[\"Condition\"]==cond), \"KO_call\"].unique())\n",
    "        for KO in KOs:\n",
    "            \n",
    "            # load dea results\n",
    "            df_tmp = pd.read_csv(os.path.join(data_path, 'DEA_condition_mixscape', 'KO15_DEA_{}_{}_m{}.csv'.format(assay, cond, KO)), index_col=0)\n",
    "\n",
    "            # set index name\n",
    "            df_tmp.index.names = [\"Gene\"]\n",
    "\n",
    "            # add column\n",
    "            df_tmp[\"Time point\"]=cond\n",
    "            df_tmp[\"KO\"]=KO\n",
    "            df_tmp[\"Assay\"]=assay\n",
    "\n",
    "            # set indices\n",
    "            df_tmp.reset_index(inplace=True)\n",
    "            df_tmp.set_index([\"Gene\", \"Time point\",\"KO\", \"Assay\"], inplace=True)\n",
    "\n",
    "            dea_dfs.append(df_tmp)\n",
    "\n",
    "# dea_df = pd.concat(dea_dfs, axis=1)\n",
    "dea_df = pd.concat(dea_dfs, axis=0)\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = dea_df,\n",
    "                     sheet_name = \"DEA_KO_effect_in_time_points\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "print(dea_df.shape)\n",
    "dea_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "98442704",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(28398, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Database</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Time point</th>\n",
       "      <th>KO</th>\n",
       "      <th>Direction</th>\n",
       "      <th>Term</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">Untreated</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">Stat1</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">up</th>\n",
       "      <th>C21-steroid hormone biosynthetic process (GO:0006700)</th>\n",
       "      <td>1/7</td>\n",
       "      <td>0.028397</td>\n",
       "      <td>0.083669</td>\n",
       "      <td>35.417989</td>\n",
       "      <td>TSPO</td>\n",
       "      <td>GO_Biological_Process_2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C21-steroid hormone metabolic process (GO:0008207)</th>\n",
       "      <td>1/9</td>\n",
       "      <td>0.036364</td>\n",
       "      <td>0.098361</td>\n",
       "      <td>27.543210</td>\n",
       "      <td>TSPO</td>\n",
       "      <td>GO_Biological_Process_2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD4-positive, alpha-beta T cell activation (GO:0035710)</th>\n",
       "      <td>1/3</td>\n",
       "      <td>0.012268</td>\n",
       "      <td>0.064605</td>\n",
       "      <td>82.666667</td>\n",
       "      <td>RSAD2</td>\n",
       "      <td>GO_Biological_Process_2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNA catabolic process (GO:0006308)</th>\n",
       "      <td>1/9</td>\n",
       "      <td>0.036364</td>\n",
       "      <td>0.098361</td>\n",
       "      <td>27.543210</td>\n",
       "      <td>ISG20</td>\n",
       "      <td>GO_Biological_Process_2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNA catabolic process, exonucleolytic (GO:0000738)</th>\n",
       "      <td>1/3</td>\n",
       "      <td>0.012268</td>\n",
       "      <td>0.064605</td>\n",
       "      <td>82.666667</td>\n",
       "      <td>ISG20</td>\n",
       "      <td>GO_Biological_Process_2021</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                              Overlap  \\\n",
       "Time point KO    Direction Term                                                         \n",
       "Untreated  Stat1 up        C21-steroid hormone biosynthetic process (GO:00...     1/7   \n",
       "                           C21-steroid hormone metabolic process (GO:0008207)     1/9   \n",
       "                           CD4-positive, alpha-beta T cell activation (GO:...     1/3   \n",
       "                           DNA catabolic process (GO:0006308)                     1/9   \n",
       "                           DNA catabolic process, exonucleolytic (GO:0000738)     1/3   \n",
       "\n",
       "                                                                                P-value  \\\n",
       "Time point KO    Direction Term                                                           \n",
       "Untreated  Stat1 up        C21-steroid hormone biosynthetic process (GO:00...  0.028397   \n",
       "                           C21-steroid hormone metabolic process (GO:0008207)  0.036364   \n",
       "                           CD4-positive, alpha-beta T cell activation (GO:...  0.012268   \n",
       "                           DNA catabolic process (GO:0006308)                  0.036364   \n",
       "                           DNA catabolic process, exonucleolytic (GO:0000738)  0.012268   \n",
       "\n",
       "                                                                               Adjusted P-value  \\\n",
       "Time point KO    Direction Term                                                                   \n",
       "Untreated  Stat1 up        C21-steroid hormone biosynthetic process (GO:00...          0.083669   \n",
       "                           C21-steroid hormone metabolic process (GO:0008207)          0.098361   \n",
       "                           CD4-positive, alpha-beta T cell activation (GO:...          0.064605   \n",
       "                           DNA catabolic process (GO:0006308)                          0.098361   \n",
       "                           DNA catabolic process, exonucleolytic (GO:0000738)          0.064605   \n",
       "\n",
       "                                                                               Odds Ratio  \\\n",
       "Time point KO    Direction Term                                                             \n",
       "Untreated  Stat1 up        C21-steroid hormone biosynthetic process (GO:00...   35.417989   \n",
       "                           C21-steroid hormone metabolic process (GO:0008207)   27.543210   \n",
       "                           CD4-positive, alpha-beta T cell activation (GO:...   82.666667   \n",
       "                           DNA catabolic process (GO:0006308)                   27.543210   \n",
       "                           DNA catabolic process, exonucleolytic (GO:0000738)   82.666667   \n",
       "\n",
       "                                                                               Genes  \\\n",
       "Time point KO    Direction Term                                                        \n",
       "Untreated  Stat1 up        C21-steroid hormone biosynthetic process (GO:00...   TSPO   \n",
       "                           C21-steroid hormone metabolic process (GO:0008207)   TSPO   \n",
       "                           CD4-positive, alpha-beta T cell activation (GO:...  RSAD2   \n",
       "                           DNA catabolic process (GO:0006308)                  ISG20   \n",
       "                           DNA catabolic process, exonucleolytic (GO:0000738)  ISG20   \n",
       "\n",
       "                                                                                                 Database  \n",
       "Time point KO    Direction Term                                                                            \n",
       "Untreated  Stat1 up        C21-steroid hormone biosynthetic process (GO:00...  GO_Biological_Process_2021  \n",
       "                           C21-steroid hormone metabolic process (GO:0008207)  GO_Biological_Process_2021  \n",
       "                           CD4-positive, alpha-beta T cell activation (GO:...  GO_Biological_Process_2021  \n",
       "                           DNA catabolic process (GO:0006308)                  GO_Biological_Process_2021  \n",
       "                           DNA catabolic process, exonucleolytic (GO:0000738)  GO_Biological_Process_2021  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# enrichments of DEA results post-mixscape KO effects within conditions -> GO BP\n",
    "enr_dfs = []\n",
    "db = \"GO_Biological_Process_2021\"\n",
    "\n",
    "for cond in conditions:\n",
    "    KOs = list(metadata.loc[(metadata[\"mixscape_class.global\"]==\"KO\") & (metadata[\"Condition\"]==cond), \"KO_call\"].unique())\n",
    "    for KO in KOs:\n",
    "        for direction in ['up','down']:\n",
    "            tmp_path = os.path.join(data_path, 'enrichment_analysis', 'DEA_condition_mixscape', '{}_m{}_{}'.format(cond, KO, direction), db, 'Enrichr_{}_{}_m{}_{}.csv'.format(db, cond, KO, direction))\n",
    "            \n",
    "            # check if exists\n",
    "            if not(os.path.exists(tmp_path)):\n",
    "                continue\n",
    "\n",
    "            # load enrichment results\n",
    "            df_tmp = pd.read_csv(tmp_path)\n",
    "            \n",
    "            # Drop columns\n",
    "            df_tmp = df_tmp.drop(['Unnamed: 0', 'Gene_set'], axis=1)\n",
    "\n",
    "            # add column\n",
    "            df_tmp[\"Time point\"]=cond\n",
    "            df_tmp[\"KO\"]=KO\n",
    "            df_tmp[\"Direction\"] = direction\n",
    "            \n",
    "            # set indices\n",
    "            df_tmp.set_index([\"Time point\", \"KO\", \"Direction\", \"Term\"], inplace=True)\n",
    "\n",
    "            enr_dfs.append(df_tmp)\n",
    "\n",
    "# dea_df = pd.concat(dea_dfs, axis=1)\n",
    "enr_df = pd.concat(enr_dfs, axis=0)\n",
    "\n",
    "# add database column\n",
    "enr_df[\"Database\"] = db\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = enr_df,\n",
    "                     sheet_name = \"enrichments_KO_in_time_points\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "print(enr_df.shape)\n",
    "enr_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "0c63edb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save Excel file\n",
    "writer.save()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c27d236",
   "metadata": {},
   "source": [
    "# CROP-seq large - TableS5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "33bd07ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = os.path.join('results', 'KO150')\n",
    "conditions = [\"untreated\", \"LO28-6h\", \"LO28-24h\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "99eed480",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create Excel writer\n",
    "writer = pd.ExcelWriter(os.path.join(results_path,'S05_CROPseq_large.xlsx'), engine='xlsxwriter')\n",
    "\n",
    "# Get the workbook and worksheet objects\n",
    "workbook = writer.book\n",
    "# define specific formatting\n",
    "index_format = workbook.add_format(xlsx_format_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "22c5b9a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add an empty sheet called LEGEND (fill with manually curated data later)\n",
    "pd.DataFrame().to_excel(writer, sheet_name=\"TableS5_legends\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "6b265763",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(28303, 22)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Condition</th>\n",
       "      <th>KO_call</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>nCount_AB</th>\n",
       "      <th>nFeature_AB</th>\n",
       "      <th>nCount_gRNA</th>\n",
       "      <th>nFeature_gRNA</th>\n",
       "      <th>batch</th>\n",
       "      <th>percent.mt</th>\n",
       "      <th>...</th>\n",
       "      <th>nFeature_SCT</th>\n",
       "      <th>S.Score</th>\n",
       "      <th>G2M.Score</th>\n",
       "      <th>Phase</th>\n",
       "      <th>mixscape_class</th>\n",
       "      <th>mixscape_class_p_ko</th>\n",
       "      <th>mixscape_class.global</th>\n",
       "      <th>joint_analysis_mixscape_class</th>\n",
       "      <th>joint_analysis_mixscape_class_p_ko</th>\n",
       "      <th>joint_analysis_mixscape_class.global</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1_AAACCTGAGAATCTCC-1</th>\n",
       "      <td>untreated</td>\n",
       "      <td>Chd7</td>\n",
       "      <td>17884</td>\n",
       "      <td>4442</td>\n",
       "      <td>987</td>\n",
       "      <td>9</td>\n",
       "      <td>358</td>\n",
       "      <td>51</td>\n",
       "      <td>A1</td>\n",
       "      <td>1.481771</td>\n",
       "      <td>...</td>\n",
       "      <td>4442</td>\n",
       "      <td>0.628917</td>\n",
       "      <td>0.252506</td>\n",
       "      <td>S</td>\n",
       "      <td>Chd7 NP</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NP</td>\n",
       "      <td>Chd7 NP</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1_AAACCTGAGCGATGAC-1</th>\n",
       "      <td>untreated</td>\n",
       "      <td>Babam1</td>\n",
       "      <td>12849</td>\n",
       "      <td>3787</td>\n",
       "      <td>1378</td>\n",
       "      <td>9</td>\n",
       "      <td>1103</td>\n",
       "      <td>56</td>\n",
       "      <td>A1</td>\n",
       "      <td>1.299712</td>\n",
       "      <td>...</td>\n",
       "      <td>3787</td>\n",
       "      <td>0.033039</td>\n",
       "      <td>0.527634</td>\n",
       "      <td>G2M</td>\n",
       "      <td>Babam1 NP</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NP</td>\n",
       "      <td>Babam1 NP</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1_AAACCTGAGTGCCAGA-1</th>\n",
       "      <td>untreated</td>\n",
       "      <td>Mov10</td>\n",
       "      <td>14479</td>\n",
       "      <td>3851</td>\n",
       "      <td>458</td>\n",
       "      <td>10</td>\n",
       "      <td>838</td>\n",
       "      <td>72</td>\n",
       "      <td>A1</td>\n",
       "      <td>1.201740</td>\n",
       "      <td>...</td>\n",
       "      <td>3851</td>\n",
       "      <td>0.027528</td>\n",
       "      <td>0.626059</td>\n",
       "      <td>G2M</td>\n",
       "      <td>Mov10 NP</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NP</td>\n",
       "      <td>Mov10 NP</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1_AAACCTGAGTGGAGAA-1</th>\n",
       "      <td>untreated</td>\n",
       "      <td>Spi1</td>\n",
       "      <td>13765</td>\n",
       "      <td>3832</td>\n",
       "      <td>1437</td>\n",
       "      <td>9</td>\n",
       "      <td>270</td>\n",
       "      <td>117</td>\n",
       "      <td>A1</td>\n",
       "      <td>1.031602</td>\n",
       "      <td>...</td>\n",
       "      <td>3832</td>\n",
       "      <td>0.578333</td>\n",
       "      <td>-0.045930</td>\n",
       "      <td>S</td>\n",
       "      <td>Spi1 NP</td>\n",
       "      <td>0.054726</td>\n",
       "      <td>NP</td>\n",
       "      <td>Spi1 NP</td>\n",
       "      <td>0.067429</td>\n",
       "      <td>NP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1_AAACCTGAGTTTAGGA-1</th>\n",
       "      <td>untreated</td>\n",
       "      <td>Jmjd1c</td>\n",
       "      <td>22985</td>\n",
       "      <td>5073</td>\n",
       "      <td>2638</td>\n",
       "      <td>10</td>\n",
       "      <td>267</td>\n",
       "      <td>42</td>\n",
       "      <td>A1</td>\n",
       "      <td>1.096367</td>\n",
       "      <td>...</td>\n",
       "      <td>5056</td>\n",
       "      <td>0.169754</td>\n",
       "      <td>-0.372522</td>\n",
       "      <td>S</td>\n",
       "      <td>Jmjd1c NP</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NP</td>\n",
       "      <td>Jmjd1c NP</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Condition KO_call  nCount_RNA  nFeature_RNA  nCount_AB  \\\n",
       "Cell barcode                                                                    \n",
       "A1_AAACCTGAGAATCTCC-1  untreated    Chd7       17884          4442        987   \n",
       "A1_AAACCTGAGCGATGAC-1  untreated  Babam1       12849          3787       1378   \n",
       "A1_AAACCTGAGTGCCAGA-1  untreated   Mov10       14479          3851        458   \n",
       "A1_AAACCTGAGTGGAGAA-1  untreated    Spi1       13765          3832       1437   \n",
       "A1_AAACCTGAGTTTAGGA-1  untreated  Jmjd1c       22985          5073       2638   \n",
       "\n",
       "                       nFeature_AB  nCount_gRNA  nFeature_gRNA batch  \\\n",
       "Cell barcode                                                           \n",
       "A1_AAACCTGAGAATCTCC-1            9          358             51    A1   \n",
       "A1_AAACCTGAGCGATGAC-1            9         1103             56    A1   \n",
       "A1_AAACCTGAGTGCCAGA-1           10          838             72    A1   \n",
       "A1_AAACCTGAGTGGAGAA-1            9          270            117    A1   \n",
       "A1_AAACCTGAGTTTAGGA-1           10          267             42    A1   \n",
       "\n",
       "                       percent.mt  ... nFeature_SCT   S.Score  G2M.Score  \\\n",
       "Cell barcode                       ...                                     \n",
       "A1_AAACCTGAGAATCTCC-1    1.481771  ...         4442  0.628917   0.252506   \n",
       "A1_AAACCTGAGCGATGAC-1    1.299712  ...         3787  0.033039   0.527634   \n",
       "A1_AAACCTGAGTGCCAGA-1    1.201740  ...         3851  0.027528   0.626059   \n",
       "A1_AAACCTGAGTGGAGAA-1    1.031602  ...         3832  0.578333  -0.045930   \n",
       "A1_AAACCTGAGTTTAGGA-1    1.096367  ...         5056  0.169754  -0.372522   \n",
       "\n",
       "                       Phase  mixscape_class mixscape_class_p_ko  \\\n",
       "Cell barcode                                                       \n",
       "A1_AAACCTGAGAATCTCC-1      S         Chd7 NP            0.000000   \n",
       "A1_AAACCTGAGCGATGAC-1    G2M       Babam1 NP            0.000000   \n",
       "A1_AAACCTGAGTGCCAGA-1    G2M        Mov10 NP            0.000000   \n",
       "A1_AAACCTGAGTGGAGAA-1      S         Spi1 NP            0.054726   \n",
       "A1_AAACCTGAGTTTAGGA-1      S       Jmjd1c NP            0.000000   \n",
       "\n",
       "                      mixscape_class.global  joint_analysis_mixscape_class  \\\n",
       "Cell barcode                                                                 \n",
       "A1_AAACCTGAGAATCTCC-1                    NP                        Chd7 NP   \n",
       "A1_AAACCTGAGCGATGAC-1                    NP                      Babam1 NP   \n",
       "A1_AAACCTGAGTGCCAGA-1                    NP                       Mov10 NP   \n",
       "A1_AAACCTGAGTGGAGAA-1                    NP                        Spi1 NP   \n",
       "A1_AAACCTGAGTTTAGGA-1                    NP                      Jmjd1c NP   \n",
       "\n",
       "                      joint_analysis_mixscape_class_p_ko  \\\n",
       "Cell barcode                                               \n",
       "A1_AAACCTGAGAATCTCC-1                           0.000000   \n",
       "A1_AAACCTGAGCGATGAC-1                           0.000000   \n",
       "A1_AAACCTGAGTGCCAGA-1                           0.000000   \n",
       "A1_AAACCTGAGTGGAGAA-1                           0.067429   \n",
       "A1_AAACCTGAGTTTAGGA-1                           0.000000   \n",
       "\n",
       "                      joint_analysis_mixscape_class.global  \n",
       "Cell barcode                                                \n",
       "A1_AAACCTGAGAATCTCC-1                                   NP  \n",
       "A1_AAACCTGAGCGATGAC-1                                   NP  \n",
       "A1_AAACCTGAGTGCCAGA-1                                   NP  \n",
       "A1_AAACCTGAGTGGAGAA-1                                   NP  \n",
       "A1_AAACCTGAGTTTAGGA-1                                   NP  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# metadata\n",
    "metadata = pd.read_csv(os.path.join(data_path, 'KO150_NORMALIZED_metadata.csv'), index_col=0)\n",
    "# set condition column\n",
    "metadata['Condition'] = metadata['hto_demux'].str.split(\"_\", expand=True)[1]\n",
    "\n",
    "# load mixscape classifications per condition\n",
    "mixscape_dfs = []\n",
    "for cond in conditions:\n",
    "    cond_tmp = pd.read_csv(os.path.join(data_path, 'mixscape', cond, 'KO150_mixscape_annotations.csv'), index_col=0)\n",
    "    mixscape_dfs.append(cond_tmp)\n",
    "    \n",
    "# load mixscape classification across condition (joint analysis)\n",
    "mixscape_all = pd.read_csv(os.path.join(data_path, 'mixscape', 'all', 'KO150_mixscape_annotations.csv'), index_col=0).add_prefix('joint_analysis_')\n",
    "\n",
    "# concatenate dfs\n",
    "mixscape_df = pd.concat(mixscape_dfs, axis=0)\n",
    "metadata = pd.concat([metadata, mixscape_df, mixscape_all], axis=1)\n",
    "\n",
    "# Drop columns\n",
    "metadata = metadata.drop(['old.ident', \"orig.ident\", \"pass_QC\", \"hto_demux\", \"nCount_HTO\", \"nFeature_HTO\"], axis=1)\n",
    "\n",
    "# add index name\n",
    "metadata.index.names = [\"Cell barcode\"]\n",
    "\n",
    "# set indices\n",
    "metadata.reset_index(inplace=True)\n",
    "metadata.set_index([\"Cell barcode\", \"Condition\", \"KO_call\"], inplace=True)\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = metadata.copy(deep=True),\n",
    "                     sheet_name = \"single_cell_annotations\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "# reset indices for downstream steps\n",
    "metadata.reset_index(inplace=True)\n",
    "metadata.set_index([\"Cell barcode\"], inplace=True)\n",
    "\n",
    "print(metadata.shape)\n",
    "metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "d339ea45",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(50848, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>P-value</th>\n",
       "      <th>Avg log2FC</th>\n",
       "      <th>Percentage expression 1</th>\n",
       "      <th>Percentage expression 2</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th>Time point</th>\n",
       "      <th>KO</th>\n",
       "      <th>Assay</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">S100a6</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">Untreated</th>\n",
       "      <th>Smc1a</th>\n",
       "      <th>RNA</th>\n",
       "      <td>1.813502e-18</td>\n",
       "      <td>-1.430115</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.692143e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ifnar1</th>\n",
       "      <th>RNA</th>\n",
       "      <td>8.592256e-02</td>\n",
       "      <td>-0.133587</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sfpq</th>\n",
       "      <th>RNA</th>\n",
       "      <td>2.182019e-01</td>\n",
       "      <td>-0.125579</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tyk2</th>\n",
       "      <th>RNA</th>\n",
       "      <td>3.080555e-01</td>\n",
       "      <td>-0.148946</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Spi1</th>\n",
       "      <th>RNA</th>\n",
       "      <td>7.399310e-06</td>\n",
       "      <td>0.562207</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.098428e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     P-value  Avg log2FC  \\\n",
       "Gene   Time point KO     Assay                             \n",
       "S100a6 Untreated  Smc1a  RNA    1.813502e-18   -1.430115   \n",
       "                  Ifnar1 RNA    8.592256e-02   -0.133587   \n",
       "                  Sfpq   RNA    2.182019e-01   -0.125579   \n",
       "                  Tyk2   RNA    3.080555e-01   -0.148946   \n",
       "                  Spi1   RNA    7.399310e-06    0.562207   \n",
       "\n",
       "                                Percentage expression 1  \\\n",
       "Gene   Time point KO     Assay                            \n",
       "S100a6 Untreated  Smc1a  RNA                        1.0   \n",
       "                  Ifnar1 RNA                        1.0   \n",
       "                  Sfpq   RNA                        1.0   \n",
       "                  Tyk2   RNA                        1.0   \n",
       "                  Spi1   RNA                        1.0   \n",
       "\n",
       "                                Percentage expression 2  Adjusted P-value  \n",
       "Gene   Time point KO     Assay                                             \n",
       "S100a6 Untreated  Smc1a  RNA                        1.0      2.692143e-14  \n",
       "                  Ifnar1 RNA                        1.0      1.000000e+00  \n",
       "                  Sfpq   RNA                        1.0      1.000000e+00  \n",
       "                  Tyk2   RNA                        1.0      1.000000e+00  \n",
       "                  Spi1   RNA                        1.0      1.098428e-01  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# DEA post-mixscape KO effects within conditions\n",
    "dea_dfs = []\n",
    "\n",
    "for assay in ['RNA', 'AB']:\n",
    "    for cond in conditions:\n",
    "        KOs = list(metadata.loc[(metadata[\"mixscape_class.global\"]==\"KO\") & (metadata[\"Condition\"]==cond), \"KO_call\"].unique())\n",
    "        for KO in KOs:\n",
    "            \n",
    "            # load dea results\n",
    "            df_tmp = pd.read_csv(os.path.join(data_path, 'DEA_condition_mixscape', 'KO150_DEA_{}_{}_{}.csv'.format(assay, cond, KO)), index_col=0)\n",
    "\n",
    "            # set index name\n",
    "            df_tmp.index.names = [\"Gene\"]\n",
    "\n",
    "            # add column\n",
    "            df_tmp[\"Time point\"]=cond\n",
    "            df_tmp[\"KO\"]=KO\n",
    "            df_tmp[\"Assay\"]=assay\n",
    "\n",
    "            # set indices\n",
    "            df_tmp.reset_index(inplace=True)\n",
    "            df_tmp.set_index([\"Gene\", \"Time point\",\"KO\", \"Assay\"], inplace=True)\n",
    "\n",
    "            dea_dfs.append(df_tmp)\n",
    "\n",
    "# dea_df = pd.concat(dea_dfs, axis=1)\n",
    "dea_df = pd.concat(dea_dfs, axis=0)\n",
    "\n",
    "\n",
    "# filter for genes that at least in one comparison were statistically significant\n",
    "genes_sig = dea_df.index[(dea_df[\"p_val_adj\"]<0.05)].get_level_values('Gene').unique().tolist()\n",
    "dea_df = dea_df.loc[genes_sig,:]\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = dea_df,\n",
    "                     sheet_name = \"DEA_KO_effect_in_time_points\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "print(dea_df.shape)\n",
    "dea_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "c9c277a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(47160, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Database</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Time point</th>\n",
       "      <th>KO</th>\n",
       "      <th>Direction</th>\n",
       "      <th>Term</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">Untreated</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">Smc1a</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">up</th>\n",
       "      <th>BMP signaling pathway (GO:0030509)</th>\n",
       "      <td>1/39</td>\n",
       "      <td>0.036174</td>\n",
       "      <td>0.076640</td>\n",
       "      <td>29.203156</td>\n",
       "      <td>PDCD4</td>\n",
       "      <td>GO_Biological_Process_2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERAD pathway (GO:0036503)</th>\n",
       "      <td>1/77</td>\n",
       "      <td>0.070249</td>\n",
       "      <td>0.104537</td>\n",
       "      <td>14.753247</td>\n",
       "      <td>FBXO2</td>\n",
       "      <td>GO_Biological_Process_2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCF-dependent proteasomal ubiquitin-dependent protein catabolic process (GO:0031146)</th>\n",
       "      <td>1/83</td>\n",
       "      <td>0.075525</td>\n",
       "      <td>0.109773</td>\n",
       "      <td>13.681186</td>\n",
       "      <td>FBXO2</td>\n",
       "      <td>GO_Biological_Process_2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>apoptotic process (GO:0006915)</th>\n",
       "      <td>1/169</td>\n",
       "      <td>0.148167</td>\n",
       "      <td>0.173093</td>\n",
       "      <td>6.680018</td>\n",
       "      <td>PDCD4</td>\n",
       "      <td>GO_Biological_Process_2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>carbohydrate derivative catabolic process (GO:1901136)</th>\n",
       "      <td>1/13</td>\n",
       "      <td>0.012196</td>\n",
       "      <td>0.046886</td>\n",
       "      <td>87.763314</td>\n",
       "      <td>FBXO2</td>\n",
       "      <td>GO_Biological_Process_2021</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                              Overlap  \\\n",
       "Time point KO    Direction Term                                                         \n",
       "Untreated  Smc1a up        BMP signaling pathway (GO:0030509)                    1/39   \n",
       "                           ERAD pathway (GO:0036503)                             1/77   \n",
       "                           SCF-dependent proteasomal ubiquitin-dependent p...    1/83   \n",
       "                           apoptotic process (GO:0006915)                       1/169   \n",
       "                           carbohydrate derivative catabolic process (GO:1...    1/13   \n",
       "\n",
       "                                                                                P-value  \\\n",
       "Time point KO    Direction Term                                                           \n",
       "Untreated  Smc1a up        BMP signaling pathway (GO:0030509)                  0.036174   \n",
       "                           ERAD pathway (GO:0036503)                           0.070249   \n",
       "                           SCF-dependent proteasomal ubiquitin-dependent p...  0.075525   \n",
       "                           apoptotic process (GO:0006915)                      0.148167   \n",
       "                           carbohydrate derivative catabolic process (GO:1...  0.012196   \n",
       "\n",
       "                                                                               Adjusted P-value  \\\n",
       "Time point KO    Direction Term                                                                   \n",
       "Untreated  Smc1a up        BMP signaling pathway (GO:0030509)                          0.076640   \n",
       "                           ERAD pathway (GO:0036503)                                   0.104537   \n",
       "                           SCF-dependent proteasomal ubiquitin-dependent p...          0.109773   \n",
       "                           apoptotic process (GO:0006915)                              0.173093   \n",
       "                           carbohydrate derivative catabolic process (GO:1...          0.046886   \n",
       "\n",
       "                                                                               Odds Ratio  \\\n",
       "Time point KO    Direction Term                                                             \n",
       "Untreated  Smc1a up        BMP signaling pathway (GO:0030509)                   29.203156   \n",
       "                           ERAD pathway (GO:0036503)                            14.753247   \n",
       "                           SCF-dependent proteasomal ubiquitin-dependent p...   13.681186   \n",
       "                           apoptotic process (GO:0006915)                        6.680018   \n",
       "                           carbohydrate derivative catabolic process (GO:1...   87.763314   \n",
       "\n",
       "                                                                               Genes  \\\n",
       "Time point KO    Direction Term                                                        \n",
       "Untreated  Smc1a up        BMP signaling pathway (GO:0030509)                  PDCD4   \n",
       "                           ERAD pathway (GO:0036503)                           FBXO2   \n",
       "                           SCF-dependent proteasomal ubiquitin-dependent p...  FBXO2   \n",
       "                           apoptotic process (GO:0006915)                      PDCD4   \n",
       "                           carbohydrate derivative catabolic process (GO:1...  FBXO2   \n",
       "\n",
       "                                                                                                 Database  \n",
       "Time point KO    Direction Term                                                                            \n",
       "Untreated  Smc1a up        BMP signaling pathway (GO:0030509)                  GO_Biological_Process_2021  \n",
       "                           ERAD pathway (GO:0036503)                           GO_Biological_Process_2021  \n",
       "                           SCF-dependent proteasomal ubiquitin-dependent p...  GO_Biological_Process_2021  \n",
       "                           apoptotic process (GO:0006915)                      GO_Biological_Process_2021  \n",
       "                           carbohydrate derivative catabolic process (GO:1...  GO_Biological_Process_2021  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# enrichments of DEA results post-mixscape KO effects within conditions\n",
    "enr_dfs = []\n",
    "db = \"GO_Biological_Process_2021\"\n",
    "\n",
    "for cond in conditions:\n",
    "    KOs = list(metadata.loc[(metadata[\"mixscape_class.global\"]==\"KO\") & (metadata[\"Condition\"]==cond), \"KO_call\"].unique())\n",
    "    \n",
    "    for KO in KOs:\n",
    "        for direction in ['up','down']:\n",
    "            tmp_path = os.path.join(data_path, 'enrichment_analysis', 'DEA_condition_mixscape', '{}_{}_{}'.format(cond, KO, direction), db, 'Enrichr_{}_{}_{}_{}.csv'.format(db, cond, KO, direction))\n",
    "            \n",
    "            # check if exists\n",
    "            if not(os.path.exists(tmp_path)):\n",
    "                continue\n",
    "\n",
    "            # load enrichment results\n",
    "            df_tmp = pd.read_csv(tmp_path)\n",
    "            \n",
    "            # Drop columns\n",
    "            df_tmp = df_tmp.drop(['Unnamed: 0', 'Gene_set'], axis=1)\n",
    "\n",
    "            # add column\n",
    "            df_tmp[\"Time point\"]=cond\n",
    "            df_tmp[\"KO\"]=KO\n",
    "            df_tmp[\"Direction\"] = direction\n",
    "            \n",
    "            # set indices\n",
    "            df_tmp.set_index([\"Time point\", \"KO\", \"Direction\", \"Term\"], inplace=True)\n",
    "\n",
    "            enr_dfs.append(df_tmp)\n",
    "\n",
    "# dea_df = pd.concat(dea_dfs, axis=1)\n",
    "enr_df = pd.concat(enr_dfs, axis=0)\n",
    "\n",
    "# add database column\n",
    "enr_df[\"Database\"] = db\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = enr_df,\n",
    "                     sheet_name = \"enrichments_KO_in_time_points\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "print(enr_df.shape)\n",
    "enr_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "8a062128",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "untreated (21, 21)\n",
      "LO28-6h (25, 25)\n",
      "LO28-24h (12, 12)\n"
     ]
    }
   ],
   "source": [
    "# Cross-prediction graph as adjacency matrix for each condition\n",
    "\n",
    "for cond in conditions:\n",
    "    tmp_path = os.path.join(data_path, 'KO_classifier', cond, 'KO150_clfLR_connprobNorm.csv')\n",
    "\n",
    "    df_tmp = pd.read_csv(tmp_path, index_col=0)\n",
    "            \n",
    "    print(cond + \" \" + str(df_tmp.shape))\n",
    "    \n",
    "    df_tmp.index.names = [cond]\n",
    "\n",
    "    # add dataframe as sheet to excel file\n",
    "    writer = df_to_excel(writer = writer,\n",
    "                         df = df_tmp,\n",
    "                         sheet_name = \"crossprediction_{}\".format(reduce(lambda s, kv: s.replace(*kv), replacements.items(), cond).replace(\" \", \"_\")),\n",
    "                         column_width = 25,\n",
    "                         index_format = index_format\n",
    "                        )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "33d8f428",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "untreated (118, 5)\n",
      "LO28-6h (164, 5)\n",
      "LO28-24h (44, 5)\n"
     ]
    }
   ],
   "source": [
    "# STRING analysis results for each condition\n",
    "\n",
    "for cond in conditions:\n",
    "    tmp_path = os.path.join(data_path, 'KO_classifier', cond, 'KO150_clfLR_graphprob_STRING_comparison.csv')\n",
    "\n",
    "    df_tmp = pd.read_csv(tmp_path, index_col=0)\n",
    "    \n",
    "    # set column names\n",
    "    df_tmp.columns = [\"STRING interaction score\", \"Average crossprediction probability\", \"Overlap\", \"Unique in crossprediction\", \"Unique in STRING\"]\n",
    "            \n",
    "    print(cond + \" \" + str(df_tmp.shape))\n",
    "    \n",
    "    df_tmp.index.names = [\"Edge\"]\n",
    "\n",
    "    # add dataframe as sheet to excel file\n",
    "    writer = df_to_excel(writer = writer,\n",
    "                         df = df_tmp,\n",
    "                         sheet_name = \"STRING_{}\".format(reduce(lambda s, kv: s.replace(*kv), replacements.items(), cond).replace(\" \", \"_\")),\n",
    "                         column_width = 25,\n",
    "                         index_format = index_format\n",
    "                        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "b35bd2c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>STRING interaction score</th>\n",
       "      <th>Average crossprediction probability</th>\n",
       "      <th>Overlap</th>\n",
       "      <th>Unique in crossprediction</th>\n",
       "      <th>Unique in STRING</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Edge</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Stat2-Tyk2</th>\n",
       "      <td>0.994</td>\n",
       "      <td>0.106390</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Spi1-Tyk2</th>\n",
       "      <td>0.205</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Spi1-Stat2</th>\n",
       "      <td>0.376</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Smc1a-Spi1</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.162550</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sfpq-Spi1</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.052689</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            STRING interaction score  Average crossprediction probability  \\\n",
       "Edge                                                                        \n",
       "Stat2-Tyk2                     0.994                             0.106390   \n",
       "Spi1-Tyk2                      0.205                             0.000000   \n",
       "Spi1-Stat2                     0.376                             0.000000   \n",
       "Smc1a-Spi1                     0.000                             0.162550   \n",
       "Sfpq-Spi1                      0.000                             0.052689   \n",
       "\n",
       "            Overlap  Unique in crossprediction  Unique in STRING  \n",
       "Edge                                                              \n",
       "Stat2-Tyk2     True                      False             False  \n",
       "Spi1-Tyk2     False                      False              True  \n",
       "Spi1-Stat2    False                      False              True  \n",
       "Smc1a-Spi1    False                       True             False  \n",
       "Sfpq-Spi1     False                       True             False  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tmp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "1ed4af1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Save Excel file\n",
    "# writer.save()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1155233",
   "metadata": {},
   "source": [
    "# Joint perturbation analysis & Integrative Analysis - included in TableS5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "a5a18f50",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "all (58, 58)\n",
      "(58, 58)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Listeria 24h+Chd4 KO</th>\n",
       "      <th>Listeria 24h+Ep400 KO</th>\n",
       "      <th>Listeria 24h+Ifnar1 KO</th>\n",
       "      <th>Listeria 24h+Ikzf1 KO</th>\n",
       "      <th>Listeria 24h+Irf9 KO</th>\n",
       "      <th>Listeria 24h+Jak1 KO</th>\n",
       "      <th>Listeria 24h+Myd88 KO</th>\n",
       "      <th>Listeria 24h+Sfpq KO</th>\n",
       "      <th>Listeria 24h+Smc1a KO</th>\n",
       "      <th>Listeria 24h+Spi1 KO</th>\n",
       "      <th>...</th>\n",
       "      <th>Untreated+Jak1 KO</th>\n",
       "      <th>Untreated+Med14 KO</th>\n",
       "      <th>Untreated+Runx1 KO</th>\n",
       "      <th>Untreated+Sf3b1 KO</th>\n",
       "      <th>Untreated+Sfpq KO</th>\n",
       "      <th>Untreated+Smc1a KO</th>\n",
       "      <th>Untreated+Spi1 KO</th>\n",
       "      <th>Untreated+Stat2 KO</th>\n",
       "      <th>Untreated+Tyk2 KO</th>\n",
       "      <th>Untreated+Yeats2 KO</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Joint analysis</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Listeria 24h+Chd4 KO</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008974</td>\n",
       "      <td>0.007204</td>\n",
       "      <td>0.025959</td>\n",
       "      <td>0.016927</td>\n",
       "      <td>0.010049</td>\n",
       "      <td>0.021910</td>\n",
       "      <td>0.049981</td>\n",
       "      <td>0.042560</td>\n",
       "      <td>0.013292</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000871</td>\n",
       "      <td>0.008110</td>\n",
       "      <td>0.008865</td>\n",
       "      <td>0.001710</td>\n",
       "      <td>0.003715</td>\n",
       "      <td>0.003435</td>\n",
       "      <td>0.002087</td>\n",
       "      <td>0.001014</td>\n",
       "      <td>0.004084</td>\n",
       "      <td>0.003158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Listeria 24h+Ep400 KO</th>\n",
       "      <td>0.008861</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.016035</td>\n",
       "      <td>0.055534</td>\n",
       "      <td>0.013130</td>\n",
       "      <td>0.047038</td>\n",
       "      <td>0.037502</td>\n",
       "      <td>0.006059</td>\n",
       "      <td>0.025426</td>\n",
       "      <td>0.015297</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001112</td>\n",
       "      <td>0.001540</td>\n",
       "      <td>0.003305</td>\n",
       "      <td>0.006563</td>\n",
       "      <td>0.003341</td>\n",
       "      <td>0.001363</td>\n",
       "      <td>0.003347</td>\n",
       "      <td>0.002193</td>\n",
       "      <td>0.002919</td>\n",
       "      <td>0.014476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Listeria 24h+Ifnar1 KO</th>\n",
       "      <td>0.001781</td>\n",
       "      <td>0.002817</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004260</td>\n",
       "      <td>0.460557</td>\n",
       "      <td>0.197412</td>\n",
       "      <td>0.003267</td>\n",
       "      <td>0.004518</td>\n",
       "      <td>0.002764</td>\n",
       "      <td>0.002279</td>\n",
       "      <td>...</td>\n",
       "      <td>0.012862</td>\n",
       "      <td>0.000647</td>\n",
       "      <td>0.000117</td>\n",
       "      <td>0.001258</td>\n",
       "      <td>0.000107</td>\n",
       "      <td>0.000536</td>\n",
       "      <td>0.000099</td>\n",
       "      <td>0.010740</td>\n",
       "      <td>0.002304</td>\n",
       "      <td>0.000278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Listeria 24h+Ikzf1 KO</th>\n",
       "      <td>0.017429</td>\n",
       "      <td>0.017485</td>\n",
       "      <td>0.005919</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.037531</td>\n",
       "      <td>0.015704</td>\n",
       "      <td>0.070946</td>\n",
       "      <td>0.015897</td>\n",
       "      <td>0.031763</td>\n",
       "      <td>0.022469</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001733</td>\n",
       "      <td>0.004647</td>\n",
       "      <td>0.002164</td>\n",
       "      <td>0.002642</td>\n",
       "      <td>0.004280</td>\n",
       "      <td>0.004644</td>\n",
       "      <td>0.002851</td>\n",
       "      <td>0.001185</td>\n",
       "      <td>0.002078</td>\n",
       "      <td>0.005183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Listeria 24h+Irf9 KO</th>\n",
       "      <td>0.002619</td>\n",
       "      <td>0.003358</td>\n",
       "      <td>0.385940</td>\n",
       "      <td>0.011316</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.164542</td>\n",
       "      <td>0.006078</td>\n",
       "      <td>0.002579</td>\n",
       "      <td>0.003927</td>\n",
       "      <td>0.000792</td>\n",
       "      <td>...</td>\n",
       "      <td>0.031174</td>\n",
       "      <td>0.000387</td>\n",
       "      <td>0.000328</td>\n",
       "      <td>0.001361</td>\n",
       "      <td>0.000063</td>\n",
       "      <td>0.000488</td>\n",
       "      <td>0.000179</td>\n",
       "      <td>0.013295</td>\n",
       "      <td>0.004183</td>\n",
       "      <td>0.000829</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  58 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Listeria 24h+Chd4 KO  Listeria 24h+Ep400 KO  \\\n",
       "Joint analysis                                                        \n",
       "Listeria 24h+Chd4 KO                0.000000               0.008974   \n",
       "Listeria 24h+Ep400 KO               0.008861               0.000000   \n",
       "Listeria 24h+Ifnar1 KO              0.001781               0.002817   \n",
       "Listeria 24h+Ikzf1 KO               0.017429               0.017485   \n",
       "Listeria 24h+Irf9 KO                0.002619               0.003358   \n",
       "\n",
       "                        Listeria 24h+Ifnar1 KO  Listeria 24h+Ikzf1 KO  \\\n",
       "Joint analysis                                                          \n",
       "Listeria 24h+Chd4 KO                  0.007204               0.025959   \n",
       "Listeria 24h+Ep400 KO                 0.016035               0.055534   \n",
       "Listeria 24h+Ifnar1 KO                0.000000               0.004260   \n",
       "Listeria 24h+Ikzf1 KO                 0.005919               0.000000   \n",
       "Listeria 24h+Irf9 KO                  0.385940               0.011316   \n",
       "\n",
       "                        Listeria 24h+Irf9 KO  Listeria 24h+Jak1 KO  \\\n",
       "Joint analysis                                                       \n",
       "Listeria 24h+Chd4 KO                0.016927              0.010049   \n",
       "Listeria 24h+Ep400 KO               0.013130              0.047038   \n",
       "Listeria 24h+Ifnar1 KO              0.460557              0.197412   \n",
       "Listeria 24h+Ikzf1 KO               0.037531              0.015704   \n",
       "Listeria 24h+Irf9 KO                0.000000              0.164542   \n",
       "\n",
       "                        Listeria 24h+Myd88 KO  Listeria 24h+Sfpq KO  \\\n",
       "Joint analysis                                                        \n",
       "Listeria 24h+Chd4 KO                 0.021910              0.049981   \n",
       "Listeria 24h+Ep400 KO                0.037502              0.006059   \n",
       "Listeria 24h+Ifnar1 KO               0.003267              0.004518   \n",
       "Listeria 24h+Ikzf1 KO                0.070946              0.015897   \n",
       "Listeria 24h+Irf9 KO                 0.006078              0.002579   \n",
       "\n",
       "                        Listeria 24h+Smc1a KO  Listeria 24h+Spi1 KO  ...  \\\n",
       "Joint analysis                                                       ...   \n",
       "Listeria 24h+Chd4 KO                 0.042560              0.013292  ...   \n",
       "Listeria 24h+Ep400 KO                0.025426              0.015297  ...   \n",
       "Listeria 24h+Ifnar1 KO               0.002764              0.002279  ...   \n",
       "Listeria 24h+Ikzf1 KO                0.031763              0.022469  ...   \n",
       "Listeria 24h+Irf9 KO                 0.003927              0.000792  ...   \n",
       "\n",
       "                        Untreated+Jak1 KO  Untreated+Med14 KO  \\\n",
       "Joint analysis                                                  \n",
       "Listeria 24h+Chd4 KO             0.000871            0.008110   \n",
       "Listeria 24h+Ep400 KO            0.001112            0.001540   \n",
       "Listeria 24h+Ifnar1 KO           0.012862            0.000647   \n",
       "Listeria 24h+Ikzf1 KO            0.001733            0.004647   \n",
       "Listeria 24h+Irf9 KO             0.031174            0.000387   \n",
       "\n",
       "                        Untreated+Runx1 KO  Untreated+Sf3b1 KO  \\\n",
       "Joint analysis                                                   \n",
       "Listeria 24h+Chd4 KO              0.008865            0.001710   \n",
       "Listeria 24h+Ep400 KO             0.003305            0.006563   \n",
       "Listeria 24h+Ifnar1 KO            0.000117            0.001258   \n",
       "Listeria 24h+Ikzf1 KO             0.002164            0.002642   \n",
       "Listeria 24h+Irf9 KO              0.000328            0.001361   \n",
       "\n",
       "                        Untreated+Sfpq KO  Untreated+Smc1a KO  \\\n",
       "Joint analysis                                                  \n",
       "Listeria 24h+Chd4 KO             0.003715            0.003435   \n",
       "Listeria 24h+Ep400 KO            0.003341            0.001363   \n",
       "Listeria 24h+Ifnar1 KO           0.000107            0.000536   \n",
       "Listeria 24h+Ikzf1 KO            0.004280            0.004644   \n",
       "Listeria 24h+Irf9 KO             0.000063            0.000488   \n",
       "\n",
       "                        Untreated+Spi1 KO  Untreated+Stat2 KO  \\\n",
       "Joint analysis                                                  \n",
       "Listeria 24h+Chd4 KO             0.002087            0.001014   \n",
       "Listeria 24h+Ep400 KO            0.003347            0.002193   \n",
       "Listeria 24h+Ifnar1 KO           0.000099            0.010740   \n",
       "Listeria 24h+Ikzf1 KO            0.002851            0.001185   \n",
       "Listeria 24h+Irf9 KO             0.000179            0.013295   \n",
       "\n",
       "                        Untreated+Tyk2 KO  Untreated+Yeats2 KO  \n",
       "Joint analysis                                                  \n",
       "Listeria 24h+Chd4 KO             0.004084             0.003158  \n",
       "Listeria 24h+Ep400 KO            0.002919             0.014476  \n",
       "Listeria 24h+Ifnar1 KO           0.002304             0.000278  \n",
       "Listeria 24h+Ikzf1 KO            0.002078             0.005183  \n",
       "Listeria 24h+Irf9 KO             0.004183             0.000829  \n",
       "\n",
       "[5 rows x 58 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Cross-prediction graph as adjacency matrix across all conditions\n",
    "\n",
    "cond = \"all\"\n",
    "tmp_path = os.path.join(data_path, 'KO_classifier', cond, 'KO150_clfLR_connprobNorm.csv')\n",
    "\n",
    "df_tmp = pd.read_csv(tmp_path, index_col=0)\n",
    "\n",
    "df_tmp.index.names = [\"Joint analysis\"]\n",
    "\n",
    "print(cond + \" \" + str(df_tmp.shape))\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = df_tmp,\n",
    "                     sheet_name = \"joint_crossprediction_analysis\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "print(df_tmp.shape)\n",
    "df_tmp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "53246214",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(20, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>KO similarity</th>\n",
       "      <th>Time point similarity</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KO gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Arid1a</th>\n",
       "      <td>0.415096</td>\n",
       "      <td>0.538811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Brd2</th>\n",
       "      <td>0.469509</td>\n",
       "      <td>0.534615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Chd4</th>\n",
       "      <td>0.453651</td>\n",
       "      <td>0.446404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dnmt1</th>\n",
       "      <td>0.411223</td>\n",
       "      <td>0.538753</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dnttip2</th>\n",
       "      <td>0.214832</td>\n",
       "      <td>0.566572</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         KO similarity  Time point similarity\n",
       "KO gene                                      \n",
       "Arid1a        0.415096               0.538811\n",
       "Brd2          0.469509               0.534615\n",
       "Chd4          0.453651               0.446404\n",
       "Dnmt1         0.411223               0.538753\n",
       "Dnttip2       0.214832               0.566572"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Cross-prediction graph scores: Within-knockout and within-condition\n",
    "\n",
    "cond = \"all\"\n",
    "tmp_path = os.path.join(data_path, 'KO_classifier', cond, 'KO150_clfLR_graphprob_scores.csv')\n",
    "\n",
    "df_tmp = pd.read_csv(tmp_path, index_col=0)\n",
    "\n",
    "df_tmp.index.names = [\"KO gene\"]\n",
    "\n",
    "# rename columns\n",
    "df_tmp.columns = [\"KO_similarity\", \"Time point_similarity\"]\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = df_tmp,\n",
    "                     sheet_name = \"joint_crossprediction_scores\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "print(df_tmp.shape)\n",
    "df_tmp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "aee2d269",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(9657, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Database</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Time point 1</th>\n",
       "      <th>KO 1</th>\n",
       "      <th>Time point 2</th>\n",
       "      <th>KO 2</th>\n",
       "      <th>Direction</th>\n",
       "      <th>Term</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">Untreated</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">Ifnar1</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">Untreated</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">Tyk2</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">down</th>\n",
       "      <th>Activation of IRF3/IRF7 mediated by TBK1/IKK epsilon</th>\n",
       "      <td>1/15</td>\n",
       "      <td>0.008057</td>\n",
       "      <td>0.019135</td>\n",
       "      <td>141.238095</td>\n",
       "      <td>IRF7</td>\n",
       "      <td>BioPlanet_2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Antiviral mechanism by interferon-stimulated genes</th>\n",
       "      <td>1/64</td>\n",
       "      <td>0.033982</td>\n",
       "      <td>0.046118</td>\n",
       "      <td>32.993304</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>BioPlanet_2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Apoptosis</th>\n",
       "      <td>1/204</td>\n",
       "      <td>0.104815</td>\n",
       "      <td>0.107648</td>\n",
       "      <td>10.252801</td>\n",
       "      <td>IRF7</td>\n",
       "      <td>BioPlanet_2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Carbohydrate metabolism</th>\n",
       "      <td>1/165</td>\n",
       "      <td>0.085555</td>\n",
       "      <td>0.090308</td>\n",
       "      <td>12.709957</td>\n",
       "      <td>SDC3</td>\n",
       "      <td>BioPlanet_2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cell adhesion molecules (CAMs)</th>\n",
       "      <td>1/57</td>\n",
       "      <td>0.030315</td>\n",
       "      <td>0.042665</td>\n",
       "      <td>37.062657</td>\n",
       "      <td>SDC3</td>\n",
       "      <td>BioPlanet_2019</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                   Overlap  \\\n",
       "Time point 1 KO 1   Time point 2 KO 2 Direction Term                                                         \n",
       "Untreated    Ifnar1 Untreated    Tyk2 down      Activation of IRF3/IRF7 mediated by TBK1/IKK ep...    1/15   \n",
       "                                                Antiviral mechanism by interferon-stimulated genes    1/64   \n",
       "                                                Apoptosis                                            1/204   \n",
       "                                                Carbohydrate metabolism                              1/165   \n",
       "                                                Cell adhesion molecules (CAMs)                        1/57   \n",
       "\n",
       "                                                                                                     P-value  \\\n",
       "Time point 1 KO 1   Time point 2 KO 2 Direction Term                                                           \n",
       "Untreated    Ifnar1 Untreated    Tyk2 down      Activation of IRF3/IRF7 mediated by TBK1/IKK ep...  0.008057   \n",
       "                                                Antiviral mechanism by interferon-stimulated genes  0.033982   \n",
       "                                                Apoptosis                                           0.104815   \n",
       "                                                Carbohydrate metabolism                             0.085555   \n",
       "                                                Cell adhesion molecules (CAMs)                      0.030315   \n",
       "\n",
       "                                                                                                    Adjusted P-value  \\\n",
       "Time point 1 KO 1   Time point 2 KO 2 Direction Term                                                                   \n",
       "Untreated    Ifnar1 Untreated    Tyk2 down      Activation of IRF3/IRF7 mediated by TBK1/IKK ep...          0.019135   \n",
       "                                                Antiviral mechanism by interferon-stimulated genes          0.046118   \n",
       "                                                Apoptosis                                                   0.107648   \n",
       "                                                Carbohydrate metabolism                                     0.090308   \n",
       "                                                Cell adhesion molecules (CAMs)                              0.042665   \n",
       "\n",
       "                                                                                                    Odds Ratio  \\\n",
       "Time point 1 KO 1   Time point 2 KO 2 Direction Term                                                             \n",
       "Untreated    Ifnar1 Untreated    Tyk2 down      Activation of IRF3/IRF7 mediated by TBK1/IKK ep...  141.238095   \n",
       "                                                Antiviral mechanism by interferon-stimulated genes   32.993304   \n",
       "                                                Apoptosis                                            10.252801   \n",
       "                                                Carbohydrate metabolism                              12.709957   \n",
       "                                                Cell adhesion molecules (CAMs)                       37.062657   \n",
       "\n",
       "                                                                                                    Genes  \\\n",
       "Time point 1 KO 1   Time point 2 KO 2 Direction Term                                                        \n",
       "Untreated    Ifnar1 Untreated    Tyk2 down      Activation of IRF3/IRF7 mediated by TBK1/IKK ep...   IRF7   \n",
       "                                                Antiviral mechanism by interferon-stimulated genes  ISG15   \n",
       "                                                Apoptosis                                            IRF7   \n",
       "                                                Carbohydrate metabolism                              SDC3   \n",
       "                                                Cell adhesion molecules (CAMs)                       SDC3   \n",
       "\n",
       "                                                                                                          Database  \n",
       "Time point 1 KO 1   Time point 2 KO 2 Direction Term                                                                \n",
       "Untreated    Ifnar1 Untreated    Tyk2 down      Activation of IRF3/IRF7 mediated by TBK1/IKK ep...  BioPlanet_2019  \n",
       "                                                Antiviral mechanism by interferon-stimulated genes  BioPlanet_2019  \n",
       "                                                Apoptosis                                           BioPlanet_2019  \n",
       "                                                Carbohydrate metabolism                             BioPlanet_2019  \n",
       "                                                Cell adhesion molecules (CAMs)                      BioPlanet_2019  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Enrichment analysis results of overlapping DEGs of edges in crossprediction graph\n",
    "enr_dfs = []\n",
    "db = \"BioPlanet_2019\"\n",
    "\n",
    "for cond1 in conditions:\n",
    "    for cond2 in conditions:\n",
    "        for KO1 in metadata['KO_call'].unique():\n",
    "            for KO2 in metadata['KO_call'].unique():\n",
    "                for direction in ['up', 'down']:\n",
    "                    edge_desc = '{}+{}-{}+{}_{}'.format(cond1, KO1, cond2, KO2, direction)\n",
    "                    tmp_path = os.path.join(data_path, 'KO_classifier', 'all', 'interpretation', 'enrichment_analysis', 'stat', edge_desc, db, 'Enrichr_{}_{}.csv'.format(db, edge_desc))\n",
    "\n",
    "                    # check if exists\n",
    "                    if not(os.path.exists(tmp_path)):\n",
    "                        continue\n",
    "                    \n",
    "                    # load enrichment results\n",
    "                    df_tmp = pd.read_csv(tmp_path)\n",
    "            \n",
    "                    # Drop columns\n",
    "                    df_tmp = df_tmp.drop(['Unnamed: 0', 'Gene_set'], axis=1)\n",
    "\n",
    "                    # add column\n",
    "                    df_tmp[\"KO 1\"] = KO1\n",
    "                    df_tmp[\"Time point 1\"] = cond1\n",
    "                    df_tmp[\"KO 2\"] = KO2\n",
    "                    df_tmp[\"Time point 2\"] = cond2\n",
    "                    df_tmp[\"Direction\"] = direction\n",
    "            \n",
    "                    # set indices\n",
    "                    df_tmp.set_index([\"Time point 1\", \"KO 1\", \"Time point 2\", \"KO 2\", \"Direction\", \"Term\"], inplace=True)\n",
    "\n",
    "                    enr_dfs.append(df_tmp)\n",
    "\n",
    "# dea_df = pd.concat(dea_dfs, axis=1)\n",
    "enr_df = pd.concat(enr_dfs, axis=0)\n",
    "\n",
    "# add database column\n",
    "enr_df[\"Database\"] = db\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = enr_df,\n",
    "                     sheet_name = \"joint_graph_edge_enrichments\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "print(enr_df.shape)\n",
    "enr_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "9c5a54d9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(288, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Listeria Cluster</th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Genes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Time point</th>\n",
       "      <th>KO</th>\n",
       "      <th>Direction</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">Untreated</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">Spi1</th>\n",
       "      <th>up</th>\n",
       "      <td>B</td>\n",
       "      <td>4/474</td>\n",
       "      <td>0.143547</td>\n",
       "      <td>0.358868</td>\n",
       "      <td>2.273010</td>\n",
       "      <td>S100A4;TMEM176A;TMEM176B;PDE4B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>up</th>\n",
       "      <td>C</td>\n",
       "      <td>2/244</td>\n",
       "      <td>0.278980</td>\n",
       "      <td>0.467925</td>\n",
       "      <td>2.415019</td>\n",
       "      <td>CTNNB1;CCL5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>up</th>\n",
       "      <td>F</td>\n",
       "      <td>1/802</td>\n",
       "      <td>0.970034</td>\n",
       "      <td>0.970034</td>\n",
       "      <td>0.420999</td>\n",
       "      <td>RPL21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>up</th>\n",
       "      <td>D</td>\n",
       "      <td>10/1260</td>\n",
       "      <td>0.040454</td>\n",
       "      <td>0.199170</td>\n",
       "      <td>2.082048</td>\n",
       "      <td>GSTM1;H2-Q6;H2-T22;H2-Q4;CTSL;FBXO32;LY6E;MGST...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>up</th>\n",
       "      <td>E</td>\n",
       "      <td>2/553</td>\n",
       "      <td>0.687566</td>\n",
       "      <td>0.747354</td>\n",
       "      <td>1.047277</td>\n",
       "      <td>KANK3;PNRC1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Listeria Cluster  Overlap   P-value  \\\n",
       "Time point KO   Direction                                       \n",
       "Untreated  Spi1 up                       B    4/474  0.143547   \n",
       "                up                       C    2/244  0.278980   \n",
       "                up                       F    1/802  0.970034   \n",
       "                up                       D  10/1260  0.040454   \n",
       "                up                       E    2/553  0.687566   \n",
       "\n",
       "                           Adjusted P-value  Odds Ratio  \\\n",
       "Time point KO   Direction                                 \n",
       "Untreated  Spi1 up                 0.358868    2.273010   \n",
       "                up                 0.467925    2.415019   \n",
       "                up                 0.970034    0.420999   \n",
       "                up                 0.199170    2.082048   \n",
       "                up                 0.747354    1.047277   \n",
       "\n",
       "                                                                       Genes  \n",
       "Time point KO   Direction                                                     \n",
       "Untreated  Spi1 up                            S100A4;TMEM176A;TMEM176B;PDE4B  \n",
       "                up                                               CTNNB1;CCL5  \n",
       "                up                                                     RPL21  \n",
       "                up         GSTM1;H2-Q6;H2-T22;H2-Q4;CTSL;FBXO32;LY6E;MGST...  \n",
       "                up                                               KANK3;PNRC1  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Enrichment analysis results of KO perturbation signatures in temporal deviating gene clusters\n",
    "cluster_to_letter = {\"LO28_cluster_1\": 'A', \"LO28_cluster_2\": 'B', \"LO28_cluster_3\": 'C', \"LO28_cluster_4\": 'D', \"LO28_cluster_5\": 'E', \"LO28_cluster_6\": 'F', \"LO28_cluster_7\": 'G'}\n",
    "listeria_changes = {\n",
    "    'G': 'A',\n",
    "    'A': 'B',\n",
    "    'B': 'C',\n",
    "    'D': 'D',\n",
    "    'E': 'E',\n",
    "    'C': 'F',\n",
    "    'F': 'G',\n",
    "}\n",
    "\n",
    "tmp_path = os.path.join(data_path, 'special_analysis', 'INT_comparison', 'GSEA_split_results.csv')\n",
    "\n",
    "df_tmp = pd.read_csv(tmp_path)\n",
    "\n",
    "# remove all but Listeria clusters\n",
    "df_tmp = df_tmp[df_tmp['Term'].str.contains('LO28')]\n",
    "\n",
    "# adapt cluster labels to letters\n",
    "df_tmp[\"Term\"] = df_tmp[\"Term\"].replace(cluster_to_letter)\n",
    "df_tmp[\"Term\"] = df_tmp[\"Term\"].replace(listeria_changes)\n",
    "\n",
    "# split columns\n",
    "df_tmp[['Time point', 'KO', 'Direction']] = df_tmp['Gene_set'].str.split(\"_\", expand=True)\n",
    "\n",
    "# Drop columns\n",
    "df_tmp = df_tmp.drop(['Unnamed: 0', 'Odds Ratio1', 'Gene_set'], axis=1)\n",
    "\n",
    "# rename column \n",
    "df_tmp.rename(columns={'Term': 'Listeria_Cluster'}, inplace=True)\n",
    "\n",
    "# set indices\n",
    "df_tmp.set_index(['Time point', 'KO', 'Direction'], inplace=True)\n",
    "\n",
    "# add dataframe as sheet to excel file\n",
    "writer = df_to_excel(writer = writer,\n",
    "                     df = df_tmp,\n",
    "                     sheet_name = \"KO_effects_enriched_in_Listeria\",\n",
    "                     column_width = 25,\n",
    "                     index_format = index_format\n",
    "                    )\n",
    "\n",
    "print(df_tmp.shape)\n",
    "df_tmp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "d7d17e4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save Excel file\n",
    "writer.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f940184",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python_basics",
   "language": "python",
   "name": "basics"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
